---
title: "Population genomics in practice"
subtitle: "What is population genomics?"
description: >-
  Discussion of @fuller_PopulationGeneticsCoral_2020 to introduce
  concepts in population genomics.
author:
  - "Per Unneberg"
format:
  nbis-quarto-revealjs:
    toc: false
    smaller: false
---

## Setup {.hidden .unnumbered .unlisted visibility="hidden"}

{{< include /slides/_knitr.qmd >}}

{{< include /slides/_rlibs.qmd >}}

```{r libs}
#| echo: false
#| eval: true
#| cache: false
library(readxl)
library(curl)
```

## Intended learning outcomes

### Course

- Present minimum toolkit of methods that should be known to anyone
  starting out in population genomics
- Sufficiently small for one-week workshop

### Lecture

- Present practical example of toolkit as applied in
  [@fuller_PopulationGeneticsCoral_2020]
- Briefly discuss baseline model
  [@johri_RecommendationsImprovingStatistical_2022]

::: {.notes }

Aim of lecture is to:

- present a practical application of commonly used methods in
  population genomics
- link population genomics to population genetics
- discuss statistical inference and the need of a baseline model with
  which to compare observations and conclusions

#### What is population genomics?

Points from [@hahn_MolecularPopulationGenetics_2019, pp. 249-250]:

- whole-genome data instead of single loci - population genomics is
  population genetics for whole-genome sequences
  - if only this, not too exciting
- major promise: enables analyses not possible for single loci or that
  require genomic context
- addresses interactions between different forces, notably selection
  and demographic history

#### Some applications

- genome-wide scans for selection
  - selection vs demography (p. 251)
- methods for genome-wide scans (p. 258)

#### Caveats

- non-independence (p. 267)
  - different statistics rely on similar input
  - overlapping peaks from different statistics *not* independent

#### General points

[@hartl1997principles p. 469-470]:

- more emphasis on differences within populations
- goal: understand differences among genomes -> requires complete
  sequence data from multiple individuals

[@li_InferenceHumanPopulation_2011, supplementary notes, p. 6] on the
use of PSMC on autosomes:

"...highly consistent except for the very recent history,
demonstrating the power of using whole-genome data."

:::

## Example: Population genetics of the coral *Acropora millepora*

:::: {.columns}

::: {.column width="50%"}

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

::: {.fragment fragment-index=1}

#### **Genome assembly** and **sampling**

Motivation: most analyses require a *reference sequence* with which to
compare *resequenced samples*

- Assemble high-quality reference genome
- Choice of populations, sampling locations

:::

:::

::: {.column width="50%"}

::: {.fragment fragment-index=1}

:::{#fig-corals-assembly}

<!-- markdownlint-disable MD013 -->

![](https://www.science.org/cms/10.1126/science.aba4674/asset/31658ca3-933a-4047-81e1-e900fd9a0bae/assets/graphic/369_aba4674_f1.jpeg){width=75% fig-align=center}

<!-- markdownlint-enable MD013 -->

Genome assembly by hybrid sequencing (A) 253 individuals from 12 reefs
(B) PCA with environmental and spatial variables (C) phenotype
distributions (D)

:::

::: { .flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

:::

:::

::::

::: {.notes }

[@fuller_PopulationGeneticsCoral_2020] is an example of a population
genomics study that applies methods that could be seen as a basic
foundation of population genomics. We believe these present a minimum
toolkit of methods that should be known to anyone starting out in
population genomics, and that is sufficiently small to be presented in
a one-week workshop. At the end of this lecture, we will discuss some
more advanced applications in population genomics.

Figure caption:

> Genome assembly and sample collection for A. millepora. (A) A de
> novo assembly for the A. millepora genome was constructed by using a
> hybrid sequencing approach. The alignment of reads generated from
> two pools of aposymbiotic larvae was used to filter out symbiont
> contigs (fig. S3). Assembled coral contigs were aligned to
> previously published linkage maps (19, 80) to create a
> chromosome-scale assembly. (B) A total of 253 individuals were
> collected from across 12 reefs on the GBR in 2017. The size of the
> circles represents the number of individuals collected at each site,
> and the circles are colored according to the bleaching severity at
> each reef. The maximum degree heating weeks (DHW) is shown across
> the region from which samples were collected (and across the GBR in
> the inset). The spatial layer of DHW represents interpolated maximum
> values from the National Oceanic and Atmospheric Administration
> Coral Reef Watch (CRW) v3.1 satellite product at a resolution of 5
> km. Each reef label is colored arbitrarily but consistent with
> labels presented in other figures. AN, Arlington Reef; FY, Fitzroy
> Reef; RL, Russell Island; CS, Coates Reef; FR, Feather Reef; NB,
> North Barnard Islands; TR, Taylor Reef; DK, Dunk Island; RB, Rib
> Reef; PA, Pandora Reef; JB, John Brewer Reef; HH, Havannah Island.
> (C) A PCA was performed for 40 environmental and spatial variables
> for each reef, with abbreviations in bold representing their
> location. The component loads for each environmental variable
> projected on the first two principal components are depicted with
> arrows. (D) The distribution of visual scores, chlorophyll abundance
> standardized by host coral protein content, and standardized
> symbiont cell densities among the collected individuals for which
> phenotype measurement was possible. A visual score of 1 indicates
> severe bleaching, whereas a score of 6 represents full
> pigmentation.at a resolution of 5 km. Each reef label is colored
> arbitrarily but consistent with labels presented in other figures.
> AN, Arlington Reef; FY, Fitzroy Reef; RL, Russell Island; CS, Coates
> Reef; FR, Feather Reef; NB, North Barnard Islands; TR, Taylor Reef;
> DK, Dunk Island; RB, Rib Reef; PA, Pandora Reef; JB, John Brewer
> Reef; HH, Havannah Island. (C) A PCA was performed for 40
> environmental and spatial variables for each reef, with
> abbreviations in bold representing their location. The component
> loads for each environmental variable projected on the first two
> principal components are depicted with arrows. (D) The distribution
> of visual scores, chlorophyll abundance standardized by host coral
> protein content, and standardized symbiont cell densities among the
> collected individuals for which phenotype measurement was possible.
> A visual score of 1 indicates severe bleaching, whereas a score of 6
> represents full pigmentation.

**Genome assembly and sampling**

Why: most analyses require a *reference sequence* with which to
compare *resequenced samples*

Points to consider:

- choice of reference individual
- the number of populations
- the number of samples (more sites better than many samples per
     population)
- the geographical distribution of samples
- sequencing depth (low-coverage often sufficient)

:::

## Example: Population genetics of the coral *Acropora millepora*

:::: {.columns}

::: {.column width="50%"}

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

#### Describe **genetic structure** and **demographic history**

Motivation:

- address basic question of why genetic structure looks the way it
  does
- demographic history may generate signals similar to selection

:::

::: {.column width="50%"}

<!-- markdownlint-disable MD013 -->

:::{#fig-corals-variation}

![](https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg){width=60% fig-align=center}

Variation and demographic history inferred from 44 resequenced
individuals. LD decay ($r^2$) from 1% of markers (A) nucleotide
diversity ($\pi$) in 1kb-windows (B) effective population sizes ({{<
var Ne >}}) estimated with PSMC (C)

:::

<!-- markdownlint-enable MD013 -->

::: {.flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

:::

::::

::: {.notes }

Figure caption:

> Properties of genetic variation and inferred demographic history in
> sampled A. millepora. (A) The decay of LD, measured as the squared
> genotypic correlation coefficient (r2), was estimated for a randomly
> chosen 1% of SNPs by using all sequenced samples. Each point
> represents the mean r2 in bins of 100 bp. (B) Nucleotide diversity
> was measured by the average pairwise differences per base pair (p)
> (86) for SNPs in 1-kb windows, using intergenic regions genome-wide.
> (C) Effective population sizes (Ne) inferred from 44 sequenced A.
> millepora genomes by using PSMC (35), assuming a mutation rate m of
> 4 × 10−9 per base pair per generation (20, 34). Although an error in
> the estimate of m will lead to a rescaling of both axes, it will not
> affect the qualitative conclusion of a decline in Ne toward the
> present. The x axis is scaled in terms of generations. This approach
> provides very little information about the recent past (here, less
> than ~104 generations).

**Variation and demographic history**

Why: summarizing *diversity* provides (indirect) information on
population size and more, as does the *linkage structure*. Estimate
*demographic history* since fluctuating population size may produce
signals similar to those of selection

- LD decay: important for imputation (e.g.,
  stephens_AccountingDecayLinkage_2005) and setting window size for
  genome scans, where a common rule of thumb is to set the size larger
  than the genome background: this ensures windows are, in some sense,
  independent
- "The extent of LD and its decay with genetic distance are useful
  parameters for determining the number of markers needed to
  successfully map a QTL, and the resolution with which the trait can
  be successfully mapped"
  [otyama_EvaluationLinkageDisequilibrium_2019]
- 0.363% average pi, but large variation.
- many psmc plots show decline in population size, which could be an
  effect of bottleneck during pleistocene. Also population divergence
  (ghost ancestral populations, splits, extinction) can affect
  population size
- in aDNA studies missingness is common (i.e., heterozygotes are
  underestimated) and has to be accounted for since coalescence times
  are affected and may influence estimate of population size

:::

## Example: Population genetics of the coral *Acropora millepora*

:::: {.columns}

::: {.column width="50%"}

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

#### Characterize **population structure**

Motivation:

1. identify populations for contrasts in e.g. selection scans
2. identify admixed individuals that should be removed from analyses
3. identify barriers to gene flow etc

:::

::: {.column width="50%"}

:::{#fig-corals-structure}

<!-- markdownlint-disable MD013 -->

![](https://www.science.org/cms/10.1126/science.aba4674/asset/4625c5a6-1e71-4b9f-9589-0fed7ab72d3d/assets/graphic/369_aba4674_f3.jpeg){width=100% fig-align=center}

<!-- markdownlint-enable MD013 -->

Characterizing population structure and gene flow across 12 refs. {{<
var fst >}} measure across geographic distance (A) PCA from LD-pruned
genome-wide SNPs for 44 resequenced samples (B) estimation of relative
effective migration surface from LD-pruned and common (MAF > 5%) SNPs
(C)

:::

::: { .flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

:::

::::

::: {.notes }

> Characterizing population structure and gene flow across 12 reefs.
> (A) There is no discernable relationship between geographic distance
> and genetic differentiation (measured as FST) across pairwise
> comparisons of sampled reefs. (B) A PCA by using LD-pruned genome-wide
> SNPs for the 44 resequenced high-coverage genomes. Each point is
> color-coded by the reef from which the individual was sampled. (C)
> EEMS (39) was used to estimate and visualize the relative effective
> migration surface by using LD-pruned and common (MAF > 5%) genome-wide
> SNPs for the 44 resequenced high-coverage genomes. The size of the
> dots reflects the number of individuals sequenced from each reef.

**Population structure:**

Why: many reasons: 1) identifying populations for contrasts in e.g.
selection scans 2) identify admixed individuals that should be removed
from analyses 3) identify barriers to gene flow etc

- no discernible relationship between geographic distance and genetic
  differentiation -> gene flow
  - for this reason, Fst between populations is low
- EEMS (Estimated Effective Migration Surfaces) models relationship
  between genetics and geography
  petkova_VisualizingSpatialPopulation_2016

Indicative of high connectivity among 12 sampled reefs.

:::

## Example: Population genetics of the coral *Acropora millepora*

:::: {.columns}

::: {.column width="50%"}

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

#### Genomic scans for **selection**

Motivation: identify loci associated with adaptation / selection

- little differentiation over reefs, however thermal regimes
- genomic scan for $\pi$ (diversity) outliers

:::

::: {.column width="50%"}

:::{#fig-corals-selection}

<!-- markdownlint-disable MD013 -->

![](https://www.science.org/cms/10.1126/science.aba4674/asset/83167ee1-c1f3-4ef8-88b3-d69a0cb93062/assets/graphic/369_aba4674_f4.jpeg){width=80% fig-align=center}

<!-- markdownlint-enable MD013 -->

Genomic scans for local adaptation detect a signal at *sacsin*.
Nucleotide diversity ($\pi$) in 1kb-window, values in top 0.01%
genome-wide in red (A, top). Close-up view around *sacsin* gene, with
predicted gene structure above (A, bottom). h12 summary statistic
showing the frequence of the two most common haplotypes (B). *sacsin*
gene tree (blue) together with random gene trees, indicating deep
genealogy at *sacsin* (C)

:::

::: { .flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

:::

::::

::: {.notes }

> Genomic scans for local adaptation detect a signal at sacsin. (A)
> (Top) Values of pairwise nucleotide diversity (π) (86) estimated in
> sliding windows of 1 kb and a step-size of 500 bp across chromosome
> 7, with points colored in red indicating regions in the top 0.01% of
> values genome-wide. A loess-smoothed trend line of π across windows
> is indicated in purple. The peak with the most extreme values falls
> in the sacsin gene region. (Bottom) A close-up view of π estimated
> across a region of chromosome 7 surrounding the sacsin gene, which
> includes all of the outlier windows indicated in red at top. π per
> base pair was calculated in sliding windows of 100 bp (gray dots);
> the loess-smoothed line is shown in blue. The predicted gene
> structure of sacsin is indicated above, as well as the flanking
> upstream and downstream genes. The 29 nonsynonymous differences
> fixed between samples of the two most common haplotypes in sacsin
> are denoted with green lines. There are also 20 fixed synonymous
> differences. Gray vertical dashed lines delimit a region that was
> masked from variant calling because of predicted repetitive
> elements. (B) The h12 summary statistic (47), which measures the
> frequency of the two most common haplotypes, is plotted across
> chromosome 7 (supplementary materials, materials and methods). Red
> dots represent h12 values in the top 0.01% genome-wide. The peak
> with the most extreme values falls in the sacsin gene region. (C) A
> gene tree for the central 1-kb region in sacsin is shown in dark
> blue; it was constructed for one randomly chosen haplotype from a
> randomly selected individual from each of the 12 sampled reefs. The
> gene tree is rooted to aligned sequences from the reference genomes
> for A. digitifera (87) and A. tenuis (draft assembly from
> <www.reefgenomics.org>). Shown in gray are gene trees for 1000
> randomly sampled 1-kb regions from across the genome for the same
> individuals. Each gene tree was inferred by using maximum likelihood
> implemented in dnaml (88).

**Selection scan**

Why: identify loci associated with adaptation / selection, which
provides potential mechanisms for adaptation, as well as information
that could be important for conservation strategies

Little differentiation across reefs -> little population structure
over hundreds of kilometers. However, there are environmental
differences (thermal regimes). Scan for pi outliers:

- points to *sacsin* gene
- h12 measures the frequency of the two most common haplotypes; red
  indicate 0.01% outlier genome-wide
- 4C: tree for central 1kb region in *sacsin* deeper than split from
  *A.digitifera* and *A.tenuis*
  - variation in *sacsin* has been maintained for long time
  - co-chaperone for heat-shock protein Hsp70

:::

## Example: Population genetics of the coral *Acropora millepora*

Study highlights common analyses in population genomics study:

::: {.incremental}

1. Genome assembly, resequencing, variant calling and filtering
2. Description of variation (e.g., $\pi$) and genetic structure (LD)
3. Description of population structure (admixture, PCA)
4. Modelling of demographic history (PSMC)
5. Genome scans for adaptive traits

:::

## Population genetics

:::: {style="display: grid; grid-template-columns: 300px 200px 200px;
grid-row-gap: 0px; grid-column-gap: 100px;"}

<div>&nbsp;</div>

:::{.circular-image}

Mutation

![](assets/images/mutation.webp)

:::

:::{.circular-image}

Selection

![](assets/images/finches.webp)

:::

<div>&nbsp;</div>

:::{.circular-image}

Recombination

![](assets/images/recombination.webp)

:::

:::{.circular-image}

Drift

![](assets/images/genetic-drift.webp)

:::

::::

::: {.notes }

[@fuller_PopulationGeneticsCoral_2020] paper has population genetics
in title -> population genetics is a key ingredient.

Population genetics focuses on the genetic basis of evolution. It is
mainly a theoretical subject, owing to the slow changes of genetic
variation. As such, it tries to explain the shape and structure of
genetic variation from theoretical predictions and models.

:::

## From population genetics to population genomics

<!-- markdownlint-disable MD013 -->

:::{}

![The variable sites at the *Drosophila melanogaster* ADH locus
[@kreitman_NucleotidePolymorphismAlcohol_1983]](assets/images/kreitman-1983-table1.webp){width=70% fig-align=center}

:::

<!-- markdownlint-enable MD013 -->

First study of natural population. However, limited to one locus.

::: {.notes }

- from locus-based studies (e.g., alcohol dehydrogenase in Drosophila
  [@kreitman_NucleotidePolymorphismAlcohol_1983]) to genome-wide
  (e.g., Drosophila population genomics
  [@begun_PopulationGenomicsWholeGenome_2007]
- note: studied loci have often not been randomly chosen, which is
  another argument for whole-genome studies
- enabler: sequencing technology

[@fuller_PopulationGeneticsCoral_2020] paper has population genetics
in title -> population genetics is a key ingredient.

Refer to Hahn's points about learning something about global patterns:

- selection acts locally, demography globally
- the structure of genetic variation and how it depends on
  - recombination landscapes and linked selection
  - demographic changes
  - identification of neutral loci

So, not simply about applying 10000 selection tests for multiple loci

All of the points above point to the importance of statistics which
implies mathematics / computational skills important

:::

## From population genetics to population genomics

:::{}

![Patterns of polymorphism and divergence [@begun_PopulationGenomicsWholeGenome_2007]](https://journals.plos.org/plosbiology/article/figure/image?size=large&download=&id=10.1371/journal.pbio.0050310.g001){width=60%}

:::

Same system but **genome-wide**. Plots represent all chromosomes and the entire genome.

::: {.notes}

@begun_PopulationGenomicsWholeGenome_2007 study: same system
(*Drosophila*) but more individuals and whole genome. All of a sudden
possible to ask questions about the general characteristics of
diversity, not just limited to single loci.

:::

## From population genetics to population genomics

:::{}

<!-- markdownlint-disable MD013 -->

![Numbers of polymorphic and fixed variants [@begun_PopulationGenomicsWholeGenome_2007]](https://journals.plos.org/plosbiology/article/figure/image?size=large&download=&id=10.1371/journal.pbio.0050310.t002){width=70% fig-align=center}

<!-- markdownlint-enable MD013 -->

:::

Novelty: now possible to do *genome-wide* characterization of
variation in different **functional contexts**

::: {.notes}

Novelty: now possible to do *genome-wide* characterization of
variation in different functional contexts

:::

## The technological revolution in **sequencing** and **computing**

:::: {.columns}

::: {.column width="50%"}

```{r }
#| label: fig-sequencing-cost
#| echo: false
#| eval: true
#| fig-cap: Sequencing cost ($) per megabase [@wetterstrandka_DNASequencingCosts]
#| fig-width: 12
#| fig-height: 8
#| out-width: 90%
url <- "https://www.genome.gov/sites/default/files/media/files/2021-11/Sequencing_Cost_Data_Table_Aug2021.xls"
if (!file.exists(basename(url))) {
  download.file(url, destfile = basename(url))
}
data <- read_excel(basename(url))
colnames(data) <- c("Date", "Mb", "Genome")
ggplot(data, aes(x = as.Date(Date), y = Mb)) +
  geom_line(size=2) +
  xlab("") +
  ylab("") +
  scale_y_continuous(trans = "log10", labels = comma, n.breaks = 6) +
  scale_x_date(breaks = "year", labels = date_format("%Y")) +
  theme(axis.text=element_text(size=24), text=element_text(size=24))
```

:::

::: {.column width="50%"}

![Moore's law](https://ourworldindata.org/uploads/2022/03/Transistor-Count-over-time.png)

:::

::::

## Statistical inference in population genomics

The data deluge requires advanced statistical methods and models to do
inference. Today data production outpaces theoretical advances.
Therefore, take care not to attach too much faith to a test that
explains data well.

::: {.fragment}

A population genomics study should aim at generating a baseline model
that takes into account the processes that shape genetic variation
[@johri_RecommendationsImprovingStatistical_2022]:

1. mutation
2. recombination
3. gene conversion
4. purifying selection acting on functional regions and its effects on
   linked variants (background selection)
5. genetic drift with demographic history and geographic structure

:::

::: {.notes}

Caution against adaptationist storytelling; always compare to a
baseline model that takes potential confounding factors into account

:::

## Applications of population genomics

:::: {style="display: grid; width: 100%; grid-template-areas: \"a b c\" \"d e f\";
grid-row-gap: 10px; grid-column-gap: 30px; align-items: center;"}

:::{.fragment}

![Conservation genomics [@webster_PopulationGenomicsInsect_2023]](https://www.annualreviews.org/docserver/fulltext/animal/11/1/av110115.f2.gif){height=200px}

:::

:::{.fragment}

![Speciation genomics
[@stankowski_WidespreadSelectionGene_2019]](/exercises/datasets/assets/images/stankowski-2019-fig1A.webp){height=200px}

:::

:::{.fragment}

![disentangle forces that create variation [@rodrigues_SharedEvolutionaryProcesses_2024]](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/genetics/226/4/10.1093_genetics_iyae006/2/m_iyae006f1.jpeg?Expires=1760610839&Signature=nii-RdEtrwqeLzTWwajHQCTXC~YJPndas9Soarpbckb8xbZswk9gmaWJDenpDMXfQc-p6-Pre65c11i6MJFbtuiEkWZyZzt3pE1vONLeGZdU~5RV4Q5UykIsYlfd~pnGiJC5mkTAQOp1OGW3RXkO0EfTbBly-~hMj627Pi9Ms6IipXxCWhVwzpuI9YlRkmqsc0HRNPFwf4cHyQS0SlgcZkCkEgJWbrthl~Ro9LuKB6bKTkIeECPKk3-7vzB2X~4X3ly3H5ce-xsFqZOnhGDCMk23B8d4wkrgBuJc7aeewJaiK~-Y7aGV7y-lNhBR-k~6whfW8l~F4-SYNhPonjr2dg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA){height=200px}

:::

:::{.fragment}

![paleogenomics (aDNA) [@vandervalk_MillionyearoldDNASheds_2021]](https://media.springernature.com/lw685/springer-static/esm/art%3A10.1038%2Fs41586-021-03224-9/MediaObjects/41586_2021_3224_Fig8_ESM.jpg?){height=200px}

:::

:::{.fragment}

![domestication [@barrera-redondo_GenomicTranscriptomicEpigenomic_2020]](https://www.frontiersin.org/files/Articles/555530/fgene-11-00742-HTML/image_m/fgene-11-00742-g002.jpg){height=200px}

:::

:::{.fragment}

![ecology [@unneberg_EcologicalGenomicsNorthern_2024]](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41467-024-50239-7/MediaObjects/41467_2024_50239_Fig1_HTML.png?){height=200px}

:::

::::

## Bibliography {.unnumbered .unlisted .smaller}

::: { #refs }
:::
