<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Population genomics in practice</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-a9a275cbfa7b8981755d3f45fb7f152b.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Population genomics in practice</h1>
  <p class="subtitle">What is population genomics?</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="intended-learning-outcomes" class="slide level2 ultitle">
<h2>Intended learning outcomes</h2>
<h3 class="ultitle" id="course">Course</h3>
<ul>
<li>Present minimum toolkit of methods that should be known to anyone starting out in population genomics</li>
<li>Sufficiently small for one-week workshop</li>
</ul>
<h3 class="ultitle" id="lecture">Lecture</h3>
<ul>
<li>Present practical example of toolkit as applied in <span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Fuller et al., 2020</a>)</span></li>
<li>Briefly discuss baseline model <span class="citation" data-cites="johri_RecommendationsImprovingStatistical_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Johri et al., 2022</a>)</span></li>
</ul>
<aside class="notes">
<p>Aim of lecture is to:</p>
<ul>
<li>present a practical application of commonly used methods in population genomics</li>
<li>link population genomics to population genetics</li>
<li>discuss statistical inference and the need of a baseline model with which to compare observations and conclusions</li>
</ul>
<h4 id="what-is-population-genomics">What is population genomics?</h4>
<p>Points from <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, pp. 249–250</a>)</span>:</p>
<ul>
<li>whole-genome data instead of single loci - population genomics is population genetics for whole-genome sequences
<ul>
<li>if only this, not too exciting</li>
</ul></li>
<li>major promise: enables analyses not possible for single loci or that require genomic context</li>
<li>addresses interactions between different forces, notably selection and demographic history</li>
</ul>
<h4 id="some-applications">Some applications</h4>
<ul>
<li>genome-wide scans for selection
<ul>
<li>selection vs demography (p.&nbsp;251)</li>
</ul></li>
<li>methods for genome-wide scans (p.&nbsp;258)</li>
</ul>
<h4 id="caveats">Caveats</h4>
<ul>
<li>non-independence (p.&nbsp;267)
<ul>
<li>different statistics rely on similar input</li>
<li>overlapping peaks from different statistics <em>not</em> independent</li>
</ul></li>
</ul>
<h4 id="general-points">General points</h4>
<p><span class="citation" data-cites="hartl1997principles">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hartl &amp; Clark, 1997, pp. 469–470</a>)</span>:</p>
<ul>
<li>more emphasis on differences within populations</li>
<li>goal: understand differences among genomes -&gt; requires complete sequence data from multiple individuals</li>
</ul>
<p><span class="citation" data-cites="li_InferenceHumanPopulation_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Li &amp; Durbin, 2011</a>, supplementary notes, p.&nbsp;6)</span> on the use of PSMC on autosomes:</p>
<p>“…highly consistent except for the very recent history, demonstrating the power of using whole-genome data.”</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-population-genetics-of-the-coral-acropora-millepora" class="slide level2 ultitle">
<h2>Example: Population genetics of the coral <em>Acropora millepora</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Motivation: corals are facing hard times and to prevent future losses of coral cover a better understanding of genetics is warranted.</p>
<div class="fragment" data-fragment-index="1">
<h4 id="genome-assembly-and-sampling"><strong>Genome assembly</strong> and <strong>sampling</strong></h4>
<p>Motivation: most analyses require a <em>reference sequence</em> with which to compare <em>resequenced samples</em></p>
<ul>
<li>Assemble high-quality reference genome</li>
<li>Choice of populations, sampling locations</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment" data-fragment-index="1">
<div id="fig-corals-assembly" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-corals-assembly-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.science.org/cms/10.1126/science.aba4674/asset/31658ca3-933a-4047-81e1-e900fd9a0bae/assets/graphic/369_aba4674_f1.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;1: Genome assembly by hybrid sequencing (A) 253 individuals from 12 reefs (B) PCA with environmental and spatial variables (C) phenotype distributions (D)"><img data-src="https://www.science.org/cms/10.1126/science.aba4674/asset/31658ca3-933a-4047-81e1-e900fd9a0bae/assets/graphic/369_aba4674_f1.jpeg" class="quarto-figure quarto-figure-center" style="width:75.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-corals-assembly-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Genome assembly by hybrid sequencing (A) 253 individuals from 12 reefs (B) PCA with environmental and spatial variables (C) phenotype distributions (D)
</figcaption>
</figure>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span></p>
</div>
</div>
</div></div>
<aside class="notes">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Fuller et al., 2020</a>)</span> is an example of a population genomics study that applies methods that could be seen as a basic foundation of population genomics. We believe these present a minimum toolkit of methods that should be known to anyone starting out in population genomics, and that is sufficiently small to be presented in a one-week workshop. At the end of this lecture, we will discuss some more advanced applications in population genomics.</p>
<p>Figure caption:</p>
<blockquote>
<p>Genome assembly and sample collection for A. millepora. (A) A de novo assembly for the A. millepora genome was constructed by using a hybrid sequencing approach. The alignment of reads generated from two pools of aposymbiotic larvae was used to filter out symbiont contigs (fig.&nbsp;S3). Assembled coral contigs were aligned to previously published linkage maps (19, 80) to create a chromosome-scale assembly. (B) A total of 253 individuals were collected from across 12 reefs on the GBR in 2017. The size of the circles represents the number of individuals collected at each site, and the circles are colored according to the bleaching severity at each reef. The maximum degree heating weeks (DHW) is shown across the region from which samples were collected (and across the GBR in the inset). The spatial layer of DHW represents interpolated maximum values from the National Oceanic and Atmospheric Administration Coral Reef Watch (CRW) v3.1 satellite product at a resolution of 5 km. Each reef label is colored arbitrarily but consistent with labels presented in other figures. AN, Arlington Reef; FY, Fitzroy Reef; RL, Russell Island; CS, Coates Reef; FR, Feather Reef; NB, North Barnard Islands; TR, Taylor Reef; DK, Dunk Island; RB, Rib Reef; PA, Pandora Reef; JB, John Brewer Reef; HH, Havannah Island. (C) A PCA was performed for 40 environmental and spatial variables for each reef, with abbreviations in bold representing their location. The component loads for each environmental variable projected on the first two principal components are depicted with arrows. (D) The distribution of visual scores, chlorophyll abundance standardized by host coral protein content, and standardized symbiont cell densities among the collected individuals for which phenotype measurement was possible. A visual score of 1 indicates severe bleaching, whereas a score of 6 represents full pigmentation.at a resolution of 5 km. Each reef label is colored arbitrarily but consistent with labels presented in other figures. AN, Arlington Reef; FY, Fitzroy Reef; RL, Russell Island; CS, Coates Reef; FR, Feather Reef; NB, North Barnard Islands; TR, Taylor Reef; DK, Dunk Island; RB, Rib Reef; PA, Pandora Reef; JB, John Brewer Reef; HH, Havannah Island. (C) A PCA was performed for 40 environmental and spatial variables for each reef, with abbreviations in bold representing their location. The component loads for each environmental variable projected on the first two principal components are depicted with arrows. (D) The distribution of visual scores, chlorophyll abundance standardized by host coral protein content, and standardized symbiont cell densities among the collected individuals for which phenotype measurement was possible. A visual score of 1 indicates severe bleaching, whereas a score of 6 represents full pigmentation.</p>
</blockquote>
<p><strong>Genome assembly and sampling</strong></p>
<p>Why: most analyses require a <em>reference sequence</em> with which to compare <em>resequenced samples</em></p>
<p>Points to consider:</p>
<ul>
<li>choice of reference individual</li>
<li>the number of populations</li>
<li>the number of samples (more sites better than many samples per population)</li>
<li>the geographical distribution of samples</li>
<li>sequencing depth (low-coverage often sufficient)</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-population-genetics-of-the-coral-acropora-millepora-1" class="slide level2 ultitle">
<h2>Example: Population genetics of the coral <em>Acropora millepora</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Motivation: corals are facing hard times and to prevent future losses of coral cover a better understanding of genetics is warranted.</p>
<h4 id="describe-genetic-structure-and-demographic-history">Describe <strong>genetic structure</strong> and <strong>demographic history</strong></h4>
<p>Motivation:</p>
<ul>
<li>address basic question of why genetic structure looks the way it does</li>
<li>demographic history may generate signals similar to selection</li>
</ul>
</div><div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div id="fig-corals-variation" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-corals-variation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;2: Variation and demographic history inferred from 44 resequenced individuals. LD decay (r^2) from 1% of markers (A) nucleotide diversity (\pi) in 1kb-windows (B) effective population sizes (\mathrm{N_e}) estimated with PSMC (C)"><img data-src="https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg" class="quarto-figure quarto-figure-center" style="width:60.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-corals-variation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Variation and demographic history inferred from 44 resequenced individuals. LD decay (<span class="math inline">r^2</span>) from 1% of markers (A) nucleotide diversity (<span class="math inline">\pi</span>) in 1kb-windows (B) effective population sizes (<span class="math inline">\mathrm{N_e}</span>) estimated with PSMC (C)
</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<div class="flushright smallest">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span></p>
</div>
</div></div>
<aside class="notes">
<p>Figure caption:</p>
<blockquote>
<p>Properties of genetic variation and inferred demographic history in sampled A. millepora. (A) The decay of LD, measured as the squared genotypic correlation coefficient (r2), was estimated for a randomly chosen 1% of SNPs by using all sequenced samples. Each point represents the mean r2 in bins of 100 bp. (B) Nucleotide diversity was measured by the average pairwise differences per base pair (p) (86) for SNPs in 1-kb windows, using intergenic regions genome-wide. (C) Effective population sizes (Ne) inferred from 44 sequenced A. millepora genomes by using PSMC (35), assuming a mutation rate m of 4 × 10−9 per base pair per generation (20, 34). Although an error in the estimate of m will lead to a rescaling of both axes, it will not affect the qualitative conclusion of a decline in Ne toward the present. The x axis is scaled in terms of generations. This approach provides very little information about the recent past (here, less than ~104 generations).</p>
</blockquote>
<p><strong>Variation and demographic history</strong></p>
<p>Why: summarizing <em>diversity</em> provides (indirect) information on population size and more, as does the <em>linkage structure</em>. Estimate <em>demographic history</em> since fluctuating population size may produce signals similar to those of selection</p>
<ul>
<li>LD decay: important for imputation (e.g., stephens_AccountingDecayLinkage_2005) and setting window size for genome scans, where a common rule of thumb is to set the size larger than the genome background: this ensures windows are, in some sense, independent</li>
<li>“The extent of LD and its decay with genetic distance are useful parameters for determining the number of markers needed to successfully map a QTL, and the resolution with which the trait can be successfully mapped” [otyama_EvaluationLinkageDisequilibrium_2019]</li>
<li>0.363% average pi, but large variation.</li>
<li>many psmc plots show decline in population size, which could be an effect of bottleneck during pleistocene. Also population divergence (ghost ancestral populations, splits, extinction) can affect population size</li>
<li>in aDNA studies missingness is common (i.e., heterozygotes are underestimated) and has to be accounted for since coalescence times are affected and may influence estimate of population size</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-population-genetics-of-the-coral-acropora-millepora-2" class="slide level2 ultitle">
<h2>Example: Population genetics of the coral <em>Acropora millepora</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Motivation: corals are facing hard times and to prevent future losses of coral cover a better understanding of genetics is warranted.</p>
<h4 id="characterize-population-structure">Characterize <strong>population structure</strong></h4>
<p>Motivation:</p>
<ol type="1">
<li>identify populations for contrasts in e.g.&nbsp;selection scans</li>
<li>identify admixed individuals that should be removed from analyses</li>
<li>identify barriers to gene flow etc</li>
</ol>
</div><div class="column" style="width:50%;">
<div id="fig-corals-structure" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-corals-structure-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.science.org/cms/10.1126/science.aba4674/asset/4625c5a6-1e71-4b9f-9589-0fed7ab72d3d/assets/graphic/369_aba4674_f3.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;3: Characterizing population structure and gene flow across 12 refs. F_{\mathrm{ST}} measure across geographic distance (A) PCA from LD-pruned genome-wide SNPs for 44 resequenced samples (B) estimation of relative effective migration surface from LD-pruned and common (MAF > 5%) SNPs (C)"><img data-src="https://www.science.org/cms/10.1126/science.aba4674/asset/4625c5a6-1e71-4b9f-9589-0fed7ab72d3d/assets/graphic/369_aba4674_f3.jpeg" class="quarto-figure quarto-figure-center" style="width:100.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-corals-structure-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Characterizing population structure and gene flow across 12 refs. <span class="math inline">F_{\mathrm{ST}}</span> measure across geographic distance (A) PCA from LD-pruned genome-wide SNPs for 44 resequenced samples (B) estimation of relative effective migration surface from LD-pruned and common (MAF &gt; 5%) SNPs (C)
</figcaption>
</figure>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span></p>
</div>
</div></div>
<aside class="notes">
<blockquote>
<p>Characterizing population structure and gene flow across 12 reefs. (A) There is no discernable relationship between geographic distance and genetic differentiation (measured as FST) across pairwise comparisons of sampled reefs. (B) A PCA by using LD-pruned genome-wide SNPs for the 44 resequenced high-coverage genomes. Each point is color-coded by the reef from which the individual was sampled. (C) EEMS (39) was used to estimate and visualize the relative effective migration surface by using LD-pruned and common (MAF &gt; 5%) genome-wide SNPs for the 44 resequenced high-coverage genomes. The size of the dots reflects the number of individuals sequenced from each reef.</p>
</blockquote>
<p><strong>Population structure:</strong></p>
<p>Why: many reasons: 1) identifying populations for contrasts in e.g. selection scans 2) identify admixed individuals that should be removed from analyses 3) identify barriers to gene flow etc</p>
<ul>
<li>no discernible relationship between geographic distance and genetic differentiation -&gt; gene flow
<ul>
<li>for this reason, Fst between populations is low</li>
</ul></li>
<li>EEMS (Estimated Effective Migration Surfaces) models relationship between genetics and geography petkova_VisualizingSpatialPopulation_2016</li>
</ul>
<p>Indicative of high connectivity among 12 sampled reefs.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-population-genetics-of-the-coral-acropora-millepora-3" class="slide level2 ultitle">
<h2>Example: Population genetics of the coral <em>Acropora millepora</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Motivation: corals are facing hard times and to prevent future losses of coral cover a better understanding of genetics is warranted.</p>
<h4 id="genomic-scans-for-selection">Genomic scans for <strong>selection</strong></h4>
<p>Motivation: identify loci associated with adaptation / selection</p>
<ul>
<li>little differentiation over reefs, however thermal regimes</li>
<li>genomic scan for <span class="math inline">\pi</span> (diversity) outliers</li>
</ul>
</div><div class="column" style="width:50%;">
<div id="fig-corals-selection" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-corals-selection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.science.org/cms/10.1126/science.aba4674/asset/83167ee1-c1f3-4ef8-88b3-d69a0cb93062/assets/graphic/369_aba4674_f4.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;4: Genomic scans for local adaptation detect a signal at sacsin. Nucleotide diversity (\pi) in 1kb-window, values in top 0.01% genome-wide in red (A, top). Close-up view around sacsin gene, with predicted gene structure above (A, bottom). h12 summary statistic showing the frequence of the two most common haplotypes (B). sacsin gene tree (blue) together with random gene trees, indicating deep genealogy at sacsin (C)"><img data-src="https://www.science.org/cms/10.1126/science.aba4674/asset/83167ee1-c1f3-4ef8-88b3-d69a0cb93062/assets/graphic/369_aba4674_f4.jpeg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-corals-selection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Genomic scans for local adaptation detect a signal at <em>sacsin</em>. Nucleotide diversity (<span class="math inline">\pi</span>) in 1kb-window, values in top 0.01% genome-wide in red (A, top). Close-up view around <em>sacsin</em> gene, with predicted gene structure above (A, bottom). h12 summary statistic showing the frequence of the two most common haplotypes (B). <em>sacsin</em> gene tree (blue) together with random gene trees, indicating deep genealogy at <em>sacsin</em> (C)
</figcaption>
</figure>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span></p>
</div>
</div></div>
<aside class="notes">
<blockquote>
<p>Genomic scans for local adaptation detect a signal at sacsin. (A) (Top) Values of pairwise nucleotide diversity (π) (86) estimated in sliding windows of 1 kb and a step-size of 500 bp across chromosome 7, with points colored in red indicating regions in the top 0.01% of values genome-wide. A loess-smoothed trend line of π across windows is indicated in purple. The peak with the most extreme values falls in the sacsin gene region. (Bottom) A close-up view of π estimated across a region of chromosome 7 surrounding the sacsin gene, which includes all of the outlier windows indicated in red at top. π per base pair was calculated in sliding windows of 100 bp (gray dots); the loess-smoothed line is shown in blue. The predicted gene structure of sacsin is indicated above, as well as the flanking upstream and downstream genes. The 29 nonsynonymous differences fixed between samples of the two most common haplotypes in sacsin are denoted with green lines. There are also 20 fixed synonymous differences. Gray vertical dashed lines delimit a region that was masked from variant calling because of predicted repetitive elements. (B) The h12 summary statistic (47), which measures the frequency of the two most common haplotypes, is plotted across chromosome 7 (supplementary materials, materials and methods). Red dots represent h12 values in the top 0.01% genome-wide. The peak with the most extreme values falls in the sacsin gene region. (C) A gene tree for the central 1-kb region in sacsin is shown in dark blue; it was constructed for one randomly chosen haplotype from a randomly selected individual from each of the 12 sampled reefs. The gene tree is rooted to aligned sequences from the reference genomes for A. digitifera (87) and A. tenuis (draft assembly from &lt;www.reefgenomics.org&gt;). Shown in gray are gene trees for 1000 randomly sampled 1-kb regions from across the genome for the same individuals. Each gene tree was inferred by using maximum likelihood implemented in dnaml (88).</p>
</blockquote>
<p><strong>Selection scan</strong></p>
<p>Why: identify loci associated with adaptation / selection, which provides potential mechanisms for adaptation, as well as information that could be important for conservation strategies</p>
<p>Little differentiation across reefs -&gt; little population structure over hundreds of kilometers. However, there are environmental differences (thermal regimes). Scan for pi outliers:</p>
<ul>
<li>points to <em>sacsin</em> gene</li>
<li>h12 measures the frequency of the two most common haplotypes; red indicate 0.01% outlier genome-wide</li>
<li>4C: tree for central 1kb region in <em>sacsin</em> deeper than split from <em>A.digitifera</em> and <em>A.tenuis</em>
<ul>
<li>variation in <em>sacsin</em> has been maintained for long time</li>
<li>co-chaperone for heat-shock protein Hsp70</li>
</ul></li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-population-genetics-of-the-coral-acropora-millepora-4" class="slide level2 ultitle">
<h2>Example: Population genetics of the coral <em>Acropora millepora</em></h2>
<p>Study highlights common analyses in population genomics study:</p>
<ol type="1">
<li class="fragment">Genome assembly, resequencing, variant calling and filtering</li>
<li class="fragment">Description of variation (e.g., <span class="math inline">\pi</span>) and genetic structure (LD)</li>
<li class="fragment">Description of population structure (admixture, PCA)</li>
<li class="fragment">Modelling of demographic history (PSMC)</li>
<li class="fragment">Genome scans for adaptive traits</li>
</ol>
</section>
<section id="population-genetics" class="slide level2 ultitle">
<h2>Population genetics</h2>
<div style="display: grid; grid-template-columns: 300px 200px 200px; grid-row-gap: 0px; grid-column-gap: 100px;">
<div>
&nbsp;
</div>
<div class="circular-image">
<p>Mutation</p>
<p><a href="assets/images/mutation.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img data-src="assets/images/mutation.webp"></a></p>
</div>
<div class="circular-image">
<p>Selection</p>
<p><a href="assets/images/finches.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="assets/images/finches.webp"></a></p>
</div>
<div>
&nbsp;
</div>
<div class="circular-image">
<p>Recombination</p>
<p><a href="assets/images/recombination.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="assets/images/recombination.webp"></a></p>
</div>
<div class="circular-image">
<p>Drift</p>
<p><a href="assets/images/genetic-drift.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img data-src="assets/images/genetic-drift.webp"></a></p>
</div>
</div>
<aside class="notes">
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Fuller et al., 2020</a>)</span> paper has population genetics in title -&gt; population genetics is a key ingredient.</p>
<p>Population genetics focuses on the genetic basis of evolution. It is mainly a theoretical subject, owing to the slow changes of genetic variation. As such, it tries to explain the shape and structure of genetic variation from theoretical predictions and models.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="from-population-genetics-to-population-genomics" class="slide level2 ultitle">
<h2>From population genetics to population genomics</h2>
<!-- markdownlint-disable MD013 -->
<div>
<div class="quarto-figure quarto-figure-center">
<figure>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="assets/images/kreitman-1983-table1.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="The variable sites at the Drosophila melanogaster ADH locus [@kreitman_NucleotidePolymorphismAlcohol_1983]"><img data-src="assets/images/kreitman-1983-table1.webp" class="quarto-figure quarto-figure-center" style="width:70.0%" alt="The variable sites at the Drosophila melanogaster ADH locus (Kreitman, 1983)"></a></p>
</figure>
</div>
<figcaption>The variable sites at the <em>Drosophila melanogaster</em> ADH locus <span class="citation" data-cites="kreitman_NucleotidePolymorphismAlcohol_1983">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kreitman, 1983</a>)</span></figcaption>
</figure>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<p>First study of natural population. However, limited to one locus.</p>
<aside class="notes">
<ul>
<li>from locus-based studies (e.g., alcohol dehydrogenase in Drosophila <span class="citation" data-cites="kreitman_NucleotidePolymorphismAlcohol_1983">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kreitman, 1983</a>)</span>) to genome-wide (e.g., Drosophila population genomics <span class="citation" data-cites="begun_PopulationGenomicsWholeGenome_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Begun et al., 2007</a>)</span></li>
<li>note: studied loci have often not been randomly chosen, which is another argument for whole-genome studies</li>
<li>enabler: sequencing technology</li>
</ul>
<p><span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Fuller et al., 2020</a>)</span> paper has population genetics in title -&gt; population genetics is a key ingredient.</p>
<p>Refer to Hahn’s points about learning something about global patterns:</p>
<ul>
<li>selection acts locally, demography globally</li>
<li>the structure of genetic variation and how it depends on
<ul>
<li>recombination landscapes and linked selection</li>
<li>demographic changes</li>
<li>identification of neutral loci</li>
</ul></li>
</ul>
<p>So, not simply about applying 10000 selection tests for multiple loci</p>
<p>All of the points above point to the importance of statistics which implies mathematics / computational skills important</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="from-population-genetics-to-population-genomics-1" class="slide level2 ultitle">
<h2>From population genetics to population genomics</h2>
<div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://journals.plos.org/plosbiology/article/figure/image?size=large&amp;download=&amp;id=10.1371/journal.pbio.0050310.g001" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Patterns of polymorphism and divergence [@begun_PopulationGenomicsWholeGenome_2007]"><img data-src="https://journals.plos.org/plosbiology/article/figure/image?size=large&amp;download=&amp;id=10.1371/journal.pbio.0050310.g001" style="width:60.0%" alt="Patterns of polymorphism and divergence (Begun et al., 2007)"></a></p>
<figcaption>Patterns of polymorphism and divergence <span class="citation" data-cites="begun_PopulationGenomicsWholeGenome_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Begun et al., 2007</a>)</span></figcaption>
</figure>
</div>
</div>
<p>Same system but <strong>genome-wide</strong>. Plots represent all chromosomes and the entire genome.</p>
<aside class="notes">
<p><span class="citation" data-cites="begun_PopulationGenomicsWholeGenome_2007">Begun et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2007</a>)</span> study: same system (<em>Drosophila</em>) but more individuals and whole genome. All of a sudden possible to ask questions about the general characteristics of diversity, not just limited to single loci.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="from-population-genetics-to-population-genomics-2" class="slide level2 ultitle">
<h2>From population genetics to population genomics</h2>
<div>
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://journals.plos.org/plosbiology/article/figure/image?size=large&amp;download=&amp;id=10.1371/journal.pbio.0050310.t002" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="Numbers of polymorphic and fixed variants [@begun_PopulationGenomicsWholeGenome_2007]"><img data-src="https://journals.plos.org/plosbiology/article/figure/image?size=large&amp;download=&amp;id=10.1371/journal.pbio.0050310.t002" class="quarto-figure quarto-figure-center" style="width:70.0%" alt="Numbers of polymorphic and fixed variants (Begun et al., 2007)"></a></p>
</figure>
</div>
<figcaption>Numbers of polymorphic and fixed variants <span class="citation" data-cites="begun_PopulationGenomicsWholeGenome_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Begun et al., 2007</a>)</span></figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<p>Novelty: now possible to do <em>genome-wide</em> characterization of variation in different <strong>functional contexts</strong></p>
<aside class="notes">
<p>Novelty: now possible to do <em>genome-wide</em> characterization of variation in different functional contexts</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-technological-revolution-in-sequencing-and-computing" class="slide level2 ultitle">
<h2>The technological revolution in <strong>sequencing</strong> and <strong>computing</strong></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-sequencing-cost" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-sequencing-cost-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="index_files/figure-revealjs/fig-sequencing-cost-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;5: Sequencing cost ($) per megabase [@wetterstrandka_DNASequencingCosts]"><img data-src="index_files/figure-revealjs/fig-sequencing-cost-1.svg" style="width:90.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-sequencing-cost-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Sequencing cost ($) per megabase <span class="citation" data-cites="wetterstrandka_DNASequencingCosts">(<a href="#/bibliography" role="doc-biblioref" onclick="">Wetterstrand, KA</a>)</span>
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://ourworldindata.org/uploads/2022/03/Transistor-Count-over-time.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Moore’s law"><img data-src="https://ourworldindata.org/uploads/2022/03/Transistor-Count-over-time.png" alt="Moore’s law"></a></p>
<figcaption>Moore’s law</figcaption>
</figure>
</div>
</div></div>
</section>
<section id="statistical-inference-in-population-genomics" class="slide level2 ultitle">
<h2>Statistical inference in population genomics</h2>
<p>The data deluge requires advanced statistical methods and models to do inference. Today data production outpaces theoretical advances. Therefore, take care not to attach too much faith to a test that explains data well.</p>
<div class="fragment">
<p>A population genomics study should aim at generating a baseline model that takes into account the processes that shape genetic variation <span class="citation" data-cites="johri_RecommendationsImprovingStatistical_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Johri et al., 2022</a>)</span>:</p>
<ol type="1">
<li>mutation</li>
<li>recombination</li>
<li>gene conversion</li>
<li>purifying selection acting on functional regions and its effects on linked variants (background selection)</li>
<li>genetic drift with demographic history and geographic structure</li>
</ol>
</div>
<aside class="notes">
<p>Caution against adaptationist storytelling; always compare to a baseline model that takes potential confounding factors into account</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="applications-of-population-genomics" class="slide level2 ultitle">
<h2>Applications of population genomics</h2>
<div style="display: grid; width: 100%; grid-template-areas: &quot;a b c&quot; &quot;d e f&quot;; grid-row-gap: 10px; grid-column-gap: 30px; align-items: center;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.annualreviews.org/docserver/fulltext/animal/11/1/av110115.f2.gif" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="Conservation genomics [@webster_PopulationGenomicsInsect_2023]"><img data-src="https://www.annualreviews.org/docserver/fulltext/animal/11/1/av110115.f2.gif" height="200" alt="Conservation genomics (Webster et al., 2023)"></a></p>
<figcaption>Conservation genomics <span class="citation" data-cites="webster_PopulationGenomicsInsect_2023">(<a href="#/bibliography" role="doc-biblioref" onclick="">Webster et al., 2023</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../../exercises/datasets/assets/images/stankowski-2019-fig1A.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-15" title="Speciation genomics [@stankowski_WidespreadSelectionGene_2019]"><img data-src="../../exercises/datasets/assets/images/stankowski-2019-fig1A.webp" height="200" alt="Speciation genomics (Stankowski et al., 2019)"></a></p>
<figcaption>Speciation genomics <span class="citation" data-cites="stankowski_WidespreadSelectionGene_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Stankowski et al., 2019</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/genetics/226/4/10.1093_genetics_iyae006/2/m_iyae006f1.jpeg?Expires=1760610839&amp;Signature=nii-RdEtrwqeLzTWwajHQCTXC~YJPndas9Soarpbckb8xbZswk9gmaWJDenpDMXfQc-p6-Pre65c11i6MJFbtuiEkWZyZzt3pE1vONLeGZdU~5RV4Q5UykIsYlfd~pnGiJC5mkTAQOp1OGW3RXkO0EfTbBly-~hMj627Pi9Ms6IipXxCWhVwzpuI9YlRkmqsc0HRNPFwf4cHyQS0SlgcZkCkEgJWbrthl~Ro9LuKB6bKTkIeECPKk3-7vzB2X~4X3ly3H5ce-xsFqZOnhGDCMk23B8d4wkrgBuJc7aeewJaiK~-Y7aGV7y-lNhBR-k~6whfW8l~F4-SYNhPonjr2dg__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" class="lightbox" data-gallery="quarto-lightbox-gallery-16" title="disentangle forces that create variation [@rodrigues_SharedEvolutionaryProcesses_2024]"><img data-src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/genetics/226/4/10.1093_genetics_iyae006/2/m_iyae006f1.jpeg?Expires=1760610839&amp;Signature=nii-RdEtrwqeLzTWwajHQCTXC~YJPndas9Soarpbckb8xbZswk9gmaWJDenpDMXfQc-p6-Pre65c11i6MJFbtuiEkWZyZzt3pE1vONLeGZdU~5RV4Q5UykIsYlfd~pnGiJC5mkTAQOp1OGW3RXkO0EfTbBly-~hMj627Pi9Ms6IipXxCWhVwzpuI9YlRkmqsc0HRNPFwf4cHyQS0SlgcZkCkEgJWbrthl~Ro9LuKB6bKTkIeECPKk3-7vzB2X~4X3ly3H5ce-xsFqZOnhGDCMk23B8d4wkrgBuJc7aeewJaiK~-Y7aGV7y-lNhBR-k~6whfW8l~F4-SYNhPonjr2dg__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" height="200" alt="disentangle forces that create variation (Rodrigues et al., 2024)"></a></p>
<figcaption>disentangle forces that create variation <span class="citation" data-cites="rodrigues_SharedEvolutionaryProcesses_2024">(<a href="#/bibliography" role="doc-biblioref" onclick="">Rodrigues et al., 2024</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://media.springernature.com/lw685/springer-static/esm/art%3A10.1038%2Fs41586-021-03224-9/MediaObjects/41586_2021_3224_Fig8_ESM.jpg?" class="lightbox" data-gallery="quarto-lightbox-gallery-17" title="paleogenomics (aDNA) [@vandervalk_MillionyearoldDNASheds_2021]"><img data-src="https://media.springernature.com/lw685/springer-static/esm/art%3A10.1038%2Fs41586-021-03224-9/MediaObjects/41586_2021_3224_Fig8_ESM.jpg?" height="200" alt="paleogenomics (aDNA) (van der Valk et al., 2021)"></a></p>
<figcaption>paleogenomics (aDNA) <span class="citation" data-cites="vandervalk_MillionyearoldDNASheds_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">van der Valk et al., 2021</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.frontiersin.org/files/Articles/555530/fgene-11-00742-HTML/image_m/fgene-11-00742-g002.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-18" title="domestication [@barrera-redondo_GenomicTranscriptomicEpigenomic_2020]"><img data-src="https://www.frontiersin.org/files/Articles/555530/fgene-11-00742-HTML/image_m/fgene-11-00742-g002.jpg" height="200" alt="domestication (Barrera-Redondo et al., 2020)"></a></p>
<figcaption>domestication <span class="citation" data-cites="barrera-redondo_GenomicTranscriptomicEpigenomic_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Barrera-Redondo et al., 2020</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41467-024-50239-7/MediaObjects/41467_2024_50239_Fig1_HTML.png?" class="lightbox" data-gallery="quarto-lightbox-gallery-19" title="ecology [@unneberg_EcologicalGenomicsNorthern_2024]"><img data-src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41467-024-50239-7/MediaObjects/41467_2024_50239_Fig1_HTML.png?" height="200" alt="ecology (Unneberg et al., 2024)"></a></p>
<figcaption>ecology <span class="citation" data-cites="unneberg_EcologicalGenomicsNorthern_2024">(<a href="#/bibliography" role="doc-biblioref" onclick="">Unneberg et al., 2024</a>)</span></figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller ultitle scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-barrera-redondo_GenomicTranscriptomicEpigenomic_2020" class="csl-entry" role="listitem">
Barrera-Redondo, J., Piñero, D., &amp; Eguiarte, L. E. (2020). Genomic, <span>Transcriptomic</span> and <span>Epigenomic Tools</span> to <span>Study</span> the <span>Domestication</span> of <span>Plants</span> and <span>Animals</span>: <span>A Field Guide</span> for <span>Beginners</span>. <em>Frontiers in Genetics</em>, <em>11</em>.
</div>
<div id="ref-begun_PopulationGenomicsWholeGenome_2007" class="csl-entry" role="listitem">
Begun, D. J., Holloway, A. K., Stevens, K., Hillier, L. W., Poh, Y.-P., Hahn, M. W., Nista, P. M., Jones, C. D., Kern, A. D., Dewey, C. N., Pachter, L., Myers, E., &amp; Langley, C. H. (2007). Population <span>Genomics</span>: <span>Whole-Genome Analysis</span> of <span>Polymorphism</span> and <span>Divergence</span> in <span>Drosophila</span> simulans. <em>PLOS Biology</em>, <em>5</em>(11), e310. <a href="https://doi.org/10.1371/journal.pbio.0050310">https://doi.org/10.1371/journal.pbio.0050310</a>
</div>
<div id="ref-fuller_PopulationGeneticsCoral_2020" class="csl-entry" role="listitem">
Fuller, Z. L., Mocellin, V. J. L., Morris, L. A., Cantin, N., Shepherd, J., Sarre, L., Peng, J., Liao, Y., Pickrell, J., Andolfatto, P., Matz, M., Bay, L. K., &amp; Przeworski, M. (2020). Population genetics of the coral <span>Acropora</span> millepora: <span>Toward</span> genomic prediction of bleaching. <em>Science</em>, <em>369</em>(6501), eaba4674. <a href="https://doi.org/10.1126/science.aba4674">https://doi.org/10.1126/science.aba4674</a>
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-hartl1997principles" class="csl-entry" role="listitem">
Hartl, D. L., &amp; Clark, A. G. (1997). <em>Principles of population genetics</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-johri_RecommendationsImprovingStatistical_2022" class="csl-entry" role="listitem">
Johri, P., Aquadro, C. F., Beaumont, M., Charlesworth, B., Excoffier, L., Eyre-Walker, A., Keightley, P. D., Lynch, M., McVean, G., Payseur, B. A., Pfeifer, S. P., Stephan, W., &amp; Jensen, J. D. (2022). Recommendations for improving statistical inference in population genomics. <em>PLOS Biology</em>, <em>20</em>(5), e3001669. <a href="https://doi.org/10.1371/journal.pbio.3001669">https://doi.org/10.1371/journal.pbio.3001669</a>
</div>
<div id="ref-kreitman_NucleotidePolymorphismAlcohol_1983" class="csl-entry" role="listitem">
Kreitman, M. (1983). Nucleotide polymorphism at the alcohol dehydrogenase locus of <span>Drosophila</span> melanogaster. <em>Nature</em>, <em>304</em>(5925), 412. <a href="https://doi.org/10.1038/304412a0">https://doi.org/10.1038/304412a0</a>
</div>
<div id="ref-li_InferenceHumanPopulation_2011" class="csl-entry" role="listitem">
Li, H., &amp; Durbin, R. (2011). Inference of human population history from individual whole-genome sequences. <em>Nature</em>, <em>475</em>(7357), 493–496. <a href="https://doi.org/10.1038/nature10231">https://doi.org/10.1038/nature10231</a>
</div>
<div id="ref-rodrigues_SharedEvolutionaryProcesses_2024" class="csl-entry" role="listitem">
Rodrigues, M. F., Kern, A. D., &amp; Ralph, P. L. (2024). Shared evolutionary processes shape landscapes of genomic variation in the great apes. <em>Genetics</em>, <em>226</em>(4), iyae006. <a href="https://doi.org/10.1093/genetics/iyae006">https://doi.org/10.1093/genetics/iyae006</a>
</div>
<div id="ref-stankowski_WidespreadSelectionGene_2019" class="csl-entry" role="listitem">
Stankowski, S., Chase, M. A., Fuiten, A. M., Rodrigues, M. F., Ralph, P. L., &amp; Streisfeld, M. A. (2019). Widespread selection and gene flow shape the genomic landscape during a radiation of monkeyflowers. <em>PLOS Biology</em>, <em>17</em>(7), e3000391. <a href="https://doi.org/10.1371/journal.pbio.3000391">https://doi.org/10.1371/journal.pbio.3000391</a>
</div>
<div id="ref-unneberg_EcologicalGenomicsNorthern_2024" class="csl-entry" role="listitem">
Unneberg, P., Larsson, M., Olsson, A., Wallerman, O., Petri, A., Bunikis, I., Vinnere Pettersson, O., Papetti, C., Gislason, A., Glenner, H., Cartes, J. E., Blanco-Bercial, L., Eriksen, E., Meyer, B., &amp; Wallberg, A. (2024). Ecological genomics in the <span>Northern</span> krill uncovers loci for local adaptation across ocean basins. <em>Nature Communications</em>, <em>15</em>(1), 6297. <a href="https://doi.org/10.1038/s41467-024-50239-7">https://doi.org/10.1038/s41467-024-50239-7</a>
</div>
<div id="ref-vandervalk_MillionyearoldDNASheds_2021" class="csl-entry" role="listitem">
van der Valk, T., Pečnerová, P., Díez-del-Molino, D., Bergström, A., Oppenheimer, J., Hartmann, S., Xenikoudakis, G., Thomas, J. A., Dehasque, M., Sağlıcan, E., Fidan, F. R., Barnes, I., Liu, S., Somel, M., Heintzman, P. D., Nikolskiy, P., Shapiro, B., Skoglund, P., Hofreiter, M., … Dalén, L. (2021). Million-year-old <span>DNA</span> sheds light on the genomic history of mammoths. <em>Nature</em>, <em>591</em>(7849), 265–269. <a href="https://doi.org/10.1038/s41586-021-03224-9">https://doi.org/10.1038/s41586-021-03224-9</a>
</div>
<div id="ref-webster_PopulationGenomicsInsect_2023" class="csl-entry" role="listitem">
Webster, M. T., Beaurepaire, A., Neumann, P., &amp; Stolle, E. (2023). Population <span>Genomics</span> for <span>Insect Conservation</span>. <em>Annual Review of Animal Biosciences</em>, <em>11</em>(1), 115–140. <a href="https://doi.org/10.1146/annurev-animal-122221-075025">https://doi.org/10.1146/annurev-animal-122221-075025</a>
</div>
<div id="ref-wetterstrandka_DNASequencingCosts" class="csl-entry" role="listitem">
Wetterstrand, KA. <em><span>DNA Sequencing Costs</span>: <span>Data</span> from the <span>NHGRI Genome Sequencing Program</span> (<span>GSP</span>)</em>. <a href="https://www.genome.gov/sequencingcostsdata">www.genome.gov/sequencingcostsdata</a>
</div>
</div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbis.se"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Population Genomics in Practice</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>