---
title: Monkeyflowers dataset
subtitle: Speciation genomics and simulation
author:
  - Per Unneberg
format: nbis-course-html
---


<!-- markdownlint-disable MD041 -->

{{< include /exercises/_knitr.qmd >}}

{{< include /exercises/_rlibs.qmd >}}

```{r }
#| label: r-libs-extra
#| echo: false
#| eval: true
#| warning: false
library(kableExtra)
```

<!-- markdownlint-enable MD041 -->

:::{}

<!-- markdownlint-disable MD013 -->

![](https://www.science.org/cms/10.1126/science.365.6456.854/asset/db271bee-8914-4712-9965-dcf1615b6a43/assets/graphic/365_854_f1.jpeg){fig-alt="The allure of monkeyflowers (DOI: 10.1126/science.365.6456.854)"}

<!-- markdownlint-enable MD013 -->

:::

## The monkeyflower model system {#sec-monkeyflower-model-system}

Monkeyflowers (*Mimulus*) have recently become a key model in
evolution and plant biology [@pennisi_AllureMonkeyflowers_2019]. The
monkeyflower system consists of 160--200 species that display an
amazing phenotypic variation. The genome is small, only 207Mbp, which
makes it an ideal candidate for genomics - and for computer exercises!

## The monkeyflower genomic landscape

Recently, @stankowski_WidespreadSelectionGene_2019 used the
monkeyflower system to investigate what forces affect the genomic
landscape. @burri_InterpretingDifferentiationLandscapes_2017 has
suggested that background selection (BGS) is one of the main causes
for correlations between genomic landscapes, and that one way to study
this phenomenon is to look at closely related taxa. This is one of the
objectives of the @stankowski_WidespreadSelectionGene_2019 paper.

They performed whole-genome resequencing of 37 individuals
from 7 subspecies and 2 ecotypes of *Mimulus aurantiacus* and its
sister taxon *M. clevelandii* (@fig-evolutionary-relationship), all
sampled in California (@fig-monkeyflower-sampling-location).

![Evolutionary relationships across the radiation](assets/images/stankowski-2019-fig1A.webp){#fig-evolutionary-relationship}

:::{.column-margin}

![Sampling locations](/slides/simulation/assets/images/stankowski-2019-figs2.webp){#fig-monkeyflower-sampling-location}

:::

Genomewide statistics, such as diversity ($\pi$), divergence
($d_{XY}$) and differentiation $F_{ST}$, were calculated within and
between taxa to generate genomic diversity landscapes. The landscapes
were highly similar across taxa, and local variation in genomic
features, such as gene density and recombination rate, was predictive
of variation in landscape patterns. These features suggest the
influence of selection, in particular BGS.

Although many characteristics were predicted by a model where BGS is
one of the main causes, there were deviations. Therefore, the authors
performed simulations in SLiM [@haller_SLiMForwardGenetic_2019] with
alternative models to see whether other factors could explain the
observed patterns.

In all, six scenarios were studied:

1. neutral evolution
2. BGS (non-neutral mutations are deleterious)
3. Bateson-Dobzhansky-Muller incompatibility (BDMI); after split, a
   fraction variants deleterious in one population, neutral in other
4. positive selection
5. BGS and positive selection
6. local adaptation; as 4 but also after split some variants
    are beneficial in one population, neutral in other

@fig-slim-simulations shows typical results of the simulations.

![Genomic landscapes simulated under different divergence histories.](/slides/simulation/assets/images/stankowski-2019-fig7.webp){#fig-slim-simulations}

In conclusion, the authors found that although BGS plays a role, it
does not sufficiently explain all observations, and that other aspects
of natural selection (such as rapid adaptation) are responsible for
the similarities between genomic landscapes.

A locus that previously had been associated with differentiation of
red and yellow ecotypes was investigated in more detail. The locus is
located on linkage group 4 (`LG4`), and we will be using both a 3Mbp
region of interest (ROI) surronding the locus, and the whole linkage
group, for different exercises.

## Data

The dataset consists of 37 samples (see @tbl-monkeyflower-data for
example information). Raw sequence reads were downloaded from
[Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra) (SRA),
bioproject
[PRJNA549183](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA549183)
and mapped to the reference sequence
[M_aurantiacus_v1_splitline_ordered.fasta](http://mimubase.org/FTP/Genomes/Maurantiacus_v1.0/).
Reads that mapped to the ROI were extracted and constitute the
sequence data that will be used during the exercises.

```{r }
#| label: tbl-monkeyflower-data
#| echo: false
#| eval: true
#| tbl-cap:
#|    Example of monkeyflower samples. See file `sampleinfo.csv`
#|    in data repository for full listing.
data <- read.csv("sampleinfo.csv", header = TRUE)
data[, c(1, 2, 3, 4, 7, 8, 9)] %>%
  head() %>%
  kable("html", escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "responsive"),
    full_width = FALSE, position = "left"
  )
```

### PDC data storage

The monkeyflower dataset is located in PDC project `{{< var
naissproject >}}` at `{{< var pg_pgip_dir >}}/data/monkeyflower`.

### Online access

In addition, a pre-compiled dataset consisting of the files relevant
for exercises is available from [`{{< var datasets.monkeyflower.url
>}}`]({{< var datasets.monkeyflower.url >}}).

### Github

The github repository [pgip-data]({{< var pgip_data.main >}}) contains
reference sequence and read data for 37 monkeyflower individuals for
the region `LG4:12,000,000-12,100,000`. The data resides in the
[data/monkeyflower/tiny]({{< var pgip_data.tiny >}}) subdirectory. This
data set is used as input data to render the website.

The repository hosts a [Snakemake workflow]({{< var pgip_data.workflow
>}}) to generate all data needed for the exercises.
