<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – DNA sequencing data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-802d3b82e08d8dc8d935d89a8ebe9074.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">DNA sequencing data</h1>
  <p class="subtitle">DNA sequencing approaches and data</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="sequencing-technologies" class="slide level2 ultitle">
<h2>Sequencing technologies</h2>
<div class="columns">
<div class="column" style="width:90%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://ngisweden.scilifelab.se/wp-content/uploads/2023/06/tech3-1024x171.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="https://ngisweden.scilifelab.se/wp-content/uploads/2023/06/tech3-1024x171.png" class="quarto-figure quarto-figure-center" style="width:80.0%"></a></p>
</figure>
</div>
</div><div class="column" data-fig-align="left" style="width:10%;">
<p><a href="https://ngisweden.scilifelab.se/wp-content/themes/ngisweden/img/NGI-logo.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img data-src="https://ngisweden.scilifelab.se/wp-content/themes/ngisweden/img/NGI-logo.svg" style="width:100.0%"></a></p>
</div></div>
<div class="columns">
<div class="column" style="width:50%;">
<h5 id="illumina-novaseq-600">Illumina NovaSeq 600</h5>
<div class="smallr">
<blockquote>
<p>Scale up and down with a tunable output of up to 6 Tb and 20B single reads in &lt; 2 days.</p>
</blockquote>
</div>
<p>Up to 2X250 bp read length. Price example: 8,000 SEK total for resequencing 3Gbp genome to 30X</p>
<div class="flushright smallest">
<p><a href="https://www.illumina.com/systems/sequencing-platforms/novaseq.html" class="uri">https://www.illumina.com/systems/sequencing-platforms/novaseq.html</a></p>
</div>
</div><div class="column" style="width:50%;">
<h5 id="pacbio-revio">PacBio Revio</h5>
<div class="smallr">
<blockquote>
<p>Up to 360 Gb of HiFi reads per day, equivalent to 1,300 human whole genomes per year.</p>
</blockquote>
</div>
<p>Tens of kilobases long HiFi reads. Price example (Sequel II): ~35kSEK per library and SMRT cell</p>
<div class="flushright smallest">
<p><a href="https://www.pacb.com/revio/" class="uri">https://www.pacb.com/revio/</a></p>
</div>
</div></div>
</section>
<section id="sequencing-approaches" class="slide level2 ultitle">
<h2>Sequencing approaches</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="cell-output-display">
<div id="fig-sequencing-cost" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-sequencing-cost-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="dna_sequencing_files/figure-revealjs/fig-sequencing-cost-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;1: DNA sequencing costs [@wetterstrandka_DNASequencingCosts]"><img data-src="dna_sequencing_files/figure-revealjs/fig-sequencing-cost-1.svg" style="width:90.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-sequencing-cost-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: DNA sequencing costs <span class="citation" data-cites="wetterstrandka_DNASequencingCosts">(<a href="#/bibliography" role="doc-biblioref" onclick="">Wetterstrand, KA</a>)</span>
</figcaption>
</figure>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<p>Despite price drop, still need to make choices regarding depth and breadth of sequencing coverage and number of samples.</p>
</div><div class="column fragment" style="width:50%;">
<div id="fig-sequencing-approaches" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-sequencing-approaches-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/sequencing_approaches_lou_2021.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;2: Comparison of common sequencing approaches. Restriction site-associated DNA sequencing (RAD-seq, top), targets regions flanking given restriction sites, but misses much of genome. Pooled sequencing (Pool-seq, middle) is cost-effective, but loses information about individuals. Low-coverage whole genome sequencing (lcWGS, bottom) is increasing in popularity, but genotyping low coverage data is problematic."><img data-src="assets/images/sequencing_approaches_lou_2021.webp" class="quarto-figure quarto-figure-center" style="width:60.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-sequencing-approaches-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Comparison of common sequencing approaches. Restriction site-associated DNA sequencing (RAD-seq, top), targets regions flanking given restriction sites, but misses much of genome. Pooled sequencing (Pool-seq, middle) is cost-effective, but loses information about individuals. Low-coverage whole genome sequencing (lcWGS, bottom) is increasing in popularity, but genotyping low coverage data is problematic.
</figcaption>
</figure>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">Lou et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span></p>
</div>
<div class="fragment">
<p>Our focus will be on Whole Genome reSequencing (WGS), mostly high-coverage.</p>
</div>
</div></div>
<aside class="notes">
<p>Despite the fact that sequencing costs have dropped dramatically (left), there still are choices to be made regarding the distribution of costs along 1) sequencing coverage depth, i.e., the mean depth of sequencing 2) sequencing coverage breadth, i.e., whether or not to do targeted or whole-genome resequencing or 3) sample size; how many individuals to sample. Whole-genome resequencing of individuals from populations to sufficient depth (30X) is still very expensive, but often needed to understand mechanisms of adaptation <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lou et al., 2021, p. 5967</a>)</span>. Various protocols have been developed to meet the challenges that cost imposes:</p>
<ol type="1">
<li>RAD-seq, restriction site-associated DNA sequencing, targets regions flanking given restriction sites. Downside: much of genome is missed</li>
<li>pool-seq, pooled sequencing. Cost-effective, but loses information about individuals</li>
<li>lcWGS, low-coverage whole genome sequencing increasing in popularity. Genotyping low coverage is problematic however.</li>
</ol>
<p>Our focus here is WGS (whole-genome resequencing), primarily high-coverage, despite the cost it may incur.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genome-assembly-and-population-resequencing" class="slide level2 ultitle">
<h2>Genome assembly and population resequencing</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h4 id="genome-assembly">Genome assembly</h4>
<div id="fig-genome-assembly" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-genome-assembly-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/oso-genome-assembly.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;3: Schematic overview of genome assembly and annotation. Population resequencing often requires a reference assembly. Assembly quality impacts downstream analyses"><img data-src="assets/images/oso-genome-assembly.webp" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-genome-assembly-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Schematic overview of genome assembly and annotation. Population resequencing often requires a <em>reference assembly</em>. Assembly quality impacts downstream analyses
</figcaption>
</figure>
</div>
<div class="flushright smallest translatey50">
<p><span class="citation" data-cites="allendorf_PopulationGenomics_2022">Allendorf et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2022</a>)</span></p>
</div>
</div><div class="column fragment" style="width:50%;">
<h4 id="population-resequencing">Population resequencing</h4>
<div id="fig-pop-reseq" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-pop-reseq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/population-resequencing.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;4: Overview of short reads mapped to a reference sequence. Reads have been colored according to population (red or yellow). There are three individuals from each population. Note the variation in sequence coverage. Mismatches in reads are highlighted as different colors with respect to population color."><img data-src="assets/images/population-resequencing.webp"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pop-reseq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Overview of short reads mapped to a reference sequence. Reads have been colored according to population (red or yellow). There are three individuals from each population. Note the variation in sequence coverage. Mismatches in reads are highlighted as different colors with respect to population color.
</figcaption>
</figure>
</div>
</div></div>
<aside class="notes">
<p>DNA sequences are generated from DNA fragments, often as paired-end reads, or long reads. For population genomics, it is often necessary to generate a <em>reference sequence</em> through <em>genome assembly</em>, the quality of which will impact the types of downstream analyses that can be done. Reads are assembled into <em>contigs</em> (contiguous sequences) that with the aid of genetic maps are pieced together into <em>scaffolds</em>. In the left figure, note that the scaffolds contain areas of unknown sequence.</p>
<p>Once a reference sequence has been produced, (short) reads from individuals are mapped to the reference (right).</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dna-sequences-in-fastq-format" class="slide level2 smaller ultitle">
<h2>DNA sequences in FASTQ format</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a></a><span class="fu">ls</span> <span class="at">--long</span> <span class="at">--human</span> fastq/<span class="pp">*</span>.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>-rw-r--r-- 1 runner runner 302K Sep 15 17:33 fastq/PUN-Y-INJ_R1.fastq.gz
-rw-r--r-- 1 runner runner 393K Sep 15 17:33 fastq/PUN-Y-INJ_R2.fastq.gz</code></pre>
</div>
</div>
<div class="fragment">
<p>Count number of lines:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a></a><span class="fu">zcat</span> fastq/PUN-Y-INJ_R1.fastq.gz <span class="kw">|</span> <span class="fu">wc</span> <span class="at">--lines</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>15768</code></pre>
</div>
</div>
</div>
<div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:30%;">
<p>Format:</p>
<ol type="1">
<li>sequence id (prefixed by <code>@</code>)</li>
<li>DNA sequence</li>
<li>separator (<code>+</code>)</li>
<li>Phred base quality scores</li>
</ol>
</div><div class="column" style="width:70%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a></a><span class="fu">zcat</span> fastq/PUN-Y-INJ_R1.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 8 <span class="kw">|</span> <span class="fu">cut</span> <span class="at">--characters</span> <span class="at">-30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@SRR9309790.10003134
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJFJJJJJJJJJJJJJJJJJ
@SRR9309790.10003222
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJ</code></pre>
</div>
</div>
</div></div>
</div>
</div>
<aside class="notes">
<p>First data encounter is usually FASTQ files from sequencing center. Files contain <strong>sequence reads</strong> that are generated by the sequencing machine. They correspond to short stretches of DNA that have been sequentially read by the machine, resulting in a string of <strong>base calls</strong> that constitute the reads. Every DNA base has an associated so-called <strong>Phred-scaled quality score</strong> from 0 to 93 that are encoded using ASCII 33 to 126.</p>
<p>The format consists of 1. @ title with id 2. raw sequence 3. +: a separator with optional description 4. the quality scores</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dna-sequence-quality-control" class="slide level2 ultitle">
<h2>DNA sequence quality control</h2>
<p>Quality values represent the probability <span class="math inline">P</span> that the call is <em>incorrect</em>. They are coded as <strong>Phred</strong> quality scores <span class="math inline">Q</span>. Here, <span class="math inline">Q=20</span> implies 1% probability of error, <span class="math inline">Q=30</span> implies 0.1% and so on. Typically you should not rely on quality values below <span class="math inline">20</span>.</p>
<p><span class="math display">
Q = -10 \log_{10} P
</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<p>A common way to do QC is with <code>fastqc</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a></a><span class="ex">fastqc</span> <span class="at">--outdir</span> fastqc <span class="at">--extract</span> fastq/<span class="pp">*</span>fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<ul>
<li>For Illumina paired-end sequencing, the <strong>second</strong> read pair usually shows a larger drop in quality towards ends</li>
<li><strong>Trimming</strong> the sequences for adapter sequence and quality is good practice (e.g., with CutAdapt <span class="citation" data-cites="martin_CutadaptRemovesAdapter_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Martin, 2011</a>)</span>)</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<p><a href="results/PUN-Y-INJ_R1_fastqc/Images/per_base_quality.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="results/PUN-Y-INJ_R1_fastqc/Images/per_base_quality.svg" style="width:80.0%"></a></p>
</div></div>
</section>
<section id="bibliography" class="slide level2 smaller ultitle scrollable">
<h2>Bibliography</h2>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-allendorf_PopulationGenomics_2022" class="csl-entry" role="listitem">
Allendorf, F. W., Funk, W. C., Aitken, S. N., Byrne, M., &amp; Luikart, G. (2022). Population <span>Genomics</span>. In F. W. Allendorf, W. C. Funk, S. N. Aitken, M. Byrne, G. Luikart, &amp; A. Antunes (Eds.), <em>Conservation and the <span>Genomics</span> of <span>Populations</span></em> (p. 0). <span>Oxford University Press</span>. <a href="https://doi.org/10.1093/oso/9780198856566.003.0004">https://doi.org/10.1093/oso/9780198856566.003.0004</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-martin_CutadaptRemovesAdapter_2011" class="csl-entry" role="listitem">
Martin, M. (2011). Cutadapt removes adapter sequences from high-throughput sequencing reads. <em>EMBnet.journal</em>, <em>17</em>(1), pp. 10–12. <a href="https://doi.org/10.14806/ej.17.1.200">https://doi.org/10.14806/ej.17.1.200</a>
</div>
<div id="ref-wetterstrandka_DNASequencingCosts" class="csl-entry" role="listitem">
Wetterstrand, KA. <em><span>DNA Sequencing Costs</span>: <span>Data</span> from the <span>NHGRI Genome Sequencing Program</span> (<span>GSP</span>)</em>. <a href="https://www.genome.gov/sequencingcostsdata">www.genome.gov/sequencingcostsdata</a>
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>DNA data</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>