<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Variant calling and genotyping</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-dd75e976e0414ec343b1912506d158bc.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Variant calling and genotyping</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="variant-and-genotype-calling" class="slide level2 ultitle">
<h2>Variant and genotype calling</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg2986/MediaObjects/41576_2011_Article_BFnrg2986_Fig1_HTML.jpg?as=webp" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg2986/MediaObjects/41576_2011_Article_BFnrg2986_Fig1_HTML.jpg?as=webp" class="quarto-figure quarto-figure-center" style="width:60.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<div class="flushright smallest translatey50">
<p><span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">Nielsen et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2011</a>)</span></p>
</div>
</div><div class="column" style="width:50%;">
<dl>
<dt>SNP calling</dt>
<dd>
Identification of polymorphic sites (<span class="math inline">&gt;1</span>% allele frequency)
</dd>
<dt>Variant calling</dt>
<dd>
Identification of variant sites (sufficient that <strong>any</strong> allele differs); single nucleotide variant (<strong>SNV</strong>)
</dd>
<dt>Genotype calling</dt>
<dd>
Determine the allele combination for each individual (aa, aA, or AA for bi-allelic variants)
</dd>
</dl>
<div class="fragment">
<p>Knowing variant sites informs us of possible genotypes and improves genotyping.</p>
<p>Example: knowing a site has <code>A</code> or <code>C</code> limits possible genotype calls to <code>AA</code>, <code>AC</code>, or <code>CC</code></p>
</div>
</div></div>
<aside class="notes">
<p>Figure caption from <span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen et al., 2011</a>)</span>:</p>
<blockquote>
<p>Pre-processing steps (shown in yellow) transform the raw data from next-generation sequencing technology into a set of aligned reads that have a measure of confidence, or quality score, associated with the bases of each read. The per-base quality scores produced by base-calling algorithms may need to be recalibrated to accurately reflect the true error rates. Depending on the number of samples and the depth of coverage, either a multi-sample calling procedure (green) or a single-sample calling procedure (orange) may then be applied to obtain SNP or genotype calls and associated quality scores. Note that the multi-sample procedure may include a linkage-based analysis, which can substantially improve the accuracy of SNP or genotype calls. Finally, post-processing (purple) uses both known data and simple heuristics to filter the set of SNPs and/or improve the associated quality scores. Optional, although recommended, steps are shown in dashed lines.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="variants-show-up-in-pileup-alignments" class="slide level2 smaller ultitle">
<h2>Variants show up in pileup alignments</h2>
<div class="columns">
<div class="column smallr" style="width:48%;">
<h5>
Sample PUN-Y-INJ
</h5>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............M...............................</span><span class="tviewcu2">.</span><span class="tviewcu4">............</span><br><span class="tviewc4">...............A...</span><span>                               </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">...............A.................</span><span>                 </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>      </span><span class="tviewc4">,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,a,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,a,,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,a,,,,,aa,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span></pre>
</div>


</div><div class="column" style="width:4%;">

</div><div class="column smallr" style="width:48%;">
<h5>
Sample PUN-R-ELF
</h5>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............A............................................</span><br><span class="tviewc4">,,,,,,,</span><span>                           </span><span class="tviewc4">..........................</span><br><span class="tviewc4">...............A.....</span><span>                  </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A...A....</span><span>               </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A.........</span><span>                                   </span><br><span class="tviewc4">...............A...........C.............</span><span>                   </span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,g,,,,,,,a,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">.............A............................................</span></pre>
</div>


</div></div>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<p>Potential variants show up as multiple mismatches in a column. Two questions arise:</p>
<ul>
<li>how do we detect variant sites?</li>
<li>how do we distinguish variants from sequencing error?</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Simple approach: filter bases on quality (e.g., Q20), call heterozygous if 20-80% bases non-reference.</p>
<p>Issues: undercalls heterozygotes, no measure of uncertainty</p>
</div>
<div class="fragment">
<p>Solution: probabilistic methods!</p>
<div class="flushright smallest">
<p><span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen et al., 2011</a>)</span></p>
</div>
</div>
</div></div>
<aside class="notes">
<p>Potential variants show up as multiple mismatches in a column. Left sample has three reads that match the reference so is probably heterozygote. For the right sample no read matches reference so most likely call is homozygote alternate.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="we-can-calculate-likelihoods-of-observed-data" class="slide level2 ultitle">
<h2>We can calculate likelihoods of observed data</h2>
<h3 class="ultitle" id="example-excluding-sequencing-error">Example (excluding sequencing error!)</h3>
<div class="columns">
<div class="column" style="width:5%;">


<title>
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">

</div>
<pre class="tviewpre"><span>   </span><br><span>TGC</span><br><span class="tviewcu4">.K.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:5%;">

</div><div class="column" style="width:90%;">
<p>Goal: calculate likelihood of observing <span class="math inline">X=</span><code>G,g,T,T,G,g,t</code> from a genotype <span class="math inline">G</span>. Denote by <span class="math inline">X_i</span> the observation at each position <span class="math inline">i</span> of <span class="math inline">X</span>. We further assume each observation <span class="math inline">X_i</span> can be treated independently. We restrict possible genotypes to the observed alleles (i.e., <code>G, T</code>). Some observations:</p>
<div class="fragment">
<p>Prob(<span class="math inline">X_1=</span><code>G</code> <em>assuming</em> genotype <span class="math inline">G</span>=<code>T,T</code>) = P(<span class="math inline">X_1=</span><code>G</code>|<code>T,T</code>) = <span class="math inline">0</span></p>
</div>
<div class="fragment">
<p>Prob(<span class="math inline">X_1=</span><code>G</code> assuming genotype <code>G,G</code>) = P(<span class="math inline">X_1=</span><code>G</code>|<code>G,G</code>) = <span class="math inline">1</span></p>
</div>
<div class="fragment">
<p>Prob(<span class="math inline">X_1=</span><code>G</code> assuming genotype <code>G,T</code>) = P(<span class="math inline">X_1=</span><code>G</code>|<code>G,T</code>) = <span class="math inline">0.5</span></p>
</div>
<div class="fragment">
<p>To get total likelihood <span class="math inline">P(X|G)</span> assuming a genotype <span class="math inline">G</span> (here <code>G,T</code>), we can multiply over all observations (reads):</p>
<p><span class="math display">
\begin{align}
    P(X|\mathsf{G,T}) &amp; = P(X_1=\mathsf{G}|\mathsf{G,T})%%
P(X_2=\mathsf{g}|\mathsf{G,T})%%
P(X_3=\mathsf{T}|\mathsf{G,T})%%
P(X_4=\mathsf{T}|\mathsf{G,T}) \\
    &amp; P(X_5=\mathsf{G}|\mathsf{G,T})%%
    P(X_6=\mathsf{g}|\mathsf{G,T})%%
    P(X_7=\mathsf{t}|\mathsf{G,T}) = 0.5^7
\end{align}
</span></p>
</div>
</div></div>
<aside class="notes">
<p>We restrict the possible genotypes to the observed alleles at a site (here G and T). If there are more than two observed alleles, a common procedure is to pick the two with highest frequencies, under the assumption that the rarest observation is a sequencing error.</p>
<p>In reality, we also need to take into account sequencing error. There are different ways of doing this (e.g. <span class="citation" data-cites="maruki_GenotypeCallingPopulationGenomic_2017">Maruki &amp; Lynch (<a href="#/bibliography" role="doc-biblioref" onclick="">2017</a>)</span>, <span class="citation" data-cites="depristo_FrameworkVariationDiscovery_2011">DePristo et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2011</a>)</span>), but we leave the details to the interested reader.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="we-can-use-bayes-theorem-to-genotype" class="slide level2 smaller ultitle">
<h2>We can use Bayes’ theorem to genotype</h2>
<h3 class="ultitle" id="example">Example</h3>
<div class="columns">
<div class="column" style="width:5%;">


<title>
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">

</div>
<pre class="tviewpre"><span>   </span><br><span>TGC</span><br><span class="tviewcu4">.K.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:5%;">

</div><div class="column" style="width:90%;">
<p>For a given site, we have a number of observations <span class="math inline">X</span>. We have shown we can calculate the likelihood of observing <span class="math inline">X</span> given a genotype <span class="math inline">G</span>, <span class="math inline">P(X|G)</span>.</p>
<div class="fragment">
<p>However; what we really want to know is the most likely genotype <span class="math inline">G</span> <em>given the data</em> <span class="math inline">X</span>, or <span class="math inline">P(G|X)</span>.</p>
</div>
<div class="fragment">
<p>Apply Bayes’ theorem:</p>
<div class="large">
<p><span class="math display">
P(G|X) \sim P(X|G)\cdot P(G)
</span></p>
</div>
</div>
<div class="fragment large">
<p><span class="math display">
\text{posterior} \sim \text{likelihood} \cdot \text{prior}
</span></p>
</div>
<div class="fragment">
<p>Consequently we need to set a prior on <span class="math inline">G</span>. If allele frequencies are known, we can constrain the frequencies; for example, if A is known to be low (<span class="math inline">\sim1</span>%) AA genotype is <em>very</em> unlikely. Otherwise, could set all equal (flat prior).</p>
</div>
</div></div>
<aside class="notes">
<p>cf <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890511" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360035890511</a></p>
<p><span class="citation" data-cites="li_SNPDetectionMassively_2009">R. Li et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2009</a>)</span>, Table 1, shows a nice numerical example of one way of setting priors. The authors assume a specific allele, G, in the <em>reference sequence</em>:</p>
<p>–G–</p>
<p>They start by calculating the frequency of <em>haploid</em> genotypes. They first determine <span class="math inline">p_G</span> by assuming a <em>heterozygous SNP rate</em> <span class="math inline">f=0.001</span>, which means 1 in a 1000 sites has G/G genotype mutated to G/X, where X is one of {A,C,T}. They assume a transition to transversion (ts/tv) ratio of 4, meaning X=A four times as often as C or T (there is an error in the text where C is taken to be the transition; the numbers in the table are correct however). This gives the following haploid genotype frequencies:</p>
<p><span class="math display">\begin{align}
p_G &amp; = 1-f = 0.999 \\
p_A &amp; = 4f/6 = 6.67\times10^{-4} \\
p_C &amp; = f/6 = 1.67\times10^{-4}\\
p_T &amp; = f/6 =1.67\times10^{-4}
\end{align}</span></p>
<p>To get the <em>diploid</em> genotypes, we simply multiply the corresponding entries, e.g., <span class="math inline">p_{AC} = p_Ap_C</span>. For homozygote ALT, we need to account for the <em>homozygous SNP rate</em> <span class="math inline">r = 0.0005</span>, where G/G mutates to X/X, for X one of {A,C,T}:</p>
<p><span class="math display">\begin{align}
p_{AA} &amp; = p_Ap_A + 4r/6 = 3.33\times10^{-4} \\
p_{AC} &amp; = p_Ap_C = 1.11\times10^{-7}\\
p_{AT} &amp; = p_Ap_T = 1.11\times10^{-7}\\
p_{CC} &amp; = p_Cp_C + r/6 = 8.34\times10^{-5} \\
p_{CG} &amp; = p_Cp_G = 1.67\times10^{-4}\\
p_{CT} &amp; = p_Cp_T = 2.78\times10^{-8}\\
p_{GG} &amp; = 1 - f - r = 0.9985 \\
p_{GT} &amp; = p_Gp_T = 1.67\times10^{-4}\\
p_{TT} &amp; = p_Tp_T + r = 8.34\times10^{-5}\\
\end{align}</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genotype-likelihoods" class="slide level2 ultitle">
<h2>Genotype likelihoods</h2>
<p>We have outlined a probabilistic approach to variant calling where we obtain a posterior probability of observing a genotype <span class="math inline">G</span> given data <span class="math inline">X</span>:</p>
<p><span class="math display">
P(G|X) \sim P(X|G)P(G)
</span></p>
<div class="fragment">
<p>Assuming a bi-allelic site, and letting <span class="math inline">H_1, H_2</span> denote the two alleles, we have three possible genotype likelihoods <span class="math inline">P(H_1H_1|X)</span>, <span class="math inline">P(H_1H_2|X)</span>, and <span class="math inline">P(H_2H_2|X)</span>.</p>
</div>
<div class="fragment">
<p>The <em>highest</em> posterior probability is typically chosen as the genotype call, with a measure confidence represented by the genotype probability or ratio between the two most probable calls.</p>
</div>
<div class="fragment">
<p>Genotype likelihoods are often represented as Phred-scaled likelihoods (again!):</p>
<p><span class="math display">
\text{QUAL} = -10 \log_{10} P(G|X)
</span></p>
</div>
</section>
<section id="variant-call-format-vcf---header" class="slide level2 smaller ultitle">
<h2>Variant Call Format (VCF) - header</h2>
<div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> vcf/allsites.vcf.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 1</span>
<span id="cb1-2"><a></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> vcf/allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##FILTER"</span></span>
<span id="cb1-3"><a></a><span class="ex">bcftools</span> view <span class="at">-h</span> vcf/allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##INFO"</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 4</span>
<span id="cb1-4"><a></a><span class="ex">bcftools</span> view <span class="at">-h</span> vcf/allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##FORMAT"</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>##fileformat=VCFv4.2
##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;
##FILTER=&lt;ID=LowQual,Description="Low quality"&gt;
##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes, for each ALT allele, in the same order as listed"&gt;
##INFO=&lt;ID=AF,Number=A,Type=Float,Description="Allele Frequency, for each ALT allele, in the same order as listed"&gt;
##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;
##INFO=&lt;ID=BaseQRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities"&gt;
##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed"&gt;
##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)"&gt;
##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description="Genotype Quality"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##FORMAT=&lt;ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block"&gt;
##FORMAT=&lt;ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles"&gt;
##FORMAT=&lt;ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group"&gt;
##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification"&gt;</code></pre>
</div>
</div>
<p><code>FILTER</code> defines applied filters , <code>INFO</code> fields provide additional information to genotypes, <code>FORMAT</code> specification fields define genotype entries, and more. NB: <code>PL</code> format definition.</p>
</div>
<div class="flushright smallr">
<p><a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf" class="uri">https://samtools.github.io/hts-specs/VCFv4.4.pdf</a></p>
</div>
</section>
<section id="variant-call-format-vcf---data" class="slide level2 ultitle">
<h2>Variant Call Format (VCF) - data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> <span class="at">--samples</span> PUN-R-ELF,PUN-Y-INJ vcf/allsites.vcf.gz <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb3-2"><a></a> <span class="fu">tail</span> <span class="at">--lines</span> 1</span>
<span id="cb3-3"><a></a><span class="ex">bcftools</span> view <span class="at">--no-header</span>  <span class="at">--samples</span> PUN-R-ELF,PUN-Y-INJ vcf/allsites.vcf.gz LG4:6886</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  PUN-R-ELF   PUN-Y-INJ
LG4 6886    .   C   T   804.48  .   AC=2;AF=0.222;AN=4;BaseQRankSum=0;DP=82;ExcessHet=1.8123;FS=1.309;MLEAC=4;MLEAF=0.222;MQ=60;MQRankSum=0;QD=20.63;ReadPosRankSum=0.577;SOR=0.44  GT:AD:DP:GQ:PGT:PID:PL:PS   0/1:3,8:11:89:.:.:293,0,89:.    0/1:2,2:4:35:.:.:35,0,35:.</code></pre>
</div>
</div>
<div>
<p><code>QUAL</code>: Phred-scaled quality score for Prob(ALT is wrong): <span class="math inline">722.43</span> (<span class="math inline">p=10^{-Q/10}=5.7e-73</span>)</p>
<p><code>INFO</code> field summarizes data for all samples. For instance:</p>
<ul>
<li>allele count 2 (<code>AC=2</code>)</li>
<li>allele frequency minor allele 0.222 (<code>AF=0.222</code>)</li>
</ul>
</div>
<div class="flushright smallest">
<p><a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf" class="uri">https://samtools.github.io/hts-specs/VCFv4.4.pdf</a></p>
</div>
</section>
<section id="variant-call-format-vcf---data-1" class="slide level2 smaller ultitle">
<h2>Variant Call Format (VCF) - data</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  PUN-R-ELF   PUN-Y-INJ
LG4 6886    .   C   T   804.48  .   AC=2;AF=0.222;AN=4;BaseQRankSum=0;DP=82;ExcessHet=1.8123;FS=1.309;MLEAC=4;MLEAF=0.222;MQ=60;MQRankSum=0;QD=20.63;ReadPosRankSum=0.577;SOR=0.44  GT:AD:DP:GQ:PGT:PID:PL:PS   0/1:3,8:11:89:.:.:293,0,89:.    0/1:2,2:4:35:.:.:35,0,35:.</code></pre>
</div>
</div>
<div class="columns">
<div class="column" style="width:80%;">
<h5>
Genotypes (<code>GT:AD:DP:GQ:PGT:PID:PL:PS</code>)
</h5>
<p>PUN-R-ELF: <code>0/1:3,8:11:50:.:.:189,0,50:.</code></p>
<p><code>GT=0/1</code>, <code>AD=3,8</code> =&gt; 3 REF, 8 ALT, <code>DP=11</code> =&gt; sequence depth = 11, <code>PL=189,0,50</code></p>
<p>PUN-Y-INJ: <code>0/1:2,2:4:45:.:.:45,0,45:.</code></p>
<p><code>GT=0/1</code>, <code>AD=2,2</code> =&gt; 2 REF, 2 ALT, <code>DP=4</code> =&gt; sequence depth = 4, <code>PL=45,0,45</code></p>
<div class="fragment" data-fragment-index="2">
<h5>
Relative genotype probabilities
</h5>
<p>Can convert Phred-scaled quality scores to probabilities as</p>
<p><span class="math display">
p = 10^{-Q/10}
</span></p>
<p>For PUN-R-ELF the relative probabilities are <span class="math inline">10^{-189/10}\approx1.26e-9</span>, <span class="math inline">10^{0}=1</span>, <span class="math inline">10^{50}=10^{-5}</span>.</p>
<p>Interpretation: <code>0/1</code> 10,000 times more likely than <code>1/1</code> (<span class="math inline">1/10^{-5}</span>)</p>
</div>
</div><div class="column" style="width:3%;">

</div><div class="column" style="width:4%;">


<title>
ELF
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
ELF
</div>
<pre class="tviewpre"><span>   </span><br><span>TCT</span><br><span class="tviewcu4">.Y.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">...</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:4%;">


<title>
INJ
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
INJ
</div>
<pre class="tviewpre"><span>   </span><br><span>TCT</span><br><span class="tviewcu4">.</span><span class="tviewcu3">Y</span><span class="tviewcu4">.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span></pre>
</div>


</div></div>
<aside class="notes">
<p>QUAL: Phred-scaled quality score for the assertion made in ALT. i.e. <span class="math inline">-10log_{10}</span> prob(call in ALT is wrong)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="gatk-best-practice" class="slide level2 ultitle">
<h2>GATK best practice</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="assets/images/gatk-best-practice.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img data-src="assets/images/gatk-best-practice.webp" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
<div class="flushright smallest">
<p><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-</a></p>
</div>
</div><div class="column" style="width:30%;">
<h5>
Pros
</h5>
<ul>
<li>Best practices</li>
<li>Large documentation</li>
<li>Variant quality score recalibration</li>
</ul>
<h5>
Cons
</h5>
<ul>
<li>Human-centric - very slow runtime on genomes with many sequences</li>
<li>Complicated setup</li>
</ul>
</div></div>
</section>
<section id="alternative-variant-callers" class="slide level2 ultitle">
<h2>Alternative variant callers</h2>
<div class="columns">
<div class="column" style="width:33%;">
<h5 id="freebayes"><a href="https://github.com/freebayes/freebayes">freebayes</a></h5>
<p>Bayesian genetic variant detector. Simpler setup.</p>
<p>May struggle in high-coverage regions.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="garrison_HaplotypebasedVariantDetection_2012">(<a href="#/bibliography" role="doc-biblioref" onclick="">Garrison &amp; Marth, 2012</a>)</span></p>
</div>
</div><div class="column" style="width:33%;">
<h5 id="bcftools"><a href="https://github.com/samtools/bcftools">bcftools</a></h5>
<p>Utilities for variant calling and manipulating VCFs and BCFs.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="danecek_TwelveYearsSAMtools_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Danecek et al., 2021</a>)</span></p>
</div>
</div><div class="column" style="width:33%;">
<h5 id="angsd"><a href="http://www.popgen.dk/angsd/index.php/ANGSD">ANGSD</a></h5>
<p>For low-coverage sequencing. Doesn’t do explicit genotyping; most methods take genotype uncertainty into account.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="korneliussen_ANGSDAnalysisNext_2014">(<a href="#/bibliography" role="doc-biblioref" onclick="">Korneliussen et al., 2014</a>)</span></p>
</div>
</div></div>
<div class="hidden">
<p>Reference bias: plot no. hets vs coverage for real data, e.g., conifer</p>
</div>
<aside class="notes">
<p>NB: samtools and GATK may actually produce different genotypes despite having identical GLs. Samtools applies prior <span class="math inline">10^{-3}</span> to het call, GATK has no prior <span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">(<a href="#/bibliography" role="doc-biblioref" onclick="">H. Li, 2014</a>)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="bibliography" class="slide level2 smaller ultitle scrollable">
<h2>Bibliography</h2>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-danecek_TwelveYearsSAMtools_2021" class="csl-entry" role="listitem">
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., Davies, R. M., &amp; Li, H. (2021). Twelve years of <span>SAMtools</span> and <span>BCFtools</span>. <em>GigaScience</em>, <em>10</em>(2), giab008. <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>
</div>
<div id="ref-depristo_FrameworkVariationDiscovery_2011" class="csl-entry" role="listitem">
DePristo, M. A., Banks, E., Poplin, R., Garimella, K. V., Maguire, J. R., Hartl, C., Philippakis, A. A., del Angel, G., Rivas, M. A., Hanna, M., McKenna, A., Fennell, T. J., Kernytsky, A. M., Sivachenko, A. Y., Cibulskis, K., Gabriel, S. B., Altshuler, D., &amp; Daly, M. J. (2011). A framework for variation discovery and genotyping using next-generation <span>DNA</span> sequencing data. <em>Nature Genetics</em>, <em>43</em>(5), 491–498. <a href="https://doi.org/10.1038/ng.806">https://doi.org/10.1038/ng.806</a>
</div>
<div id="ref-garrison_HaplotypebasedVariantDetection_2012" class="csl-entry" role="listitem">
Garrison, E., &amp; Marth, G. (2012). Haplotype-based variant detection from short-read sequencing. <em>arXiv:1207.3907 [q-Bio]</em>. <a href="http://arxiv.org/abs/1207.3907">http://arxiv.org/abs/1207.3907</a>
</div>
<div id="ref-korneliussen_ANGSDAnalysisNext_2014" class="csl-entry" role="listitem">
Korneliussen, T. S., Albrechtsen, A., &amp; Nielsen, R. (2014). <span>ANGSD</span>: <span>Analysis</span> of <span>Next Generation Sequencing Data</span>. <em>BMC Bioinformatics</em>, <em>15</em>(1), 356. <a href="https://doi.org/10.1186/s12859-014-0356-4">https://doi.org/10.1186/s12859-014-0356-4</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-li_SNPDetectionMassively_2009" class="csl-entry" role="listitem">
Li, R., Li, Y., Fang, X., Yang, H., Wang, J., Kristiansen, K., &amp; Wang, J. (2009). <span>SNP</span> detection for massively parallel whole-genome resequencing. <em>Genome Research</em>, <em>19</em>(6), 1124–1132. <a href="https://doi.org/10.1101/gr.088013.108">https://doi.org/10.1101/gr.088013.108</a>
</div>
<div id="ref-maruki_GenotypeCallingPopulationGenomic_2017" class="csl-entry" role="listitem">
Maruki, T., &amp; Lynch, M. (2017). Genotype <span>Calling</span> from <span>Population-Genomic Sequencing Data</span>. <em>G3 Genes|Genomes|Genetics</em>, <em>7</em>(5), 1393–1404. <a href="https://doi.org/10.1534/g3.117.039008">https://doi.org/10.1534/g3.117.039008</a>
</div>
<div id="ref-nielsen_GenotypeSNPCalling_2011" class="csl-entry" role="listitem">
Nielsen, R., Paul, J. S., Albrechtsen, A., &amp; Song, Y. S. (2011). Genotype and <span>SNP</span> calling from next-generation sequencing data. <em>Nature Reviews Genetics</em>, <em>12</em>(6), 443–451. <a href="https://doi.org/10.1038/nrg2986">https://doi.org/10.1038/nrg2986</a>
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Variant calling</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>