<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/all.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/latex-fontsize.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Genetic diversity</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-dd75e976e0414ec343b1912506d158bc.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Genetic diversity</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="what-determines-diversity-levels" class="slide level2 ultitle">
<h2>What determines diversity levels?</h2>
<div class="columns">
<div class="column" style="width:30%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="assets/images/leffler-2012-fig1.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="assets/images/leffler-2012-fig1.webp" class="quarto-figure quarto-figure-center" style="width:55.0%"></a></p>
</figure>
</div>
</div><div class="column" style="width:70%;">
<p>The usual questions:</p>
<blockquote>
<p>What evolutionary forces maintain genetic diversity in natural populations? How do diversity levels relate to census population sizes…? Do low levels of diversity limit adaptation to selective pressures?</p>
</blockquote>
<div class="flushright smallest">
<p><span class="citation" data-cites="leffler_RevisitingOldRiddle_2012">Leffler et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2012</a>)</span></p>
</div>
<p><br></p>
<div class="fragment">
<p>After <em>allozyme</em> era, the study of genetic diversity was largely neglected due to lack of genome-wide data, but with advent of population genomics becoming a hot topic again.</p>
<div class="flushright smallest">
<p><span class="citation" data-cites="ellegren_DeterminantsGeneticDiversity_2016">Ellegren &amp; Galtier (<a href="#/bibliography" role="doc-biblioref" onclick="">2016</a>)</span></p>
</div>
<p><br></p>
</div>
<div class="fragment">
<p>Lewontin’s paradox: genetic diversity range smaller than variation among species in population size</p>
</div>
</div></div>
<aside class="notes">
<p>Riddle: why is genetic diversity range so narrow? Possibly there are lower and upper limits.</p>
<p>Lower limit: censoring effect, i.e., when diversity passes a lower limit, population is driven to extinction due to inability to adapt</p>
<p>Upper limit: functional/structural constraints, e.g., impaired chromosome pairing or reproductive incompatibilities</p>
<p>Linked selection, that is, variation-reducing selection. Larger populations have higher influx of new mutations, so if more draft, higher reduction in diversity. Would require strong frequent levels of selection; little support for this in literature.</p>
<p>Purifying selection. Nearly neutral model could explain narrow range.</p>
<p><span class="citation" data-cites="ellegren_DeterminantsGeneticDiversity_2016">(<a href="#/bibliography" role="doc-biblioref" onclick="">Ellegren &amp; Galtier, 2016</a>)</span> Upper limit could in part be explained by linked selection <span class="citation" data-cites="corbett-detig_NaturalSelectionConstrains_2015">(<a href="#/bibliography" role="doc-biblioref" onclick="">Corbett-Detig et al., 2015</a>)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="factors-that-influence-genetic-diversity" class="slide level2 ultitle">
<h2>Factors that influence genetic diversity</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-genetic-diversity-determinants-Ellegren-fig2" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-genetic-diversity-determinants-Ellegren-fig2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg.2016.58/MediaObjects/41576_2016_Article_BFnrg201658_Fig2_HTML.jpg?as=webp" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;1: Overview of determinants of genetic diversity [@ellegren_DeterminantsGeneticDiversity_2016, Fig 2]"><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg.2016.58/MediaObjects/41576_2016_Article_BFnrg201658_Fig2_HTML.jpg?as=webp" style="width:90.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-genetic-diversity-determinants-Ellegren-fig2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Overview of determinants of genetic diversity <span class="citation" data-cites="ellegren_DeterminantsGeneticDiversity_2016">(<a href="#/bibliography" role="doc-biblioref" onclick="">Ellegren &amp; Galtier, 2016</a>, Fig 2)</span>
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<h4 id="genetic-drift">Genetic drift</h4>
<p>Reduces diversity at loss <span class="math inline">\propto \frac{1}{N}</span></p>
<h4 id="selection">Selection</h4>
<p>Adaptive selection decreases variation, more so if acting on new mutations compared to standing variation.</p>
<p>Balancing selection may increase variation.</p>
<h4 id="recombination">Recombination</h4>
<p>Low recombination rates lead to less “reshuffling” of variation and hence lower diversity.</p>
</div></div>
<aside class="notes">
<p>Note: lower recombination lowers diversity presumably because more neutral variants linked to selected sites. There are intriguing differences between taxa <span class="citation" data-cites="leffler_RevisitingOldRiddle_2012">(<a href="#/bibliography" role="doc-biblioref" onclick="">Leffler et al., 2012</a>)</span></p>
<p><span class="citation" data-cites="ellegren_DeterminantsGeneticDiversity_2016">Ellegren &amp; Galtier (<a href="#/bibliography" role="doc-biblioref" onclick="">2016</a>)</span> point out that with advent of population genomic data, we are at the brink of understanding Lewontin’s paradox, namely <span class="citation" data-cites="charlesworth_HowCanWe_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Charlesworth &amp; Jensen, 2022</a>)</span></p>
<blockquote>
<p>the observation that the range of levels of genetic diversity in natural populations appears to be far smaller than the extent of variation among species in population size</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="why-we-measure-patterns-of-genetic-variation" class="slide level2 ultitle">
<h2>Why we measure patterns of genetic variation</h2>
<p>Genetic variation patterns are informative of evolutionary and demographic processes. We often use <strong>summary statistics</strong> to describe the patterns, and to estimate parameters such as <strong>effective population size</strong> and <strong>mutation rate</strong> from genetic variation data (<span class="math inline">\theta = 4N_e\mu</span>)</p>
<div class="fragment">
<div>
<p>Often critical first step of analysis, such as</p>
<ul>
<li>exploratory study</li>
<li>test of an evolutionary hypothesis</li>
<li>training of machine-learning models</li>
</ul>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="korunes_PixyUnbiasedEstimation_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Korunes &amp; Samuk, 2021</a>)</span></p>
</div>
</div>
</section>
<section id="genetic-diversity-in-conservation-biology" class="slide level2 smaller ultitle">
<h2>Genetic diversity in conservation biology</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-kuderna-2023" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-kuderna-2023-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/kuderna-2023-fig2.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;2: @kuderna_GlobalCatalogWholegenome_2023, Fig 2"><img data-src="assets/images/kuderna-2023-fig2.webp" class="quarto-figure quarto-figure-center" style="width:100.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-kuderna-2023-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: <span class="citation" data-cites="kuderna_GlobalCatalogWholegenome_2023">Kuderna et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2023</a>)</span>, Fig 2
</figcaption>
</figure>
</div>
<blockquote>
<p>…no global relationship between numerically coded IUCN extinction risk categories and estimated heterozygosity…</p>
</blockquote>
</div><div class="column" style="width:50%;">
<div>
<p>Low genetic diversity symptom of past genetic drift inbreeding (higher levels of homozygosity), caused by low <span class="math inline">N_e</span></p>
<div class="flushright smallest">
<p><span class="citation" data-cites="garcia-dorado_NeutralGeneticDiversity_2021">García-Dorado &amp; Caballero (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span></p>
</div>
</div>
<p><br><br></p>
<div>
<p>However: if population decline is rapid, may be too little time for inbreeding to occur <span class="math inline">\Rightarrow</span> genetic diversity within species not necessarily aligned to extinction risk</p>
<div class="flushright smallest">
<p><span class="citation" data-cites="lewis_BiggestEverStudy_2023">Lewis (<a href="#/bibliography" role="doc-biblioref" onclick="">2023</a>)</span></p>
</div>
</div>
</div></div>
<aside class="notes">
<p>Take home <span class="citation" data-cites="lewis_BiggestEverStudy_2023">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lewis, 2023</a>)</span>: genetic diversity within a species does not necessarily align with its extinction risk. Possible explanation: population decline has been so rapid that there hasn’t been time for inbreeding to occur. Other factors (e.g., habitat destruction) greater threat.</p>
<p>Caption:</p>
<blockquote>
<p>Runs of homozygosity and impact of extinction risk on diversity (A) Relationship between IUCN extinction risk categories and heterozygosity. Solid black circles and bars denote median and IQR. (B) Partition into threatened (T: VU, EN, CR) and nonthreatened (N: LC, NT) categories for all families with more than one species in either partition. Significant differences (p &lt; 0.05, one-sided rank-sum test) are marked with an asterisk. (C) Median number of tracts of homozygosity versus median proportion of the genome in runs of homozygosity per species. Species with a fraction over 1/3 are highlighted. Solid black dots within highlights denote threatened species (VU, EN, CR).</p>
</blockquote>
<p>On Figure 2A:</p>
<blockquote>
<p>We investigated whether genetic diversity estimates are correlated with extinction risk in primates, a subject of previous debate (17, 33, 34). Despite our broad sampling, we find no global relationship between numerically coded IUCN extinction risk categories and estimated heterozygosity [p &gt; 0.05, phylogenetic generalized least squares (PGLS)] (Fig. 2A) (16).</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section>
<section id="measuring-genetic-diversity" class="title-slide slide level1 center">
<h1>Measuring genetic diversity</h1>

</section>
<section id="nucleotide-diversity" class="slide level2 ultitle">
<h2>Nucleotide diversity</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>sample 0:   00100
sample 1:   00001
sample 2:   01010
sample 3:   10010</code></pre>
</div>
</div>
<div class="fragment" data-fragment-index="2">
<p>Diversity: for each site, count and sum differences between all (unique) pairs of samples, and divide by unique pairs. For <span class="math inline">n</span> samples, there are <span class="math inline">n \choose 2</span> such pairs.</p>
</div>
<div class="fragment" data-fragment-index="3">
<p>Example: for site 0, start comparing samples 0-1 (0 diffs), samples 0-2 (0), samples 0-3 (1), samples 1-2 (0) and so on. Call these differences <span class="math inline">k_{ij}</span>. Then</p>
</div>
<div class="fragment" data-fragment-index="4">
<p><span class="math display">
\pi = \frac{\sum_{i&lt;j}k_{ij}}{n \choose 2}
</span></p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment" data-fragment-index="1">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreediversity" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreediversity {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div>
</div></div>
</section>
<section id="wattersons-theta_w" class="slide level2 ultitle">
<h2>Watterson’s <span class="math inline">\theta_W</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>sample 0:   00100
sample 1:   00001
sample 2:   01010
sample 3:   10010</code></pre>
</div>
</div>
<div class="fragment" data-fragment-index="1">
<p>Alternative measure of diversity: simply count the number of segregating sites (<span class="math inline">S</span>). However, must correct for the number of samples <span class="math inline">n</span> as we expect that more samples <span class="math inline">\Rightarrow</span> more sites</p>
</div>
<div class="fragment" data-fragment-index="2">
<p><span class="math display">
\theta_W = \frac{S}{a} = \frac{S}{\sum_{i=1}^{n-1}\frac{1}{i}}
</span></p>
</div>
<div class="fragment" data-fragment-index="3">
<p>Important: under neutrality, <span class="math inline">\theta = E(\pi) = E(\theta_W)</span>. Difference between two the basis for <strong>Tajima’s D</strong> that is a test for selection</p>
</div>
</div><div class="column" style="width:50%;">
<div>
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreethetaw" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreethetaw {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div>
</div></div>
<aside class="notes">
<p>Intuitive example to understand of Tajimas: balancing selection has long internal branches -&gt; long haplotype. This will elevate pi since many samples will be different for many sites.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="calculating-diversity-measures---pi-and-theta_w" class="slide level2 ultitle">
<h2>Calculating diversity measures - <span class="math inline">\pi</span> and <span class="math inline">\theta_W</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>sample 0:   00100
sample 1:   00001
sample 2:   01010
sample 3:   10010</code></pre>
</div>
</div>
<div class="small">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a></a><span class="im">from</span> trees <span class="im">import</span> ts_small_mut <span class="im">as</span> ts</span>
<span id="cb4-2"><a></a></span>
<span id="cb4-3"><a></a><span class="co"># Calculate correction factor a for Watterson's</span></span>
<span id="cb4-4"><a></a><span class="co"># theta: the larger the sample size, the more</span></span>
<span id="cb4-5"><a></a><span class="co"># segregating sites we expect to see</span></span>
<span id="cb4-6"><a></a>a <span class="op">=</span> <span class="bu">sum</span>([<span class="dv">1</span><span class="op">/</span>i <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, ts.num_samples)])</span>
<span id="cb4-7"><a></a></span>
<span id="cb4-8"><a></a>pi <span class="op">=</span> ts.diversity()</span>
<span id="cb4-9"><a></a>thetaW <span class="op">=</span> ts.num_sites <span class="op">/</span> a <span class="op">/</span> ts.sequence_length</span>
<span id="cb4-10"><a></a></span>
<span id="cb4-11"><a></a><span class="bu">print</span>(<span class="ss">f"Diversity:           </span><span class="sc">{</span>pi<span class="sc">:.6f}</span><span class="ss">"</span>,</span>
<span id="cb4-12"><a></a>      <span class="ss">f"Watterson's theta:   </span><span class="sc">{</span>thetaW<span class="sc">:.6f}</span><span class="ss">"</span>,</span>
<span id="cb4-13"><a></a>      <span class="ss">f"Sequence length:     </span><span class="sc">{</span>ts<span class="sc">.</span>sequence_length<span class="sc">:.0f}</span><span class="ss">"</span>,</span>
<span id="cb4-14"><a></a>      sep<span class="op">=</span><span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Diversity:           0.000267
Watterson's theta:   0.000273
Sequence length:     10000</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<div>
<svg baseprofile="full" height="500" id="msprimecalctreediversity" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimecalctreediversity {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}}</style>
</defs><g class="tree-sequence"><g class="background"><path d="M20,0 l460,0 l0,338.2 l0,25 l0,5 l-460,0 l0,-5 l0,-25 l0,-338.2z"></path></g><g class="axes"><g class="x-axis"><g class="title" transform="translate(250 400)"><text class="lab" text-anchor="middle" transform="translate(0 -11)">Genome position</text></g><line class="ax-line" x1="20" x2="480" y1="363.2" y2="363.2"></line><g class="ticks"><g class="tick" transform="translate(20 363.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">0</text></g></g><g class="tick" transform="translate(480 363.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">10000</text></g></g></g><g class="site s0" transform="translate(22.852 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m0"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s1" transform="translate(119.36 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m1"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s2" transform="translate(271.712 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m2"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s3" transform="translate(311.134 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m3"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s4" transform="translate(348.808 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m4"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g></g></g><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 283.291)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 11.3093)"><path class="edge" d="M 0 0 V -11.3093 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -1.95306)"><line x1="0" x2="0" y1="0" y2="1.95306"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 11.3093)"><path class="edge" d="M 0 0 V -11.3093 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -6.88775)"><line x1="0" x2="0" y1="0" y2="6.88775"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -283.291 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 169.503)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 125.097)"><path class="edge" d="M 0 0 V -125.097 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -17.6968)"><line x1="0" x2="0" y1="0" y2="17.6968"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 125.097)"><path class="edge" d="M 0 0 V -125.097 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -87.3704)"><line x1="0" x2="0" y1="0" y2="87.3704"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -169.503 H -105"></path><g class="mut m3 s3" transform="translate(0 -132.108)"><line x1="0" x2="0" y1="0" y2="132.108"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div>
</div></div>
</section>
<section id="divergence---dxy" class="slide level2 ultitle">
<h2>Divergence - dXY</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Sample 0:   00100
Sample 1:   00001

Sample 2:   01010
Sample 3:   10010</code></pre>
</div>
</div>
<p>Divergence: for each site, count and sum differences between all pairs of samples <em>between</em> two populations</p>
<div class="fragment">
<p>Example: for site 0, compare samples 0-2 (0 diffs), samples 0-3 (1 diff), samples 1-2 (0 diffs), samples 1-3 (1 diff), and so on. Call differences <span class="math inline">k_ij</span>, let <span class="math inline">n_X</span>, <span class="math inline">n_Y</span> be sample size in populations <span class="math inline">X</span>, <span class="math inline">Y</span>. Then</p>
</div>
<div class="fragment">
<p><span class="math display">
d_{XY} = \frac{1}{n_Xn_Y}\sum_{i=1}^{n_X}\sum_{i=1}^{n_Y}k_{ij}
</span></p>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreedivergence" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreedivergence {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}.n4 .edge {stroke: blue }}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div></div>
</section>
<section id="divergence---dxy-1" class="slide level2 ultitle">
<h2>Divergence - dXY</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Sample 0:   00100
Sample 1:   00001

Sample 2:   01010
Sample 3:   10010</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a></a><span class="im">from</span> trees <span class="im">import</span> ts_small_mut <span class="im">as</span> ts</span>
<span id="cb8-2"><a></a></span>
<span id="cb8-3"><a></a>sample_sets <span class="op">=</span> [[<span class="dv">0</span>, <span class="dv">1</span>], [<span class="dv">2</span>, <span class="dv">3</span>]]</span>
<span id="cb8-4"><a></a></span>
<span id="cb8-5"><a></a>dxy <span class="op">=</span> ts.divergence(sample_sets<span class="op">=</span>sample_sets)</span>
<span id="cb8-6"><a></a></span>
<span id="cb8-7"><a></a><span class="bu">print</span>(<span class="ss">f"Divergence:   </span><span class="sc">{</span>dxy<span class="sc">:.6f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Divergence:   0.000300</code></pre>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreedivergence" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreedivergence {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}.n4 .edge {stroke: blue }}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div></div>
<aside class="notes">
<p><span class="math inline">F_ST</span> (later) and related statistics are strongly affected by within-subpopulation genetic variance. <span class="citation" data-cites="nei_AnalysisGeneDiversity_1973">Nei (<a href="#/bibliography" role="doc-biblioref" onclick="">1973</a>)</span> proposed dXY as an alternative that does not depend on the levels of diversity <em>within</em> subpopulations.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="differentiation---afd-allele-frequency-difference" class="slide level2 smaller ultitle">
<h2>Differentiation - AFD (allele frequency difference)</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Sample 0:   00100
Sample 1:   00001

Sample 2:   01010
Sample 3:   10010</code></pre>
</div>
</div>
<p>Allele Frequency Difference (AFD) proposed as intuitive alternative to <span class="math inline">F_{\mathrm{ST}}</span>. For each site, count the difference in allele frequency between two populations.</p>
<div class="fragment">
<p>Example: site 0, frequency in blue is 0, in black 1/2, so difference=1/2, site 1, frequency in blue 0, in black 1/2, and so on</p>
</div>
<div class="fragment">
<p><span class="math display">
AFD = \frac{1}{2}\sum_{i=1}^n| (f_{i1} - f_{i2})|
</span></p>
<p>where <span class="math inline">n</span> is the number of different alleles (<span class="math inline">n=2</span> for biallelic SNPs), <span class="math inline">f_{ij}</span> is the frequency of allele <span class="math inline">i</span> in population <span class="math inline">j</span></p>
<div class="flushright smallest">
<p><span class="citation" data-cites="berner_AlleleFrequencyDifference_2019">Berner (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span></p>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreeafd" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreeafd {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}.n4 .edge {stroke: blue }}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div></div>
<aside class="notes">
<p>AFD proposed as a more intuitive and easier to understand statistic than Fst</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="differentiation--" class="slide level2 smaller ultitle">
<h2>Differentiation - <span class="math inline">F_{\mathrm{ST}}</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Sample 0:   00100
Sample 1:   00001

Sample 2:   01010
Sample 3:   10010</code></pre>
</div>
</div>
<p><span class="math inline">F_{\mathrm{ST}}</span> is another measure of differention among subpopulations <span class="citation" data-cites="wright_EvolutionMendelianPopulations_1931">(<a href="#/bibliography" role="doc-biblioref" onclick="">Wright, 1931</a>)</span>. It ranges between 0 and 1 and has the interpretation 0: no differentiation, 1: complete fixation of alternate alleles in subpopulations</p>
<div class="fragment">
<p>Example: site 3 has <span class="math inline">F_{\mathrm{ST}}</span>=1 as it is fixed in black, not present in blue</p>
</div>
<div class="fragment">
<p>There are <em>many</em> ways to express and calculate <span class="math inline">F_{\mathrm{ST}}</span>. Example:</p>
<p><span class="math display">
F_{\mathrm{ST}} = \frac{h_{\mathrm{T}} - h_{\mathrm{S}}}{h_{\mathrm{T}}}
</span></p>
<p>where <span class="math inline">h_{\mathrm{T}}</span> is the expected heterozygosity in the total population, <span class="math inline">h_{\mathrm{S}}</span> the <em>average</em> of expected heterozygosities across subpopulations. For site 3, <span class="math inline">h_{\mathrm{S}}=0</span>, <span class="math inline">h_{\mathrm{T}}=2/3</span>.</p>
</div>
<div class="fragment">
<p>Caveats: strongly influenced by <em>within</em> subpopulation levels of variation. Therefore considered <em>relative</em> measure (cf <span class="math inline">d_{\mathrm{XY}}</span>, which is an <em>absolute</em> measure).</p>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreefst" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreefst {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}.n4 .edge {stroke: blue }}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 318.678)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -2.19703)"><line x1="0" x2="0" y1="0" y2="2.19703"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 12.722)"><path class="edge" d="M 0 0 V -12.722 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -7.74814)"><line x1="0" x2="0" y1="0" y2="7.74814"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -318.678 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 190.676)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -19.9074)"><line x1="0" x2="0" y1="0" y2="19.9074"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 140.724)"><path class="edge" d="M 0 0 V -140.724 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -98.2843)"><line x1="0" x2="0" y1="0" y2="98.2843"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -190.676 H -105"></path><g class="mut m3 s3" transform="translate(0 -148.61)"><line x1="0" x2="0" y1="0" y2="148.61"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div></div>
<aside class="notes">
<p>Wright’s fst measures differentiation among subpopulations. Interpretation: 0 no differentiation, 1 complete fixation of alternate alleles. Has therefore become a popular statistic for quantifying differentiation from molecular data.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="differentiation---1" class="slide level2 smaller ultitle">
<h2>Differentiation - <span class="math inline">F_{\mathrm{ST}}</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Sample 0:   00100
Sample 1:   00001

Sample 2:   01010
Sample 3:   10010</code></pre>
</div>
</div>
<div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a></a><span class="im">from</span> trees <span class="im">import</span> ts_small_mut <span class="im">as</span> ts</span>
<span id="cb13-2"><a></a></span>
<span id="cb13-3"><a></a>sample_sets <span class="op">=</span> [[<span class="dv">0</span>, <span class="dv">1</span>], [<span class="dv">2</span>, <span class="dv">3</span>]]</span>
<span id="cb13-4"><a></a>win <span class="op">=</span> [<span class="dv">0</span>] <span class="op">+</span> [<span class="bu">int</span>(x.position)<span class="op">+</span><span class="dv">1</span> <span class="cf">for</span> x <span class="kw">in</span> ts.sites()]</span>
<span id="cb13-5"><a></a>_ <span class="op">=</span> win.pop()</span>
<span id="cb13-6"><a></a>win <span class="op">=</span> win <span class="op">+</span> [ts.sequence_length]</span>
<span id="cb13-7"><a></a></span>
<span id="cb13-8"><a></a>fst <span class="op">=</span> ts.Fst(sample_sets<span class="op">=</span>sample_sets)</span>
<span id="cb13-9"><a></a>fst_sites <span class="op">=</span> ts.Fst(sample_sets<span class="op">=</span>sample_sets,</span>
<span id="cb13-10"><a></a>                   windows<span class="op">=</span>win)</span>
<span id="cb13-11"><a></a></span>
<span id="cb13-12"><a></a><span class="bu">print</span>(<span class="ss">f"Site id:       </span><span class="sc">{</span><span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">5</span>))<span class="sc">}</span><span class="ss">"</span>,</span>
<span id="cb13-13"><a></a>      <span class="ss">f"Fst per site:  </span><span class="sc">{</span>fst_sites<span class="sc">}</span><span class="ss">"</span>,</span>
<span id="cb13-14"><a></a>      <span class="ss">f"Overall Fst:   </span><span class="sc">{</span>fst<span class="sc">:.6f}</span><span class="ss">"</span>,</span>
<span id="cb13-15"><a></a>      sep<span class="op">=</span><span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Site id:       [0, 1, 2, 3, 4]
Fst per site:  [0. 0. 0. 1. 0.]
Overall Fst:   0.200000</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="smallest">
<p>Tree plot with mutations and sequences <a href="../../recipes/slides/index.html#sec-recipe-diversity-msprime-trees"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></p>
</div>
<svg baseprofile="full" height="500" id="msprimetreefst" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}#msprimetreefst {.edge {stroke: black; stroke-width: 3px}.node > .lab {text-anchor: end; transform: rotate(-45deg) translate(-5px, 8px);}.node > .sym, .node:not(.leaf) > .lab {display: none}.node:not(.leaf) > .lab {display: none}.n4 .edge {stroke: blue }}</style>
</defs><g class="tree-sequence"><g class="background"><path d="M20,0 l460,0 l0,338.2 l0,25 l0,5 l-460,0 l0,-5 l0,-25 l0,-338.2z"></path></g><g class="axes"><g class="x-axis"><g class="title" transform="translate(250 400)"><text class="lab" text-anchor="middle" transform="translate(0 -11)">Genome position</text></g><line class="ax-line" x1="20" x2="480" y1="363.2" y2="363.2"></line><g class="ticks"><g class="tick" transform="translate(20 363.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">0</text></g></g><g class="tick" transform="translate(480 363.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">10000</text></g></g></g><g class="site s0" transform="translate(22.852 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m0"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s1" transform="translate(119.36 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m1"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s2" transform="translate(271.712 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m2"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s3" transform="translate(311.134 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m3"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s4" transform="translate(348.808 363.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m4"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g></g></g><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n6 p0 root" transform="translate(230 26.8)"><g class="a6 c2 node n4 p0" transform="translate(-105 283.291)"><g class="a4 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-52.5 11.3093)"><path class="edge" d="M 0 0 V -11.3093 H 52.5"></path><g class="mut m2 s2" transform="translate(0 -1.95306)"><line x1="0" x2="0" y1="0" y2="1.95306"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0: 00100</text></g><g class="a4 i0 leaf m4 node n1 p0 s4 sample" transform="translate(52.5 11.3093)"><path class="edge" d="M 0 0 V -11.3093 H -52.5"></path><g class="mut m4 s4" transform="translate(0 -6.88775)"><line x1="0" x2="0" y1="0" y2="6.88775"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">4</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1: 00001</text></g><path class="edge" d="M 0 0 V -283.291 H 105"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7)"></text></g><g class="a6 c2 m3 node n5 p0 s3" transform="translate(105 169.503)"><g class="a5 i1 leaf m1 node n2 p0 s1 sample" transform="translate(-52.5 125.097)"><path class="edge" d="M 0 0 V -125.097 H 52.5"></path><g class="mut m1 s1" transform="translate(0 -17.6968)"><line x1="0" x2="0" y1="0" y2="17.6968"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2: 01010</text></g><g class="a5 i1 leaf m0 node n3 p0 s0 sample" transform="translate(52.5 125.097)"><path class="edge" d="M 0 0 V -125.097 H -52.5"></path><g class="mut m0 s0" transform="translate(0 -87.3704)"><line x1="0" x2="0" y1="0" y2="87.3704"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">0</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3: 10010</text></g><path class="edge" d="M 0 0 V -169.503 H -105"></path><g class="mut m3 s3" transform="translate(0 -132.108)"><line x1="0" x2="0" y1="0" y2="132.108"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7)"></text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div></div>
<aside class="notes">

<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="many-programs-treat-missing-data-as-invariant" class="slide level2 ultitle">
<h2>Many programs treat missing data as invariant</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://onlinelibrary.wiley.com/cms/asset/2d12ae05-e2cd-496e-a1d6-f3842bb59e22/men13326-fig-0001-m.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="@korunes_PixyUnbiasedEstimation_2021, Fig 1"><img data-src="https://onlinelibrary.wiley.com/cms/asset/2d12ae05-e2cd-496e-a1d6-f3842bb59e22/men13326-fig-0001-m.jpg" class="quarto-figure quarto-figure-center" style="width:80.0%" alt="Korunes &amp; Samuk (2021), Fig 1"></a></p>
</figure>
</div>
<figcaption><span class="citation" data-cites="korunes_PixyUnbiasedEstimation_2021">Korunes &amp; Samuk (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span>, Fig 1</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:50%;">
<p>Diversity:</p>
<p><span class="math display">
\pi = \frac{\sum_{i&lt;j}k_{ij}}{n \choose 2}
</span></p>
<p>Divergence:</p>
<p><span class="math display">
d_{XY} = \frac{1}{n_Xn_Y}\sum_{i=1}^{n_X}\sum_{j=1}^{n_Y}k_{ij}
</span></p>
<p>Here, <span class="math inline">n</span> is the number of samples, <span class="math inline">k_{ij}</span> tally of allelic differences between two haplotypes within (<span class="math inline">\pi</span>) a population or between (<span class="math inline">d_{XY}</span>) populations</p>
</div></div>
<aside class="notes">
<p>Paper caption:</p>
<blockquote>
<p>The logic and input/ouput of pixy demonstrated with a simple haploid example. (a) Comparison of two methods for computing π (or dXY) in the face of missing data. These methods follow the first expression of Equation 1 but differ in how they calculate the numerator and denominator. In Case 1, all missing data is assumed to be present but invariant. This results in a deflated estimate of π. In Case 2, missing data are simply omitted from the calculation, both in terms of the number of sites (the final denominator) and the component denominators for each site (the n choose two terms). This results in an unbiased estimate of π. (b) The adjusted π method (Case 2) as implemented for VCFs in pixy. The example VCF (input) contains the same four haplotypes as (a). Invariant sites are represented as sites with no ALT allele, and greyed-out sites are those that failed to pass a genotype filter requiring a minimum number of reads covering the genotype (Depth ≥ 10 in this case)</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="missing-data-may-bias-diversity-measures-downwards" class="slide level2 ultitle">
<h2>Missing data may bias diversity measures downwards</h2>
<div>
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://onlinelibrary.wiley.com/cms/asset/dc499ed3-0c1f-4260-95a1-cb85a0894d10/men13326-fig-0004-m.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="@korunes_PixyUnbiasedEstimation_2021, Fig 4"><img data-src="https://onlinelibrary.wiley.com/cms/asset/dc499ed3-0c1f-4260-95a1-cb85a0894d10/men13326-fig-0004-m.jpg" class="quarto-figure quarto-figure-center" style="width:45.0%" alt="Korunes &amp; Samuk (2021), Fig 4"></a></p>
</figure>
</div>
<figcaption><span class="citation" data-cites="korunes_PixyUnbiasedEstimation_2021">Korunes &amp; Samuk (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span>, Fig 4</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<aside class="notes">
<p>Real data: strong tendency of commonly adopted tools to <strong>underestimate</strong> pi, dxy (even for 30X data sets)</p>
<p>Caption:</p>
<blockquote>
<p>Comparisons of estimates of π from whole genome data derived from 18 Anopheles gambiae individuals from the Ag1000G Burkina Faso (BFS) population. Each panel (a–d) depicts the estimates of π for the X chromosome performed using pixy (y-axis) and four other methods (x-axis, a–d). Points are coloured according to the proportion of missing data (of any type) calculated by pixy. The 1:1 line is shown in red</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="nucleotide-diversity-landscapes" class="slide level2 ultitle">
<h2>Nucleotide diversity landscapes</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> vcf/allsites.vcf.gz <span class="at">--window-pi</span> 1000 <span class="at">--out</span> results/out</span>
<span id="cb15-2"><a></a><span class="ex">csvtk</span> plot line <span class="at">--tabs</span> results/out.windowed.pi <span class="at">-x</span> BIN_START <span class="at">-y</span> PI <span class="dt">\</span></span>
<span id="cb15-3"><a></a>   <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> <span class="st">"Position (bp)"</span> <span class="dt">\</span></span>
<span id="cb15-4"><a></a>   <span class="at">--ylab</span> <span class="st">"Diversity"</span> <span class="at">--title</span> LG4 <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="dt">\</span></span>
<span id="cb15-5"><a></a>   <span class="op">&gt;</span> results/out.windowed.pi.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div>
<p><a href="results/out.windowed.pi.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="results/out.windowed.pi.png" style="width:80.0%"></a></p>
</div>
<aside class="notes">
<p>Window-based plot is seque into genome scans</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="on-genome-scans" class="title-slide slide level1 center">
<h1>On genome scans</h1>

</section>
<section id="genetic-basis-of-adaptation-and-genome-scans" class="slide level2 ultitle">
<h2>Genetic basis of adaptation and genome scans</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="assets/images/storz-2005-fig1.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="assets/images/storz-2005-fig1.webp" class="quarto-figure quarto-figure-center" style="width:60.0%"></a></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p>Fundamental questions:</p>
<ul>
<li>How many genes are involved in the evolution of adaptive traits?</li>
<li>What is the distribution of phenotypic effects among successive allelic substitutions?</li>
<li>Is adaptation typically based on standing variation or new mutations?</li>
<li>What is the relative importance of additive vs.&nbsp;nonadditive effects on adaptive trait variation?</li>
<li>And what is the relative importance of structural vs.&nbsp;regulatory changes in phenotypic evolution?</li>
</ul>
<div class="flushright smallest">
<p><span class="citation" data-cites="storz_INVITEDREVIEWUsing_2005">Storz (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span>, Fig 1</p>
</div>
</div></div>
<aside class="notes">
<p><span class="citation" data-cites="storz_INVITEDREVIEWUsing_2005">Storz (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span> lays out genomic scans as an alternative approach to GWAS and QTL mapping for finding genetic determinants of adaptation. The reason is GWAS/QTL requires a phenotype whereas genome scans are data-driven.</p>
<p>Caption (fig 1):</p>
<blockquote>
<p>Fig. 1 (a) Effects of genetic hitch-hiking along a recombining chromosome. Hori- zontal lines depict a population sample of homologous chromosomes, and filled symbols depict neutral mutations. In this example an advantageous mutation (open symbol) arises and is rapidly driven to fixation by positive selection. Although the mutation is recombined against new genetic backgrounds during the course of the selective sweep, a sizable fraction of the ancestral haplotype (shown in red) also becomes fixed. Consequently, neutral variants that were initially linked to the advantageous mutation undergo a dramatic increase in frequency as a result of hitchhiking. (b) In this example, locus 3 has been rendered monomorphic by a selective sweep. Sampled gene copies (denoted by tips of the gene tree) share a very recent common ancestor, and π = 0 (where π = nucleotide diversity; Nei &amp; Li 1979). comparison, unlinked, neutrally evolving regions of the genome (loci 1, 2, and 4) are characterized by deeper genealogies, and higher levels of nucleotide diversity (π = 0.019–0.020). Note that the gene trees depict the true genealogies of the samples, not the genealogies inferred from observed variation.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example" class="slide level2 smaller ultitle">
<h2>Example</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> vcf/allsites.vcf.gz <span class="at">--weir-fst-pop</span> PUN-Y.txt <span class="dt">\</span></span>
<span id="cb16-2"><a></a>   <span class="at">--weir-fst-pop</span> PUN-R.txt <span class="dt">\</span></span>
<span id="cb16-3"><a></a>   <span class="at">--fst-window-size</span> 1000 <span class="dt">\</span></span>
<span id="cb16-4"><a></a>   <span class="at">--out</span> results/out</span>
<span id="cb16-5"><a></a><span class="ex">csvtk</span> plot line <span class="at">--tabs</span> results/out.windowed.weir.fst <span class="dt">\</span></span>
<span id="cb16-6"><a></a>   <span class="at">-x</span> BIN_START <span class="at">-y</span> MEAN_FST <span class="dt">\</span></span>
<span id="cb16-7"><a></a>   <span class="at">--point-size</span> 2 <span class="at">--xlab</span> <span class="st">"Position (bp)"</span> <span class="dt">\</span></span>
<span id="cb16-8"><a></a>   <span class="at">--ylab</span> <span class="st">"Fst"</span> <span class="at">--title</span> <span class="st">"LG4: PUN-Y vs PUN-R"</span> <span class="dt">\</span></span>
<span id="cb16-9"><a></a>   <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="at">--scatter</span> <span class="dt">\</span></span>
<span id="cb16-10"><a></a>   <span class="op">&gt;</span> results/out.windowed.weir.fst.mean.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="results/out.windowed.weir.fst.mean.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img data-src="results/out.windowed.weir.fst.mean.png" class="quarto-figure quarto-figure-center" style="width:70.0%"></a></p>
</figure>
</div>
</div>
</section>
<section id="z-scores-can-help-identifying-outliers" class="slide level2 smaller ultitle">
<h2>Z-scores can help identifying outliers</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"results/out.windowed.weir.fst"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-2"><a></a>x <span class="ot">&lt;-</span> data<span class="sc">$</span>MEAN_FST</span>
<span id="cb17-3"><a></a>z <span class="ot">&lt;-</span> (x <span class="sc">-</span> <span class="fu">mean</span>(x))<span class="sc">/</span><span class="fu">sd</span>(x)</span>
<span id="cb17-4"><a></a><span class="fu">plot</span>(<span class="at">x =</span> data<span class="sc">$</span>BIN_START, <span class="at">y =</span> z, <span class="at">xlab =</span> <span class="st">"Position (bp)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure>
<p><a href="diversity_files/figure-revealjs/r-zscore-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img data-src="diversity_files/figure-revealjs/r-zscore-1.svg" style="width:100.0%"></a></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<p>Raw data can be converted to <strong>Z-scores</strong> to highlight outliers. A Z-score is a measure of how far a data point is from the mean in terms of the number of standard deviations:</p>
<p><span class="math display">
Z = \frac{X - \mu}{\sigma}
</span></p>
<p>Threshold of a couple of standard deviations common.</p>
</div></div>
</section>
<section id="ld-decay-and-choice-of-window-size" class="slide level2 ultitle">
<h2>LD decay and choice of window size</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img data-src="https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:50%;">
<p>Properties of genetic variation and inferred demographic history in sampled <em>A. millepora</em>. <span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span>, Figure 2. Upper left plot illustrates LD as a function of physical distance. Here, choosing a window size 20-30kb would ensure that most windows are independent.</p>
</div></div>
</section>
<section id="annotations" class="slide level2 smaller ultitle">
<h2>Annotations</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a></a><span class="ex">csvtk</span> filter2 <span class="at">--tabs</span> gff/annotation.gff <span class="at">--filter</span> <span class="st">' $3 == "CDS" '</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb18-2"><a></a> <span class="ex">csvtk</span> mutate2 <span class="at">--tabs</span> <span class="at">-H</span> <span class="at">-e</span> <span class="st">'$4 - 12000000'</span> <span class="at">-w</span> 0 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb18-3"><a></a> <span class="ex">csvtk</span> mutate2 <span class="at">--tabs</span> <span class="at">-H</span> <span class="at">-e</span> <span class="st">'$5 - 12000000'</span> <span class="at">-w</span> 0 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb18-4"><a></a> <span class="ex">csvtk</span> cut <span class="at">--tabs</span> <span class="at">--fields</span> 1,10,11 <span class="kw">|</span> <span class="ex">bedtools</span> sort <span class="kw">|</span> <span class="ex">bedtools</span> merge <span class="dt">\</span></span>
<span id="cb18-5"><a></a>    <span class="op">&gt;</span> bed/CDS.bed <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb18-6"><a></a><span class="fu">head</span> <span class="at">-n</span> 3 bed/CDS.bed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LG4 12032   12121
LG4 12214   12658
LG4 12774   12830</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a></a><span class="ex">pixy</span> <span class="at">--vcf</span> vcf/allsites.vcf.gz <span class="at">--stats</span> pi <span class="dt">\</span></span>
<span id="cb20-2"><a></a> <span class="at">--populations</span> populations.txt <span class="at">--output_prefix</span> cds <span class="dt">\</span></span>
<span id="cb20-3"><a></a> <span class="at">--output_folder</span> results <span class="at">--bed_file</span> bed/CDS.bed</span>
<span id="cb20-4"><a></a><span class="ex">pixy</span> <span class="at">--vcf</span> vcf/allsites.vcf.gz <span class="at">--stats</span> pi <span class="dt">\</span></span>
<span id="cb20-5"><a></a> <span class="at">--populations</span> populations.txt <span class="at">--window_size</span> 1000 <span class="dt">\</span></span>
<span id="cb20-6"><a></a> <span class="at">--output_prefix</span> all <span class="at">--output_folder</span> results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a></a><span class="ex">csvtk</span> summary <span class="at">--tabs</span> <span class="at">-i</span> <span class="at">-g</span> pop <span class="dt">\</span></span>
<span id="cb21-2"><a></a>   <span class="at">-f</span> avg_pi:mean results/cds_pi.txt <span class="at">-w</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pop avg_pi:mean
PUN-R   0.00247
PUN-Y   0.00462</code></pre>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a></a><span class="ex">csvtk</span> summary <span class="at">--tabs</span> <span class="at">-i</span> <span class="at">-g</span> pop <span class="dt">\</span></span>
<span id="cb23-2"><a></a>   <span class="at">-f</span> avg_pi:mean results/all_pi.txt <span class="at">-w</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pop avg_pi:mean
PUN-R   0.00893
PUN-Y   0.00805</code></pre>
</div>
</div>
</div></div>
</section></section>
<section>
<section id="genome-diversity-and-differentiation-landscapes" class="title-slide slide level1 center">
<h1>Genome diversity and differentiation landscapes</h1>

</section>
<section id="dissecting-differentiation-landscapes" class="slide level2 ultitle">
<h2>Dissecting differentiation landscapes</h2>
<div>
<div id="fig-burri-2017" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-burri-2017-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/burri-2017-fig1.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="Figure&nbsp;3: @burri_DissectingDifferentiationLandscapes_2017 Fig. 1"><img data-src="assets/images/burri-2017-fig1.webp"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-burri-2017-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: <span class="citation" data-cites="burri_DissectingDifferentiationLandscapes_2017">Burri (<a href="#/bibliography" role="doc-biblioref" onclick="">2017</a>)</span> Fig. 1
</figcaption>
</figure>
</div>
</div>
<aside class="notes">

<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="monkeyflower-genomic-landscape" class="slide level2 ultitle">
<h2>Monkeyflower genomic landscape</h2>
<div>
<!-- markdownlint-disable MD013 -->
<div id="fig-stankowski-2019" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-stankowski-2019-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/stankowski-2019-fig1.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;4: @stankowski_WidespreadSelectionGene_2019 Fig. 1"><img data-src="assets/images/stankowski-2019-fig1.webp" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stankowski-2019-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: <span class="citation" data-cites="stankowski_WidespreadSelectionGene_2019">Stankowski et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span> Fig. 1
</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<aside class="notes">

<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="exercise" class="slide level2 ultitle">
<h2>Exercise</h2>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller ultitle scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-berner_AlleleFrequencyDifference_2019" class="csl-entry" role="listitem">
Berner, D. (2019). Allele <span>Frequency Difference AFD</span> to <span>FST</span> for <span>Quantifying Genetic Population Differentiation</span>. <em>Genes</em>, <em>10</em>(4), 308. <a href="https://doi.org/10.3390/genes10040308">https://doi.org/10.3390/genes10040308</a>
</div>
<div id="ref-burri_DissectingDifferentiationLandscapes_2017" class="csl-entry" role="listitem">
Burri, R. (2017). Dissecting differentiation landscapes: A linked selection’s perspective. <em>Journal of Evolutionary Biology</em>, <em>30</em>(8), 1501–1505. <a href="https://doi.org/10.1111/jeb.13108">https://doi.org/10.1111/jeb.13108</a>
</div>
<div id="ref-charlesworth_HowCanWe_2022" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Jensen, J. D. (2022). How <span>Can We Resolve Lewontin</span>’s <span>Paradox</span>? <em>Genome Biology and Evolution</em>, <em>14</em>(7), evac096. <a href="https://doi.org/10.1093/gbe/evac096">https://doi.org/10.1093/gbe/evac096</a>
</div>
<div id="ref-corbett-detig_NaturalSelectionConstrains_2015" class="csl-entry" role="listitem">
Corbett-Detig, R. B., Hartl, D. L., &amp; Sackton, T. B. (2015). Natural <span>Selection Constrains Neutral Diversity</span> across <span>A Wide Range</span> of <span>Species</span>. <em>PLOS Biology</em>, <em>13</em>(4), e1002112. <a href="https://doi.org/10.1371/journal.pbio.1002112">https://doi.org/10.1371/journal.pbio.1002112</a>
</div>
<div id="ref-ellegren_DeterminantsGeneticDiversity_2016" class="csl-entry" role="listitem">
Ellegren, H., &amp; Galtier, N. (2016). Determinants of genetic diversity. <em>Nature Reviews Genetics</em>, <em>17</em>(7), 422–433. <a href="https://doi.org/10.1038/nrg.2016.58">https://doi.org/10.1038/nrg.2016.58</a>
</div>
<div id="ref-fuller_PopulationGeneticsCoral_2020" class="csl-entry" role="listitem">
Fuller, Z. L., Mocellin, V. J. L., Morris, L. A., Cantin, N., Shepherd, J., Sarre, L., Peng, J., Liao, Y., Pickrell, J., Andolfatto, P., Matz, M., Bay, L. K., &amp; Przeworski, M. (2020). Population genetics of the coral <span>Acropora</span> millepora: <span>Toward</span> genomic prediction of bleaching. <em>Science</em>, <em>369</em>(6501), eaba4674. <a href="https://doi.org/10.1126/science.aba4674">https://doi.org/10.1126/science.aba4674</a>
</div>
<div id="ref-garcia-dorado_NeutralGeneticDiversity_2021" class="csl-entry" role="listitem">
García-Dorado, A., &amp; Caballero, A. (2021). Neutral genetic diversity as a useful tool for conservation biology. <em>Conservation Genetics</em>, <em>22</em>(4), 541–545. <a href="https://doi.org/10.1007/s10592-021-01384-9">https://doi.org/10.1007/s10592-021-01384-9</a>
</div>
<div id="ref-korunes_PixyUnbiasedEstimation_2021" class="csl-entry" role="listitem">
Korunes, K. L., &amp; Samuk, K. (2021). Pixy: <span>Unbiased</span> estimation of nucleotide diversity and divergence in the presence of missing data. <em>Molecular Ecology Resources</em>, <em>21</em>(4), 1359–1368. <a href="https://doi.org/10.1111/1755-0998.13326">https://doi.org/10.1111/1755-0998.13326</a>
</div>
<div id="ref-kuderna_GlobalCatalogWholegenome_2023" class="csl-entry" role="listitem">
Kuderna, L. F. K., Gao, H., Janiak, M. C., Kuhlwilm, M., Orkin, J. D., Bataillon, T., Manu, S., Valenzuela, A., Bergman, J., Rousselle, M., Silva, F. E., Agueda, L., Blanc, J., Gut, M., de Vries, D., Goodhead, I., Harris, R. A., Raveendran, M., Jensen, A., … Marques Bonet, T. (2023). A global catalog of whole-genome diversity from 233 primate species. <em>Science</em>, <em>380</em>(6648), 906–913. <a href="https://doi.org/10.1126/science.abn7829">https://doi.org/10.1126/science.abn7829</a>
</div>
<div id="ref-leffler_RevisitingOldRiddle_2012" class="csl-entry" role="listitem">
Leffler, E. M., Bullaughey, K., Matute, D. R., Meyer, W. K., Ségurel, L., Venkat, A., Andolfatto, P., &amp; Przeworski, M. (2012). Revisiting an <span>Old Riddle</span>: <span>What Determines Genetic Diversity Levels</span> within <span>Species</span>? <em>PLOS Biology</em>, <em>10</em>(9), e1001388. <a href="https://doi.org/10.1371/journal.pbio.1001388">https://doi.org/10.1371/journal.pbio.1001388</a>
</div>
<div id="ref-lewis_BiggestEverStudy_2023" class="csl-entry" role="listitem">
Lewis, D. (2023). Biggest ever study of primate genomes has surprises for humanity. <em>Nature</em>. <a href="https://doi.org/10.1038/d41586-023-01776-6">https://doi.org/10.1038/d41586-023-01776-6</a>
</div>
<div id="ref-nei_AnalysisGeneDiversity_1973" class="csl-entry" role="listitem">
Nei, M. (1973). Analysis of <span>Gene Diversity</span> in <span>Subdivided Populations</span>. <em>Proceedings of the National Academy of Sciences</em>, <em>70</em>(12), 3321–3323. <a href="https://doi.org/10.1073/pnas.70.12.3321">https://doi.org/10.1073/pnas.70.12.3321</a>
</div>
<div id="ref-stankowski_WidespreadSelectionGene_2019" class="csl-entry" role="listitem">
Stankowski, S., Chase, M. A., Fuiten, A. M., Rodrigues, M. F., Ralph, P. L., &amp; Streisfeld, M. A. (2019). Widespread selection and gene flow shape the genomic landscape during a radiation of monkeyflowers. <em>PLOS Biology</em>, <em>17</em>(7), e3000391. <a href="https://doi.org/10.1371/journal.pbio.3000391">https://doi.org/10.1371/journal.pbio.3000391</a>
</div>
<div id="ref-storz_INVITEDREVIEWUsing_2005" class="csl-entry" role="listitem">
Storz, J. F. (2005). <span>INVITED REVIEW</span>: <span>Using</span> genome scans of <span>DNA</span> polymorphism to infer adaptive population divergence. <em>Molecular Ecology</em>, <em>14</em>(3), 671–688. <a href="https://doi.org/10.1111/j.1365-294X.2005.02437.x">https://doi.org/10.1111/j.1365-294X.2005.02437.x</a>
</div>
<div id="ref-wright_EvolutionMendelianPopulations_1931" class="csl-entry" role="listitem">
Wright, S. (1931). Evolution in <span>Mendelian Populations</span>. <em>Genetics</em>, <em>16</em>(2), 97–159. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1201091/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1201091/</a>
</div>
</div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Genetic diversity</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>