---
title: Genetic diversity
subtitle: Theory and practice
author:
  - Per Unneberg
format:
  nbis-quarto-revealjs:
    footer: Genetic diversity
    toc: false
---

## Setup  {visibility="hidden" .hidden .unnumbered .unlisted}

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}

```{bash }
#| label: make-results-directory
#| echo: false
#| eval: true
mkdir -p results bed
```

# Genetic diversity

## What determines diversity levels?

:::: {.columns}

::: {.column width="30%"}

![](assets/images/leffler-2012-fig1.webp){width=55% fig-align=center}

:::

::: {.column width="70%"}

The usual questions:

> What evolutionary forces maintain genetic diversity in natural
> populations? How do diversity levels relate to census population
> sizes...? Do low levels of diversity limit adaptation to
> selective pressures?

::: {.flushright .smallest}

@leffler_RevisitingOldRiddle_2012

:::

<br>

::: {.fragment}

After *allozyme* era, the study of genetic diversity was largely
neglected due to lack of genome-wide data, but with advent of
population genomics becoming a hot topic again.

::: {.flushright .smallest}

@ellegren_DeterminantsGeneticDiversity_2016

:::

<br>

:::

::: {.fragment}

Lewontin's paradox: genetic diversity range smaller than variation
among species in population size

:::

:::

::::

::: {.notes }

Riddle: why is genetic diversity range so narrow? Possibly there are
lower and upper limits.

Lower limit: censoring effect, i.e., when diversity passes a lower
limit, population is driven to extinction due to inability to adapt

Upper limit: functional/structural constraints, e.g., impaired
chromosome pairing or reproductive incompatibilities

Linked selection, that is, variation-reducing selection. Larger
populations have higher influx of new mutations, so if more draft,
higher reduction in diversity. Would require strong frequent levels of
selection; little support for this in literature.

Purifying selection. Nearly neutral model could explain narrow range.

[@ellegren_DeterminantsGeneticDiversity_2016] Upper limit could in
part be explained by linked selection
[@corbett-detig_NaturalSelectionConstrains_2015]

:::

## Factors that influence genetic diversity

:::: {.columns}

::: {.column width="50%"}

```{r }
#| label: fig-genetic-diversity-determinants-Ellegren-fig2
#| echo: false
#| eval: true
#| out-width: 90%
#| fig-cap:
#|    Overview of determinants of genetic diversity
#|    [@ellegren_DeterminantsGeneticDiversity_2016, Fig 2]
knitr::include_graphics("https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg.2016.58/MediaObjects/41576_2016_Article_BFnrg201658_Fig2_HTML.jpg?as=webp")
```

:::

::: {.column width="50%"}

#### Genetic drift

Reduces diversity at loss $\propto \frac{1}{N}$

#### Selection

Adaptive selection decreases variation, more so if acting on new
mutations compared to standing variation.

Balancing selection may increase variation.

#### Recombination

Low recombination rates lead to less "reshuffling" of variation and
hence lower diversity.

:::

::::

::: {.notes }

Note: lower recombination lowers diversity presumably because more
neutral variants linked to selected sites. There are intriguing
differences between taxa [@leffler_RevisitingOldRiddle_2012]

@ellegren_DeterminantsGeneticDiversity_2016 point out that with advent
of population genomic data, we are at the brink of understanding
Lewontin's paradox, namely [@charlesworth_HowCanWe_2022]

> the observation that the range of levels of genetic diversity in
> natural populations appears to be far smaller than the extent of
> variation among species in population size

:::

## Why we measure patterns of genetic variation

Genetic variation patterns are informative of evolutionary and
demographic processes. We often use **summary statistics** to describe
the patterns, and to estimate parameters such as **effective
population size** and **mutation rate** from genetic variation data
($\theta = 4N_e\mu$)

::: {.fragment}

:::{}

Often critical first step of analysis, such as

- exploratory study
- test of an evolutionary hypothesis
- training of machine-learning models

:::

::: {.flushright .smallest}

[@korunes_PixyUnbiasedEstimation_2021]

:::

:::

## Genetic diversity in conservation biology {.smaller}

:::: {.columns}

::: {.column width="50%"}

![@kuderna_GlobalCatalogWholegenome_2023, Fig
2](assets/images/kuderna-2023-fig2.webp){#fig-kuderna-2023 width=100%
fig-align=center}

> ...no global relationship between numerically coded IUCN extinction
> risk categories and estimated heterozygosity...

:::

::: {.column width="50%"}

:::{}

Low genetic diversity symptom of past genetic drift inbreeding (higher
levels of homozygosity), caused by low $N_e$

::: {.flushright .smallest}

@garcia-dorado_NeutralGeneticDiversity_2021

:::

:::

<br><br>

:::{}

However: if population decline is rapid, may be too little time for
inbreeding to occur $\Rightarrow$ genetic diversity within species not
necessarily aligned to extinction risk

::: {.flushright .smallest}

@lewis_BiggestEverStudy_2023

:::

:::

:::

::::

::: {.notes}

Take home [@lewis_BiggestEverStudy_2023]: genetic diversity within a
species does not necessarily align with its extinction risk. Possible
explanation: population decline has been so rapid that there hasn't
been time for inbreeding to occur. Other factors (e.g., habitat
destruction) greater threat.

Caption:

> Runs of homozygosity and impact of extinction risk on diversity (A)
> Relationship between IUCN extinction risk categories and
> heterozygosity. Solid black circles and bars denote median and IQR.
> (B) Partition into threatened (T: VU, EN, CR) and nonthreatened (N:
> LC, NT) categories for all families with more than one species in
> either partition. Significant differences (p < 0.05, one-sided
> rank-sum test) are marked with an asterisk. (C) Median number of
> tracts of homozygosity versus median proportion of the genome in
> runs of homozygosity per species. Species with a fraction over 1/3
> are highlighted. Solid black dots within highlights denote
> threatened species (VU, EN, CR).

On Figure 2A:

> We investigated whether genetic diversity estimates are correlated
> with extinction risk in primates, a subject of previous debate (17,
> 33, 34). Despite our broad sampling, we find no global relationship
> between numerically coded IUCN extinction risk categories and
> estimated heterozygosity [p > 0.05, phylogenetic generalized least
> squares (PGLS)] (Fig. 2A) (16).

:::

# Measuring genetic diversity

## Nucleotide diversity

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-ts-small-pi
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
print("\n".join([f"{k}:   {v}" for k, v in zip(x.index, x.values)]))
```

::: {.fragment fragment-index=2}

Diversity: for each site, count and sum differences between all
(unique) pairs of samples, and divide by unique pairs. For $n$
samples, there are $n \choose 2$ such pairs.

:::

::: {.fragment fragment-index=3}

Example: for site 0, start comparing samples 0-1 (0 diffs), samples
0-2 (0), samples 0-3 (1), samples 1-2 (0) and so on. Call these
differences $k_{ij}$. Then

:::

::: {.fragment fragment-index=4}

$$
\pi = \frac{\sum_{i<j}k_{ij}}{n \choose 2}
$$

:::

:::

::: {.column width="50%"}

::: {.fragment fragment-index=1}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-diversity
#| echo: false
#| eval: true
#| fig-format: svg
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    id_string="msprimetreediversity",
    show_sequences=True,
    x_axis=False))
```

:::

:::

::::

## Watterson's $\theta_W$

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-ts-small-theta
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
print("\n".join([f"{k}:   {v}" for k, v in zip(x.index, x.values)]))
```

::: {.fragment fragment-index=1}

Alternative measure of diversity: simply count the number of
segregating sites ($S$). However, must correct for the number of
samples $n$ as we expect that more samples $\Rightarrow$ more sites

:::

::: {.fragment fragment-index=2}

$$
\theta_W = \frac{S}{a} = \frac{S}{\sum_{i=1}^{n-1}\frac{1}{i}}
$$

:::

::: {.fragment fragment-index=3}

Important: under neutrality, $\theta = E(\pi) = E(\theta_W)$. Difference
between two the basis for **Tajima's D** that is a test for selection

:::

:::

::: {.column width="50%"}

::: {}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-thetaw
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    id_string="msprimetreethetaw",
    show_sequences=True,
    x_axis=False))
```

:::

:::

::::

::: {.notes}

Intuitive example to understand of Tajimas: balancing selection has
long internal branches -> long haplotype. This will elevate pi since
many samples will be different for many sites.

:::

## Calculating diversity measures - $\pi$ and $\theta_W$

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-genotypematrix-and-results
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
print("\n".join([f"{k}:   {v}" for k, v in zip(x.index, x.values)]))
```

::: {.small}

```{python }
#| label: msprime-tree-1-diversity
#| echo: true
#| eval: true
from trees import ts_small_mut as ts

# Calculate correction factor a for Watterson's
# theta: the larger the sample size, the more
# segregating sites we expect to see
a = sum([1/i for i in range(1, ts.num_samples)])

pi = ts.diversity()
thetaW = ts.num_sites / a / ts.sequence_length

print(f"Diversity:           {pi:.6f}",
      f"Watterson's theta:   {thetaW:.6f}",
      f"Sequence length:     {ts.sequence_length:.0f}",
      sep="\n")
```

:::

:::

::: {.column width="50%"}

::: { .smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

::: {}

```{python }
#| label: fig-msprime-calculate-diversity-theta
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    id_string="msprimecalctreediversity",
    show_sequences=True,
    x_axis=True))
```

:::

:::

::::

## Divergence - dXY

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-divergence-1
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
for i, v in enumerate(x):
    print(f"Sample {i}:   {v}")
    if i == 1:
        print()
print("\n")
```

Divergence: for each site, count and sum differences between all pairs
of samples *between* two populations

::: {.fragment}

Example: for site 0, compare samples 0-2 (0 diffs), samples 0-3 (1
diff), samples 1-2 (0 diffs), samples 1-3 (1 diff), and so on. Call
differences $k_ij$, let $n_X$, $n_Y$ be sample size in populations
$X$, $Y$. Then

:::

::: {.fragment}

$$
d_{XY} = \frac{1}{n_Xn_Y}\sum_{i=1}^{n_X}\sum_{i=1}^{n_Y}k_{ij}
$$

:::

:::

::: {.column width="50%"}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-small-divergence-2
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
css_extra = ".n4 .edge {stroke: blue }"
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    css_extra=css_extra,
    id_string="msprimetreedivergence",
    show_sequences=True,
    x_axis=False))
```

:::

::::

## Divergence - dXY

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-divergence-2
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
for i, v in enumerate(x):
    print(f"Sample {i}:   {v}")
    if i == 1:
        print()
print("\n")
```

```{python }
#| label: msprime-tree-divergence-2
#| echo: true
#| eval: true
from trees import ts_small_mut as ts

sample_sets = [[0, 1], [2, 3]]

dxy = ts.divergence(sample_sets=sample_sets)

print(f"Divergence:   {dxy:.6f}")
```

:::

::: {.column width="50%"}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-calc-divergence
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
css_extra = ".n4 .edge {stroke: blue }"
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    css_extra=css_extra,
    id_string="msprimetreedivergence",
    show_sequences=True,
    x_axis=False))
```

:::

::::

::: {.notes}

$F_ST$ (later) and related statistics are strongly affected by
within-subpopulation genetic variance. @nei_AnalysisGeneDiversity_1973
proposed dXY as an alternative that does not depend on the levels of
diversity *within* subpopulations.

:::

## Differentiation - AFD (allele frequency difference) {.smaller}

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-afd
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
for i, v in enumerate(x):
    print(f"Sample {i}:   {v}")
    if i == 1:
        print()
print("\n")
```

Allele Frequency Difference (AFD) proposed as intuitive alternative to
{{< var fst >}}. For each site, count the difference in allele
frequency between two populations.

::: {.fragment}

Example: site 0, frequency in blue is 0, in black 1/2, so
difference=1/2, site 1, frequency in blue 0, in black 1/2, and so on

:::

::: {.fragment}

$$
AFD = \frac{1}{2}\sum_{i=1}^n| (f_{i1} - f_{i2})|
$$

where $n$ is the number of different alleles ($n=2$ for biallelic
SNPs), $f_{ij}$ is the frequency of allele $i$ in population $j$

::: {.flushright .smallest}

@berner_AlleleFrequencyDifference_2019

:::

:::

:::

::: {.column width="50%"}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-calc-afd
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
css_extra = ".n4 .edge {stroke: blue }"
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    css_extra=css_extra,
    id_string="msprimetreeafd",
    show_sequences=True,
    x_axis=False))
```

:::

::::

::: {.notes}

AFD proposed as a more intuitive and easier to understand statistic than Fst

:::

## Differentiation - {{< var fst >}} {.smaller}

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-fst-1
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
for i, v in enumerate(x):
    print(f"Sample {i}:   {v}")
    if i == 1:
        print()
print("\n")
```

{{< var fst >}} is another measure of differention among
subpopulations [@wright_EvolutionMendelianPopulations_1931]. It ranges
between 0 and 1 and has the interpretation 0: no differentiation, 1:
complete fixation of alternate alleles in subpopulations

::: {.fragment}

Example: site 3 has {{< var fst >}}=1 as it is fixed in black, not
present in blue

:::

::: {.fragment}

There are *many* ways to express and calculate {{< var fst >}}.
Example:

$$
F_{\mathrm{ST}} = \frac{h_{\mathrm{T}} - h_{\mathrm{S}}}{h_{\mathrm{T}}}
$$

where $h_{\mathrm{T}}$ is the expected heterozygosity in the total
population, $h_{\mathrm{S}}$ the *average* of expected
heterozygosities across subpopulations. For site 3,
$h_{\mathrm{S}}=0$, $h_{\mathrm{T}}=2/3$.

:::

::: {.fragment}

Caveats: strongly influenced by *within* subpopulation levels of
variation. Therefore considered *relative* measure (cf {{< var dxy
>}}, which is an *absolute* measure).

:::

:::

::: {.column width="50%"}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-calc-fst-1
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
css_extra = ".n4 .edge {stroke: blue }"
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    css_extra=css_extra,
    id_string="msprimetreefst",
    show_sequences=True,
    x_axis=False))
```

:::

::::

::: {.notes}

Wright's fst measures differentiation among subpopulations.
Interpretation: 0 no differentiation, 1 complete fixation of alternate
alleles. Has therefore become a popular statistic for quantifying
differentiation from molecular data.

:::

## Differentiation - {{< var fst >}} {.smaller}

:::: {.columns}

::: {.column width="50%"}

```{python }
#| label: msprime-tree-small-fst-2
#| echo: false
#| eval: true
from trees import ts_small_mut, genotypes
x = genotypes(ts_small_mut)
for i, v in enumerate(x):
    print(f"Sample {i}:   {v}")
    if i == 1:
        print()
print("\n")
```

:::{}

```{python }
#| label: msprime-tree-calculate-fst
#| echo: true
#| eval: true
from trees import ts_small_mut as ts

sample_sets = [[0, 1], [2, 3]]
win = [0] + [int(x.position)+1 for x in ts.sites()]
_ = win.pop()
win = win + [ts.sequence_length]

fst = ts.Fst(sample_sets=sample_sets)
fst_sites = ts.Fst(sample_sets=sample_sets,
                   windows=win)

print(f"Site id:       {list(range(5))}",
      f"Fst per site:  {fst_sites}",
      f"Overall Fst:   {fst:.6f}",
      sep="\n")
```

:::

:::

::: {.column width="50%"}

::: {.smallest}

Tree plot with mutations and sequences [{{< fa rectangle-list >}}]({{<
var recipes.slides >}}#sec-recipe-diversity-msprime-trees)

:::

```{python }
#| label: fig-msprime-tree-calc-fst-2
#| echo: false
#| eval: true
#| fig-cap: |
#|     [{{< fa rectangle-list >}}]({{< var recipes.slides
#|     >}}#sec-recipe-diversity-msprime-trees)
#| fig-format: svg
#| output: asis
from trees import ts_small_mut as ts
from trees import plot_tree
css_extra = ".n4 .edge {stroke: blue }"
print(plot_tree(
    ts,
    size=(500,400),
    canvas_size=(1200, 500),
    css_extra=css_extra,
    id_string="msprimetreefst",
    show_sequences=True,
    x_axis=True))
```

:::

::::

::: {.notes}

:::

## Many programs treat missing data as invariant

:::: {.columns}

::: {.column width="50%"}

<!-- markdownlint-disable MD013 -->

![@korunes_PixyUnbiasedEstimation_2021, Fig 1](https://onlinelibrary.wiley.com/cms/asset/2d12ae05-e2cd-496e-a1d6-f3842bb59e22/men13326-fig-0001-m.jpg){width=80% fig-align=center}

<!-- markdownlint-enable MD013 -->

:::

::: {.column width="50%"}

Diversity:

$$
\pi = \frac{\sum_{i<j}k_{ij}}{n \choose 2}
$$

Divergence:

$$
d_{XY} = \frac{1}{n_Xn_Y}\sum_{i=1}^{n_X}\sum_{j=1}^{n_Y}k_{ij}
$$

Here, $n$ is the number of samples, $k_{ij}$ tally of allelic
differences between two haplotypes within ($\pi$) a population or
between ($d_{XY}$) populations

:::

::::

::: {.notes}

Paper caption:

> The logic and input/ouput of pixy demonstrated with a simple haploid
> example. (a) Comparison of two methods for computing π (or dXY) in
> the face of missing data. These methods follow the first expression
> of Equation 1 but differ in how they calculate the numerator and
> denominator. In Case 1, all missing data is assumed to be present but
> invariant. This results in a deflated estimate of π. In Case 2,
> missing data are simply omitted from the calculation, both in terms
> of the number of sites (the final denominator) and the component
> denominators for each site (the n choose two terms). This results in
> an unbiased estimate of π. (b) The adjusted π method (Case 2) as
> implemented for VCFs in pixy. The example VCF (input) contains the
> same four haplotypes as (a). Invariant sites are represented as sites
> with no ALT allele, and greyed-out sites are those that failed to
> pass a genotype filter requiring a minimum number of reads covering
> the genotype (Depth ≥ 10 in this case)

:::

## Missing data may bias diversity measures downwards

:::{}

<!-- markdownlint-disable MD013 -->

![@korunes_PixyUnbiasedEstimation_2021, Fig 4](https://onlinelibrary.wiley.com/cms/asset/dc499ed3-0c1f-4260-95a1-cb85a0894d10/men13326-fig-0004-m.jpg){width=45% fig-align=center}

<!-- markdownlint-enable MD013 -->

:::

::: {.notes}

Real data: strong tendency of commonly adopted tools to
**underestimate** pi, dxy (even for 30X data sets)

Caption:

> Comparisons of estimates of π from whole genome data derived from 18
> Anopheles gambiae individuals from the Ag1000G Burkina Faso (BFS)
> population. Each panel (a–d) depicts the estimates of π for the X
> chromosome performed using pixy (y-axis) and four other methods
> (x-axis, a–d). Points are coloured according to the proportion of
> missing data (of any type) calculated by pixy. The 1:1 line is shown
> in red

:::

## Nucleotide diversity landscapes

```{bash }
#| label: monkeyflower-diversity-landscape
#| echo: true
#| eval: true
#| results: hide
vcftools --gzvcf vcf/allsites.vcf.gz --window-pi 1000 --out results/out
csvtk plot line --tabs results/out.windowed.pi -x BIN_START -y PI \
   --point-size 0.01 --xlab "Position (bp)" \
   --ylab "Diversity" --title LG4 --width 9.0 --height 3.5 \
   > results/out.windowed.pi.png
```

:::{}

![](results/out.windowed.pi.png){width=80%}

:::

::: {.notes}

Window-based plot is seque into genome scans

:::

# On genome scans

## Genetic basis of adaptation and genome scans

:::: {.columns}

::: {.column width="50%"}

![](assets/images/storz-2005-fig1.webp){width=60% fig-align=center}

:::

::: {.column width="50%"}

Fundamental questions:

- How many genes are involved in the evolution of adaptive traits?
- What is the distribution of phenotypic effects among successive
  allelic substitutions?
- Is adaptation typically based on standing variation or new
  mutations?
- What is the relative importance of additive vs. nonadditive effects
  on adaptive trait variation?
- And what is the relative importance of structural vs. regulatory
  changes in phenotypic evolution?

::: {.flushright .smallest}

@storz_INVITEDREVIEWUsing_2005, Fig 1

:::

:::

::::

::: {.notes}

@storz_INVITEDREVIEWUsing_2005 lays out genomic scans as an
alternative approach to GWAS and QTL mapping for finding genetic
determinants of adaptation. The reason is GWAS/QTL requires a
phenotype whereas genome scans are data-driven.

Caption (fig 1):

> Fig. 1 (a) Effects of genetic hitch-hiking along a recombining
> chromosome. Hori- zontal lines depict a population sample of
> homologous chromosomes, and filled symbols depict neutral mutations.
> In this example an advantageous mutation (open symbol) arises and is
> rapidly driven to fixation by positive selection. Although the
> mutation is recombined against new genetic backgrounds during the
> course of the selective sweep, a sizable fraction of the ancestral
> haplotype (shown in red) also becomes fixed. Consequently, neutral
> variants that were initially linked to the advantageous mutation
> undergo a dramatic increase in frequency as a result of hitchhiking.
> (b) In this example, locus 3 has been rendered monomorphic by a
> selective sweep. Sampled gene copies (denoted by tips of the gene
> tree) share a very recent common ancestor, and π = 0 (where π =
> nucleotide diversity; Nei & Li 1979). comparison, unlinked,
> neutrally evolving regions of the genome (loci 1, 2, and 4) are
> characterized by deeper genealogies, and higher levels of nucleotide
> diversity (π = 0.019–0.020). Note that the gene trees depict the
> true genealogies of the samples, not the genealogies inferred from
> observed variation.

:::

## Example {.smaller}

```{bash }
#| label: genome-scan-example-fst
#| echo: true
#| eval: true
#| results: hide
vcftools --gzvcf vcf/allsites.vcf.gz --weir-fst-pop PUN-Y.txt \
   --weir-fst-pop PUN-R.txt \
   --fst-window-size 1000 \
   --out results/out
csvtk plot line --tabs results/out.windowed.weir.fst \
   -x BIN_START -y MEAN_FST \
   --point-size 2 --xlab "Position (bp)" \
   --ylab "Fst" --title "LG4: PUN-Y vs PUN-R" \
   --width 9.0 --height 3.5 --scatter \
   > results/out.windowed.weir.fst.mean.png
```

:::{}

![](results/out.windowed.weir.fst.mean.png){width=70% fig-align=center}

:::

## Z-scores can help identifying outliers {.smaller}

:::: {.columns}

::: {.column width="60%"}

```{r }
#| label: r-zscore
#| echo: true
#| eval: true
#| out-width: 100%
#| fig-width: 12
#| fig-height: 4
data <- read.table("results/out.windowed.weir.fst", header = TRUE)
x <- data$MEAN_FST
z <- (x - mean(x)) / sd(x)
plot(x = data$BIN_START, y = z, xlab = "Position (bp)")
```

:::

::: {.column width="40%"}

Raw data can be converted to **Z-scores** to highlight outliers. A
Z-score is a measure of how far a data point is from the mean in terms
of the number of standard deviations:

$$
Z = \frac{X - \mu}{\sigma}
$$

Threshold of a couple of standard deviations common.

:::

::::

## LD decay and choice of window size

:::: {.columns}

::: {.column width="50%"}

<!-- markdownlint-disable MD013 -->

![](https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg){width=80% fig-align=center}

<!-- markdownlint-enable MD013 -->

:::

::: {.column width="50%"}

Properties of genetic variation and inferred demographic history in
sampled *A. millepora*. @fuller_PopulationGeneticsCoral_2020,
Figure 2. Upper left plot illustrates LD as a function of physical
distance. Here, choosing a window size 20-30kb would ensure that most
windows are independent.

:::

::::

## Annotations {.smaller}

```{bash }
#| label: pixy-prepare-annotation
#| echo: true
#| eval: true
csvtk filter2 --tabs gff/annotation.gff --filter ' $3 == "CDS" ' |\
 csvtk mutate2 --tabs -H -e '$4 - 12000000' -w 0 |\
 csvtk mutate2 --tabs -H -e '$5 - 12000000' -w 0 |\
 csvtk cut --tabs --fields 1,10,11 | bedtools sort | bedtools merge \
    > bed/CDS.bed 2>/dev/null
head -n 3 bed/CDS.bed
```

```{bash }
#| label: pixy-calculate-by-annotation
#| echo: true
#| eval: true
#| results: hide
pixy --vcf vcf/allsites.vcf.gz --stats pi \
 --populations populations.txt --output_prefix cds \
 --output_folder results --bed_file bed/CDS.bed
pixy --vcf vcf/allsites.vcf.gz --stats pi \
 --populations populations.txt --window_size 1000 \
 --output_prefix all --output_folder results
```

:::: {.columns}

::: {.column width="50%"}

```{bash }
#| label: avg-pi-cds
#| echo: true
#| eval: true
csvtk summary --tabs -i -g pop \
   -f avg_pi:mean results/cds_pi.txt -w 5
```

:::

::: {.column width="50%"}

```{bash }
#| label: avg-pi-all
#| echo: true
#| eval: true
csvtk summary --tabs -i -g pop \
   -f avg_pi:mean results/all_pi.txt -w 5
```

:::

::::

# Genome diversity and differentiation landscapes

## Dissecting differentiation landscapes

:::{}

![@burri_DissectingDifferentiationLandscapes_2017 Fig. 1](assets/images/burri-2017-fig1.webp){#fig-burri-2017}

:::

::: {.notes}

:::

## Monkeyflower genomic landscape

:::{}

<!-- markdownlint-disable MD013 -->

![@stankowski_WidespreadSelectionGene_2019 Fig.
1](assets/images/stankowski-2019-fig1.webp){#fig-stankowski-2019
width=80% fig-align=center}

<!-- markdownlint-enable MD013 -->

:::

::: {.notes}

:::

## Exercise

## Bibliography {.unnumbered .unlisted .smaller}

::: { #refs }
:::
