<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Variant filtering</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-802d3b82e08d8dc8d935d89a8ebe9074.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Variant filtering</h1>
  <p class="subtitle">From raw variant calls to high-quality call sets</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="why-we-need-to-filter-variants" class="slide level2 ultitle">
<h2>Why we need to filter variants</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-Li-2014-2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Li-2014-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/Li-2014-F2.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;1: Overlap in raw variant calls for different combinations of read mappers and variant callers."><img data-src="assets/images/Li-2014-F2.webp" class="quarto-figure quarto-figure-center" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Li-2014-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Overlap in <strong>raw</strong> variant calls for different combinations of read mappers and variant callers.
</figcaption>
</figure>
</div>
<p>Error rate of variant calls (SNPs and INDELs) largely unknown. Two major sources of error are</p>
<ol type="1">
<li>erroneous realignment in low-complexity regions</li>
<li>incomplete reference sequence</li>
</ol>
</div><div class="column" style="width:50%;">
<div id="fig-Li-2014-3" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Li-2014-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/images/Li-2014-F3.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;2: Overlap in filtered variant calls for different combinations of read mappers and variant callers"><img data-src="assets/images/Li-2014-F3.webp" class="quarto-figure quarto-figure-center" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Li-2014-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Overlap in <strong>filtered</strong> variant calls for different combinations of read mappers and variant callers
</figcaption>
</figure>
</div>
</div><div class="flushright smallr translatey50">
<p><span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">Li (<a href="#/bibliography" role="doc-biblioref" onclick="">2014</a>)</span></p>
</div>
</div>
<aside class="notes">
<p>Error rate of variant calls (SNPs and INDELs) and leading causal artifacts remain unclear and largely unknown <span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">(<a href="#/bibliography" role="doc-biblioref" onclick="">Li, 2014</a>)</span>. This study compared several read mappers and variant callers on a haploid data set. For comparison, the following variant filters were applied:</p>
<ol type="1">
<li>low-complexity (LC) filter</li>
<li>maximum depth (MD)</li>
<li>allele balance (AD) - filter sites where non-reference (ALT) too low</li>
<li>double strand (DS) - filter if ALT too low on one strand</li>
<li>Fisher strand (FS) - filter if strong REF/ALT correlation with +/- strand</li>
<li>quality filter (QU) - variant quality threshold</li>
</ol>
<p>Here, AB, DS and FS filters are <em>caller dependent</em>!</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="manual-filtering-sets-thresholds-on-context-statistics" class="slide level2 smaller ultitle">
<h2>Manual filtering sets thresholds on context statistics</h2>
<!-- markdownlint-disable MD013 -->
<div class="small">
<div id="tbl-key-data-filters" class="quarto-float quarto-figure quarto-figure-center" data-tbl-colwidths="[30,30,40]">
<figure class="quarto-float quarto-float-tbl">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-key-data-filters-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: Key data filters <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(Table 3 <a href="#/bibliography" role="doc-biblioref" onclick="">Lou et al., 2021, p. 5974</a>)</span>
</figcaption>
<div aria-describedby="tbl-key-data-filters-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top">
<colgroup>
<col style="width: 30%">
<col style="width: 30%">
<col style="width: 40%">
</colgroup>
<thead>
<tr class="header">
<th>Category</th>
<th>Filter</th>
<th>Recommendation (examples)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>General filters</td>
<td>Base quality</td>
<td>Recalibrate / &lt;Q20</td>
</tr>
<tr class="even">
<td></td>
<td>Mapping quality</td>
<td>MAQ &lt; 20 / improper pairs</td>
</tr>
<tr class="odd">
<td></td>
<td>Minimum depth and/ or number of individuals</td>
<td>Varies; e.g.&nbsp;&lt;50% individuals, &lt;0.8X average depth</td>
</tr>
<tr class="even">
<td></td>
<td>Maximum depth</td>
<td>1-2 sd above median depth</td>
</tr>
<tr class="odd">
<td></td>
<td>Duplicate reads</td>
<td>Remove</td>
</tr>
<tr class="even">
<td></td>
<td>Indels</td>
<td>Realign reads / haplotype-based caller / exclude bases flanking indels</td>
</tr>
<tr class="odd">
<td></td>
<td>Overlapping sections of paired-­end reads</td>
<td>Soft-clip to avoid double-counting</td>
</tr>
<tr class="even">
<td>Filters on polymorphic sites</td>
<td><span class="math inline">p</span>-value</td>
<td><span class="math inline">10^{-6}</span></td>
</tr>
<tr class="odd">
<td></td>
<td>SNPs with more than two alleles</td>
<td>Filter; methods often assume bi-allelic sites</td>
</tr>
<tr class="even">
<td></td>
<td>Minimum minor allele frequency (MAF)</td>
<td>1%-10% for some analyses (PCA/admixture/LD/<span class="math inline">\mathsf{F_{ST}}</span>)</td>
</tr>
<tr class="odd">
<td>Restricting analysis to a predefined site list</td>
<td>List of global SNPs</td>
<td>Use global call set for analyses requiring shared sites</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<h4 id="procedure">Procedure</h4>
<p>Look at annotations (<em>context statistics</em>) and set thresholds.</p>
<p>Example: filter all sites with MAF&lt;1%</p>
<aside class="notes">
<p>NB: bypassing recommendations often means doing custom analyses. For instance, <span class="citation" data-cites="talla_DissectingEffectsSelection_2019">Talla et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span> include GATK tri-allelic sites due to different bi-allelic pairs segregating in different subpopulations (e.g.&nbsp;A/G in pop 1, A/T in pop 2)</p>
<h4 id="verbose-explanations-of-filters">Verbose explanations of filters</h4>
<p>Source: <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lou et al., 2021</a>)</span></p>
<blockquote>
<p>Base quality scores are factored into the calculation of genotype likelihoods, so if they accurately reflect the probability of sequencing error, bases with low scores also carry useful information. However, base quality scores are sometimes miscalibrated, so noise may be reduced if bases with scores below a threshold (e.g., 20) are either trimmed off prior to analysis or ignored. Alternatively, all base quality scores can be recalibrated based on estimated error profiles in the data (see Section 3.1).</p>
<p>Mapping quality is not considered in genotype likelihood estimation in currently available tools, so it is often advisable to remove low-­confidence and/or nonuniquely mapped reads prior to analysis (e.g., reads with mapping quality &lt;20). Filtering out reads that do not map in proper pairs should also further increase confidence in reads being mapped to the correct location, but could cause biases in regions with structural variation.</p>
<p>To avoid sites with low or confounding data support in downstream analysis, minimum depth and/or minimum number of individual filters can be used to exclude sites with much reduced sequencing coverage compared to the rest of the genome (e.g., regions with low unique mapping rates, such as repetitive sequences). Appropriate thresholds will vary between data sets, but could, for example, exclude sites with read data for &lt;50% of individuals (globally or within each population), or with &lt;0.8× average depth across individuals (after filtering on mapping quality)</p>
<p>Maximum depth filters are used to exclude sites with exceptionally high coverage (e.g., regions that are susceptible to dubious mapping, such as copy number variants). Common maximum depth thresholds could be one or two standard deviations above the median genome-­wide depth.</p>
<p>PCR and optical duplicates can give inflated impressions of how many unique molecules have been sequenced, which—­particularly in the presence of preferential amplification of one allele—­ could bias genotype likelihood estimation. We therefore recommend removing duplicate reads prior to any analysis.</p>
<p>Reads mapped across indels are frequently misaligned, especially if the ends of reads span an indel. To avoid false SNP calls, we recommend either using dedicated tools to realign reads covering indels, using a haplotype-­based variant caller (e.g., freebayes or gatk) to estimate genotype likelihoods, or excluding bases flanking indels.</p>
<p>If the DNA insert in a library fragment is shorter than the combined length of paired reads, there will be a section of overlap between the forward and reverse reads. While some variant callers (e.g., gatk) account for the pseudoreplication in overlapping ends of read pairs, the current implementation of angsd treats each end of a read pair as independent (this may change in a future release (T. Korneliussen, personal communication)). When treated as independent, read support for overlapping sections will be “double counted,” which may bias genotype likelihoods. A conservative approach is to soft-­clip one of the overlapping read ends.</p>
<p>The significance threshold (often in the form of maximum p-­value) can be adjusted to fine-­tune the sensitivity of polymorphism detection, with lower p-­values leading to fewer, but higher confidence, SNP calls. A commonly used cut-­off is 10 −6.</p>
<p>Most software programs for downstream analyses assume that all SNPs are biallelic, so SNPs with more than two alleles can be filtered out in the SNP identification step to avoid violation of such assumptions.</p>
<p>For many types of analysis, such as PCA, admixture analysis, detection of FST outliers and estimation of LD, low-­frequency SNPs are uninformative and can even bias results (e.g.&nbsp;Linck &amp; Battey, 2019; Roesti et al., 2012). For those types of analysis, imposing a minimum MAF filter of 1%–­10% can substantially speed up computation time. Appropriate thresholds depend on coverage, sample size (how many copies does an MAF threshold correspond to) and the type of downstream analysis.</p>
<p>For comparison of parameter estimates for multiple populations, it is important to ensure that data are obtained for a shared set of sites and that SNP polarization (which allele we track the frequency of) is consistent. For programs such as angsd where population-­specific estimates are obtained by analysing the data from each population separately, a good strategy is to first conduct a global SNP calling with all samples and then restrict population-­specific analysis to those SNPs with consistent major and minor allele designations (-­doMajorMinor 3 in angsd) no MAF or SNP p-­value filter (because that would incorrectly generate “missing data” if a site is fixed in a particular population).</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="guidelines-what-guidelines" class="slide level2 smaller ultitle">
<h2>Guidelines? What guidelines?</h2>
<h5 id="gatk-hard-filters">GATK hard filters</h5>
<blockquote>
<p>However, because we want to help, we have formulated some generic recommendations that should at least provide a starting point for people to experiment with their data.</p>
</blockquote>
<div class="fragment">
<div class="columns small">
<div class="column" style="width:50%;">
<h6 id="snps">SNPs</h6>
<pre><code>QualByDepth (QD) &lt; 2.0
RMSMappingQuality (MQ) &lt; 40.0
FisherStrand (FS) &gt; 60.0
StrandOddsRatio (SOR) &gt; 3.0
MappingQualityRankSumTest (MQRankSum) &lt; -12.5
ReadPosRankSumTest (ReadPosRankSum) &lt; -8.0</code></pre>
</div><div class="column" style="width:50%;">
<h6 id="indels">Indels</h6>
<pre><code>QualByDepth (QD) &lt; 2.0
ReadPosRankSum (ReadPosRankSumTest) &lt; -20.0
InbreedingCoeff &lt; -0.8
FisherStrand (FS) &gt; 200.0
StrandOddsRatio (SOR) &gt; 10.0</code></pre>
</div></div>
</div>
<div class="fragment">
<blockquote>
<p>That said, you ABSOLUTELY SHOULD NOT expect to run these commands and be done with your analyses.</p>
</blockquote>
<div class="flushright smallest translatey50">
<p><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037499012" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360037499012</a></p>
</div>
</div>
<div class="fragment">
<h5 id="on-rad-seq-filtering">On RAD-seq filtering</h5>
<blockquote>
<p>… the effects of SNP filtering practices on population genetic inference have received much less attention</p>
</blockquote>
<div class="flushright smallest translatey50">
<p>There Is No ‘Rule of Thumb’: Genomic Filter Settings for a Small Plant Population to Obtain Unbiased Gene Flow Estimates <span class="citation" data-cites="nazareno_ThereNoRule_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nazareno &amp; Knowles, 2021</a>)</span></p>
</div>
</div>
<aside class="notes">
<p>General guidelines on manual filters are not discussed much in the literature, simply due to the fact that there is no set of rule of thumbs. Every problem requires its own settings, as the GATK developers maintain.</p>
<p>GATK guidelines explained (see <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890471" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360035890471</a>):</p>
<ul>
<li>QualByDepth (QFD): variant confidence (QUAL) divided by unfiltered depth</li>
<li>FisherStrand (FS): checks for strand bias (i.e., if minor allele occurs more often on one strand)</li>
<li>StrandOddsRatio (SOR): alternative strand bias test</li>
<li>RMSMappingQuality (MQ): root mean square mapping quality over all reads</li>
<li>MappingQualytRankSumTest (MQRankSum): compares mapping qualities of ref and alt alleles</li>
<li>ReadPosRankSumTest (ReadPosRankSum): looks at site position within reads</li>
<li>InbreedingCoeff: population-level statistics that requires at least 10 individuals</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="what-about-machine-learning" class="slide level2 ultitle">
<h2>What about machine learning?</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><a href="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fng.806/MediaObjects/41588_2011_Article_BFng806_Fig4_HTML.jpg?as=webp" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fng.806/MediaObjects/41588_2011_Article_BFng806_Fig4_HTML.jpg?as=webp" style="width:100.0%"></a></p>
<div class="flushright smallest">
<p><span class="citation" data-cites="depristo_FrameworkVariationDiscovery_2011">DePristo et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2011</a>)</span></p>
</div>
</div><div class="column" style="width:50%;">
<h5>
Variant Quality Score Recalibration
</h5>
<p>Motivation: look at context statistics and integrate over <em>multiple</em> dimensions</p>
<ul>
<li>training data: subset of known variants (from validated resources, e.g.&nbsp;1000 Genomes)</li>
<li>compile multiple statistics (allele depth, read count, quality, …)</li>
<li>fit Gaussian mixture model</li>
<li>reassign quality scores to variant call set</li>
</ul>
<div class="fragment">
<p>Caveat: database of known variants often <em>not</em> known for non-model organisms.</p>
</div>
</div></div>
<aside class="notes">
<p>Key take home: thresholds that previously were binary yes/no filters now depend on context; for instance, an AD (allele depth) cutoff of 4 will in VQSR sometimes pass, sometimes not, depending on other information</p>
<p>Figure legend:</p>
<ol type="a">
<li class="fragment">Relationship in the HiSeq call set between strand bias and quality by depth for genomic locations in HapMap3 (red) and dbSNP (orange) used for training the variant quality score recalibrator (left), (b) and the same annotations applied to differentiate likely true positive (green) from false positive (purple) new SNPs. (c–e) Quality tranches in the recalibrated HiSeq (c), exome (d) and low-pass CEU (e) calls beginning with (top) the highest quality but smallest call set with an estimated false positive rate among new SNP calls of &lt;1/1000 to a more comprehensive call set (bottom) that includes effectively all true positives in the raw call set along with more false positive calls for a cumulative false positive rate of 10%. Each successive call set contains within it the previous tranche’s true- and false-positive calls (shaded bars) as well as tranche-specific calls of both classes (solid bars). The tranche selected for further analyses here is indicated.</li>
</ol>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section>
<section id="filtering-vcf-with-variant-sites" class="title-slide slide level1 center">
<h1>Filtering VCF with variant sites</h1>

</section>
<section id="monkeyflower-call-set-with-variant-sites-only" class="slide level2 ultitle">
<h2>Monkeyflower call set with variant sites only</h2>
<div>
<p>Variant statistics for a variant only file from 10 Monkeyflower individuals.</p>
<div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a></a><span class="ex">bcftools</span> stats variantsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"^SN"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  12622
SN  0   number of no-ALTs:  0
SN  0   number of SNPs: 10422
SN  0   number of MNPs: 0
SN  0   number of indels:   2221
SN  0   number of others:   0
SN  0   number of multiallelic sites:   1038
SN  0   number of multiallelic SNP sites:   236</code></pre>
</div>
</div>
</div>
<p>Use <code>vcftools</code> to compile data to generate summary statistics</p>
<p>Plot and select thresholds</p>
</div>
</section>
<section id="mean-depth-and-variant-quality-distribution" class="slide level2 ultitle">
<h2>Mean depth and variant quality distribution</h2>
<div class="columns smallr">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>vcf <span class="ot">&lt;-</span> <span class="st">"variantsites.vcf.gz"</span></span>
<span id="cb5-2"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--site-depth 2&gt;/dev/null"</span>))</span>
<span id="cb5-3"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.ldepth"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-4"><a></a>x <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">table</span>(data<span class="sc">$</span>SUM_DEPTH))</span>
<span id="cb5-5"><a></a>lower <span class="ot">&lt;-</span> <span class="fl">0.8</span> <span class="sc">*</span> <span class="fu">median</span>(data<span class="sc">$</span>SUM_DEPTH)</span>
<span id="cb5-6"><a></a>upper <span class="ot">&lt;-</span> <span class="fu">median</span>(data<span class="sc">$</span>SUM_DEPTH) <span class="sc">+</span> <span class="dv">1</span> <span class="sc">*</span> <span class="fu">sd</span>(data<span class="sc">$</span>SUM_DEPTH)</span>
<span id="cb5-7"><a></a>xupper <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(upper<span class="sc">/</span><span class="dv">100</span>) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb5-8"><a></a><span class="fu">ggplot</span>(x, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.numeric</span>(<span class="fu">as.character</span>(Var1)), <span class="at">y =</span> Freq)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Depth"</span>) <span class="sc">+</span></span>
<span id="cb5-9"><a></a>    <span class="fu">ylab</span>(<span class="st">"bp"</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, xupper) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> lower, <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb5-10"><a></a>    <span class="at">size =</span> <span class="fl">1.3</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> upper, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">1.3</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Example threshold: 0.8X median depth, median depth + 2sd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-mean-depth" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-mean-depth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-mean-depth-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;3: Depth uneven. High coverage often repetitive sequence. Too low coverage will bias SNP calling due to undersampling of alleles."><img data-src="variant_filtering_files/figure-revealjs/fig-mean-depth-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mean-depth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Depth uneven. High coverage often repetitive sequence. Too low coverage will bias SNP calling due to undersampling of alleles.
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--site-quality 2&gt;/dev/null"</span>))</span>
<span id="cb6-2"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.lqual"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-3"><a></a><span class="fu">ggplot</span>(<span class="fu">subset</span>(data, QUAL <span class="sc">&lt;</span> <span class="dv">1000</span>), <span class="fu">aes</span>(<span class="at">x =</span> QUAL)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">"white"</span>,</span>
<span id="cb6-4"><a></a>    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Quality value"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">30</span>,</span>
<span id="cb6-5"><a></a>    <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">1.3</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Example threshold: Q30"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-variant-quality-score-distribution" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-variant-quality-score-distribution-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-variant-quality-score-distribution-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;4: Filter variants with too low quality (Q30=0.001% chance of being wrong)"><img data-src="variant_filtering_files/figure-revealjs/fig-variant-quality-score-distribution-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-variant-quality-score-distribution-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Filter variants with too low quality (Q30=0.001% chance of being wrong)
</figcaption>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="missing-data-per-individual-and-site" class="slide level2 ultitle">
<h2>Missing data per individual and site</h2>
<div class="columns smallr">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--missing-indv 2&gt;/dev/null"</span>))</span>
<span id="cb7-2"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.imiss"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-3"><a></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> F_MISS, <span class="at">y =</span> INDV)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Missing data per individual"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-missing-indv" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-missing-indv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-missing-indv-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;5: Missing number of sites per individual. Too many would indicate poor sample quality."><img data-src="variant_filtering_files/figure-revealjs/fig-missing-indv-1.svg" class="quarto-figure quarto-figure-center" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-missing-indv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Missing number of sites per individual. Too many would indicate poor sample quality.
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--missing-site 2&gt;/dev/null"</span>))</span>
<span id="cb8-2"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.lmiss"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-3"><a></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> F_MISS)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">bins =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a></a>    <span class="fu">xlab</span>(<span class="st">"F_MISS"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.25</span>, <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb8-5"><a></a>    <span class="at">size =</span> <span class="fl">1.3</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Missing data per site: example threshold F_MISS=0.25"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-missing-site" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-missing-site-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-missing-site-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;6: Fraction missing calls per site. Could warrant separate filters when comparing populations (e.g., total missing 0.2, but population A has 0.1 missing, population B 0.4)."><img data-src="variant_filtering_files/figure-revealjs/fig-missing-site-1.svg" class="quarto-figure quarto-figure-center" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-missing-site-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Fraction missing calls per site. Could warrant separate filters when comparing populations (e.g., total missing 0.2, but population A has 0.1 missing, population B 0.4).
</figcaption>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="minor-allele-frequency-and-heterozygosity" class="slide level2 ultitle">
<h2>Minor allele frequency and heterozygosity</h2>
<div class="columns smallr">
<div class="column" style="width:50%;">
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--freq2 --max-alleles 2 2&gt;/dev/null"</span>))</span>
<span id="cb9-2"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.frq"</span>, <span class="at">skip =</span> <span class="dv">1</span>)</span>
<span id="cb9-3"><a></a><span class="fu">colnames</span>(data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"CHROM"</span>, <span class="st">"POS"</span>, <span class="st">"N_ALLELES"</span>, <span class="st">"N_CHR"</span>, <span class="st">"FREQ1"</span>, <span class="st">"FREQ2"</span>)</span>
<span id="cb9-4"><a></a>data<span class="sc">$</span>MAF <span class="ot">&lt;-</span> <span class="fu">apply</span>(data, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">as.numeric</span>(<span class="fu">min</span>(x[<span class="dv">5</span>], x[<span class="dv">6</span>])))</span>
<span id="cb9-5"><a></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> MAF)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">bins =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb9-6"><a></a>    <span class="fu">xlab</span>(<span class="st">"MAF"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.1</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">1.3</span>) <span class="sc">+</span></span>
<span id="cb9-7"><a></a>    <span class="fu">ggtitle</span>(<span class="st">"Minor allele frequency: example threshold MAF=0.1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-minor-allele-frequency" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-minor-allele-frequency-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-minor-allele-frequency-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;7: Minor allele frequency distribution"><img data-src="variant_filtering_files/figure-revealjs/fig-minor-allele-frequency-1.svg" class="quarto-figure quarto-figure-center" style="width:100.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-minor-allele-frequency-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Minor allele frequency distribution
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:35%;">
<svg baseprofile="full" height="350" id="msprime-tree" version="1.1" width="250" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.x-regions rect {fill: yellow; stroke: black; opacity: 0.5}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.polytomy line {stroke: black; stroke-dasharray: 1px, 1px}.polytomy text {paint-order:stroke;stroke-width:0.3em;stroke:white}.edge {stroke: black; stroke-width: 3px}.mut .sym{stroke:black} .mut .lab{fill:black}.mut.m0 .sym{stroke:red} .mut.m0 .lab{fill:red}.mut.m4 .sym{stroke:red} .mut.m4 .lab{fill:red}.mut.m5 .sym{stroke:red} .mut.m5 .lab{fill:red}.node > .sym, .node:not(.leaf) > .lab {display: none}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n22 p0 root" transform="translate(113.411 26.8)"><g class="a22 c2 m2 m3 m6 node n20 p0 s2 s3 s6" transform="translate(-47.3698 207.029)"><g class="a20 c2 node n16 p0" transform="translate(-21.25 42.7476)"><g class="a16 i3 leaf node n6 p0 sample" transform="translate(10.625 31.6239)"><path class="edge" d="M 0 0 V -31.6239 H -10.625"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a16 c2 node n14 p0" transform="translate(-10.625 11.1848)"><g class="a14 i0 leaf node n0 p0 sample" transform="translate(-7.08333 20.4391)"><path class="edge" d="M 0 0 V -20.4391 H 7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a14 i5 leaf m5 node n11 p0 s5 sample" transform="translate(7.08333 20.4391)"><path class="edge" d="M 0 0 V -20.4391 H -7.08333"></path><g class="mut m5 s5" transform="translate(0 -19.9892)"><line x1="0" x2="0" y1="0" y2="19.9892"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab rgt" transform="translate(7 0)">5</text></g><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><path class="edge" d="M 0 0 V -11.1848 H 10.625"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><path class="edge" d="M 0 0 V -42.7476 H 21.25"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><g class="a20 c2 node n17 p0" transform="translate(21.25 31.2231)"><g class="a17 i2 leaf node n5 p0 sample" transform="translate(10.625 43.1483)"><path class="edge" d="M 0 0 V -43.1483 H -10.625"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a17 c2 node n13 p0" transform="translate(-10.625 34.194)"><g class="a13 i0 leaf node n1 p0 sample" transform="translate(-7.08333 8.95427)"><path class="edge" d="M 0 0 V -8.95427 H 7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a13 i1 leaf node n2 p0 sample" transform="translate(7.08333 8.95427)"><path class="edge" d="M 0 0 V -8.95427 H -7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><path class="edge" d="M 0 0 V -34.194 H 10.625"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><path class="edge" d="M 0 0 V -31.2231 H -21.25"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab rgt" transform="translate(5 -7)"></text></g><path class="edge" d="M 0 0 V -207.029 H 47.3698"></path><g class="mut m6 s6" transform="translate(0 -192.786)"><line x1="0" x2="0" y1="0" y2="192.786"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab lft" transform="translate(-7 0)">6</text></g><g class="mut m3 s3" transform="translate(0 -93.6026)"><line x1="0" x2="0" y1="0" y2="93.6026"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab lft" transform="translate(-7 0)">3</text></g><g class="mut m2 s2" transform="translate(0 -5.15485)"><line x1="0" x2="0" y1="0" y2="5.15485"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab lft" transform="translate(-7 0)">2</text></g><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><g class="a22 c2 m1 m7 node n21 p0 s1 s7" transform="translate(47.3698 174.459)"><g class="a21 i5 leaf m0 m4 node n10 p0 s0 s4 sample" transform="translate(22.1354 106.941)"><path class="edge" d="M 0 0 V -106.941 H -22.1354"></path><g class="mut m0 s0" transform="translate(0 -72.1134)"><line x1="0" x2="0" y1="0" y2="72.1134"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab rgt" transform="translate(7 0)">0</text></g><g class="mut m4 s4" transform="translate(0 -42.0215)"><line x1="0" x2="0" y1="0" y2="42.0215"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab rgt" transform="translate(7 0)">4</text></g><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a21 c2 node n19 p0" transform="translate(-22.1354 34.1086)"><g class="a19 c2 node n15 p0" transform="translate(-19.4792 46.346)"><g class="a15 i1 leaf node n3 p0 sample" transform="translate(-7.08333 26.4862)"><path class="edge" d="M 0 0 V -26.4862 H 7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a15 i3 leaf node n7 p0 sample" transform="translate(7.08333 26.4862)"><path class="edge" d="M 0 0 V -26.4862 H -7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><path class="edge" d="M 0 0 V -46.346 H 19.4792"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><g class="a19 c2 node n18 p0" transform="translate(19.4792 16.611)"><g class="a18 i4 leaf node n8 p0 sample" transform="translate(10.625 56.2212)"><path class="edge" d="M 0 0 V -56.2212 H -10.625"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a18 c2 node n12 p0" transform="translate(-10.625 52.7313)"><g class="a12 i2 leaf node n4 p0 sample" transform="translate(-7.08333 3.48991)"><path class="edge" d="M 0 0 V -3.48991 H 7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><g class="a12 i4 leaf node n9 p0 sample" transform="translate(7.08333 3.48991)"><path class="edge" d="M 0 0 V -3.48991 H -7.08333"></path><rect class="sym" height="10" width="10" x="-5" y="-5"></rect><text class="lab" transform="translate(0 9)"></text></g><path class="edge" d="M 0 0 V -52.7313 H 10.625"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><path class="edge" d="M 0 0 V -16.611 H -19.4792"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab rgt" transform="translate(5 -7)"></text></g><path class="edge" d="M 0 0 V -34.1086 H 22.1354"></path><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab lft" transform="translate(-5 -7)"></text></g><path class="edge" d="M 0 0 V -174.459 H -47.3698"></path><g class="mut m7 s7" transform="translate(0 -148.516)"><line x1="0" x2="0" y1="0" y2="148.516"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab rgt" transform="translate(7 0)">7</text></g><g class="mut m1 s1" transform="translate(0 -30.3651)"><line x1="0" x2="0" y1="0" y2="30.3651"></line><path class="sym" d="M -5,-5 l 10,10 M -5,5 l 10,-10"></path><text class="lab rgt" transform="translate(7 0)">1</text></g><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab rgt" transform="translate(5 -7)"></text></g><circle class="sym" cx="0" cy="0" r="5"></circle><text class="lab" transform="translate(0 -9)"></text></g></g></g></g></g>
</svg>
<p>n=12; mutations 0, 4, 5 (red) are singletons and would fail MAF&lt;=0.1</p>
</div></div>
<p>Reasonable cutoff 0.05-0.1 for PCA, population structure.</p>
<p>But! Statistics based on diversity or the SFS should <strong>not</strong> be filtered on MAF</p>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a><span class="fu">system</span>(<span class="fu">paste</span>(<span class="st">"vcftools --gzvcf"</span>, vcf, <span class="st">"--het 2&gt;/dev/null"</span>))</span>
<span id="cb10-2"><a></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"out.het"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-3"><a></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> F, <span class="at">y =</span> INDV)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Inbreeding coefficient"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heterozygosity" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-heterozygosity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="variant_filtering_files/figure-revealjs/fig-heterozygosity-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;8: Genome-wide inbreeding coefficient by individual"><img data-src="variant_filtering_files/figure-revealjs/fig-heterozygosity-1.svg" class="quarto-figure quarto-figure-center" style="width:60.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heterozygosity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Genome-wide inbreeding coefficient by individual
</figcaption>
</figure>
</div>
</div>
</div>
<ul>
<li>F=0: Hardy-Weinberg Equilibrium</li>
<li>F&gt;0: deficit of heterozygotes; inbreeding, Wahlund effect (population substructure), allele dropout</li>
<li>F&lt;0: surplus of heterozygotes; could be sample contamination, poor sequence quality (mismapping)</li>
</ul>
</div></div>
</section></section>
<section>
<section id="filtering-vcf-with-invariant-sites" class="title-slide slide level1 center">
<h1>Filtering VCF with invariant sites</h1>

</section>
<section id="monkeyflower-call-set-with-invariant-sites" class="slide level2 ultitle">
<h2>Monkeyflower call set with invariant sites</h2>
<div>
<p>Variant statistics for an all sites file from 10 Monkeyflower individuals.</p>
<div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a></a><span class="ex">bcftools</span> stats allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"^SN"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  10035
SN  0   number of no-ALTs:  9286
SN  0   number of SNPs: 344
SN  0   number of MNPs: 0
SN  0   number of indels:   117
SN  0   number of others:   0
SN  0   number of multiallelic sites:   49
SN  0   number of multiallelic SNP sites:   5</code></pre>
</div>
</div>
<p>Filtering as before but excluding MAF, variant quality filters</p>
</div>
</div>
</section></section>
<section>
<section id="filtering-and-population-genetics" class="title-slide slide level1 center">
<h1>Filtering and population genetics</h1>

</section>
<section id="next-generation-data-filtering-in-the-genomics-era" class="slide level2 ultitle">
<h2>Next-generation data filtering in the genomics era</h2>
<div id="fig-hemstrom" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-hemstrom-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41576-024-00738-6/MediaObjects/41576_2024_738_Fig2_HTML.png?as=webp" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;9: Post-variant filtering challenges. a) joint filtering of missing data can obscure population quality differences b) SNP number varies threefold if MAF within sample-group filtering is applied; would be missed if genome-wide filtering c) population structure inflates HWP which may inflate study-wide filtering on HWP d) LD thinning may obscure inversions"><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41576-024-00738-6/MediaObjects/41576_2024_738_Fig2_HTML.png?as=webp" style="width:45.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-hemstrom-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Post-variant filtering challenges. a) joint filtering of missing data can obscure population quality differences b) SNP number varies threefold if MAF within sample-group filtering is applied; would be missed if genome-wide filtering c) population structure inflates HWP which may inflate study-wide filtering on HWP d) LD thinning may obscure inversions
</figcaption>
</figure>
</div>
<div class="flushright">
<p><span class="citation" data-cites="hemstrom_NextgenerationDataFiltering_2024">Hemstrom et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2024</a>)</span>, Fig. 2</p>
</div>
<aside class="notes">
<p>Figure legend:</p>
<blockquote>
<p>a, Missing data, which can occur across loci and individuals. Data from monarch butterflies114 are used to show that the percentage of missing data can be high (21–100% per locus) when missing data filtering occurs within sample-groups, but is much lower if performed jointly across all samples (19–56%), the latter of which can obscure the differences in data quality among populations. b, Data from a study on yellow perch102 showing that the number of single-nucleotide polymorphisms (SNPs) varies threefold among populations if a minor allele frequency (MAF) filter of 0.01 is applied within sample-groups, which would be missed if the same filter was applied study-wide. c, Loci can deviate from Hardy–Weinberg proportions (HWP) due to homozygote or heterozygote excesses owing to biological causes, which filtering may obscure. For example, unintentionally combining two divergent sample-groups will result in higher FIS (fixation index, individuals versus subpopulation) and many more loci out of HWP (owing to Wahlund effects)39. Filtering study-wide would cause the erroneous removal of loci in such cases. d, Linkage disequilibrium (LD) thinning (that is, filtering out loci with high LD; red points) can obscure an inversion (blue points) in three-spined stickleback haplotypes141. FST, fixation index, subpopulation versus total; MAC, minor allele count.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller ultitle scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-depristo_FrameworkVariationDiscovery_2011" class="csl-entry" role="listitem">
DePristo, M. A., Banks, E., Poplin, R., Garimella, K. V., Maguire, J. R., Hartl, C., Philippakis, A. A., del Angel, G., Rivas, M. A., Hanna, M., McKenna, A., Fennell, T. J., Kernytsky, A. M., Sivachenko, A. Y., Cibulskis, K., Gabriel, S. B., Altshuler, D., &amp; Daly, M. J. (2011). A framework for variation discovery and genotyping using next-generation <span>DNA</span> sequencing data. <em>Nature Genetics</em>, <em>43</em>(5), 491–498. <a href="https://doi.org/10.1038/ng.806">https://doi.org/10.1038/ng.806</a>
</div>
<div id="ref-hemstrom_NextgenerationDataFiltering_2024" class="csl-entry" role="listitem">
Hemstrom, W., Grummer, J. A., Luikart, G., &amp; Christie, M. R. (2024). Next-generation data filtering in the genomics era. <em>Nature Reviews Genetics</em>, 1–18. <a href="https://doi.org/10.1038/s41576-024-00738-6">https://doi.org/10.1038/s41576-024-00738-6</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-nazareno_ThereNoRule_2021" class="csl-entry" role="listitem">
Nazareno, A. G., &amp; Knowles, L. L. (2021). There <span>Is No</span> <span>“<span>Rule</span> of <span>Thumb</span>”</span>: <span>Genomic Filter Settings</span> for a <span>Small Plant Population</span> to <span>Obtain Unbiased Gene Flow Estimates</span>. <em>Frontiers in Plant Science</em>, <em>12</em>. <a href="https://www.frontiersin.org/articles/10.3389/fpls.2021.677009">https://www.frontiersin.org/articles/10.3389/fpls.2021.677009</a>
</div>
<div id="ref-talla_DissectingEffectsSelection_2019" class="csl-entry" role="listitem">
Talla, V., Soler, L., Kawakami, T., Dincă, V., Vila, R., Friberg, M., Wiklund, C., &amp; Backström, N. (2019). Dissecting the <span>Effects</span> of <span>Selection</span> and <span>Mutation</span> on <span>Genetic Diversity</span> in <span>Three Wood White</span> (<span>Leptidea</span>) <span>Butterfly Species</span>. <em>Genome Biology and Evolution</em>, <em>11</em>(10), 2875–2886. <a href="https://doi.org/10.1093/gbe/evz212">https://doi.org/10.1093/gbe/evz212</a>
</div>
</div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Variant filtering</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>