---
title: Variant calling index
description: List of exercises
format:
  nbis-course-html:
    number-sections: false
author:
  - Per Unneberg
exercise:
  dir: variant-calling
  venv: e-variant-calling
listing:
  id: vc-listing
  type: table
  contents:
    - introduction.qmd
    - data_qc.qmd
    - read_mapping.qmd
    - variant_calling_workflow.qmd
---

## About

A generic variant calling workflow consists of the following basic
steps:

1. read quality control and filtering
2. read mapping
3. removal / marking of duplicate reads
4. joint / sample-based variant calling and genotyping

There are different tweaks and additions to each of these steps,
depending on application and method. The variant calling exercises
here present the basic steps to go from raw data to variant calls.

The exercises are based on the [Monkeyflowers
dataset](/exercises/datasets/monkeyflowers.qmd). Make sure to read the
dataset document before running any commands as it will give you the
biological background and general information about where to find and
how to setup the data. We will focus on the *red* and *yellow*
ecotypes in what follows.

## Intended learning outcomes

- Perform qc on sequencing reads and interpret results
- Prepare reference for read mapping
- Map reads to reference
- Mark duplicates
- Perform raw variant calling to generate a set of sites to exclude
  from recalibration
- Perform base quality score recalibration
- Perform variant calling on base recalibrated data
- Do genotyping on all samples and combine results to a raw variant
  call set

## Listing

:::{#vc-listing}
:::

## Additional material

[Variant calling, long description](variant_calling_long.qmd){#variant-calling-long}

: Describes all steps of a standard variant calling workflow from data
  preparation to final summary QC. All commands are run manually
  without the aid of a workflow manager.
