---
title: Data and definitions
subtitle: Populations genetics foundations
description: >-
  Introduction to population genetic data and definitions.
author:
  - Per Unneberg
format:
  nbis-quarto-revealjs:
    toc: false
    smaller: false
    footer: Population genetics foundations - data
module:
  - variant calling
---

## Setup  {.hidden .unnumbered .unlisted visibility="hidden"}

{{< include /slides/_knitr.qmd >}}

{{< include /slides/_rlibs.qmd >}}

## DNA variation

![The variable sites at the *Drosophila melanogaster* ADH locus
[@kreitman_NucleotidePolymorphismAlcohol_1983]](/slides/pgip/assets/images/kreitman-1983-table1.webp){width=100%
fig-align=center}\

::: {.notes }

Goal of section: look at the data that forms the foundation for
population genomic analyses

From [@nei_MolecularEvolutionPhylogenetics_2000, p. 231]:

> The main subject of population genetics is to study the generation
> and maintenance of genetic polymorphism and to understand the
> mechanisms of evolution at the population level

[@casillas_MolecularPopulationGenetics_2017, p. 1026]:

> Big data samples of complete genome sequences of many individuals
> from natural populations of many species have transformed population
> genetics inferences on samples of loci to population genomics: the
> analysis of genome-wide patterns of DNA variation within and between
> species.

[@gillespie_PopulationGeneticsConcise_2004, p. 1]

> Population geneticists spend most of their time doing one of two
> things: describing the genetic structure of populations or
> theorizing on the evolutionary forces acting on populations. On a
> good day, these two activities mesh and true insights emerge.

:::

## DNA variation - monomorphic sites

::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| * |   | * | * |   | * | * |   | * | *  |    | *  |    | *  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

The alignment has 4 DNA sequences where each sequence has length
$L=15$. A site where all nucleotides (alleles) are identical is called
a *monomorphic* site (indicated with asterisks above). There are 9
monomorphic sites.

## DNA variation - segregating sites

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

A site where there are different nucleotides (alleles) is called a
*segregating* site (indicated with asterisks above), often denoted
*S*. There are $S=6$ segregating sites.

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

Alternative names for *segregating site* are:

:::

::: {.incremental}

- *polymorphism*
- *mutation*
- *single nucleotide polymorphism (SNP)*

:::

:::

::: {.column width="50%"}

::: {.fragment}

*mutation* here and onwards refers to the process that generates new
variation and the new variants generated by this process

In contrast to *mutation* which corresponds to within-species
variation, a *substitution* refers to DNA differences between species.

:::

:::

::::

## DNA variation - major and minor alleles

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

Much of the nucleotide variation we study consists of *bi-allelic*
SNPs. The most common variant is called the *major* allele, and the
least common the *minor* allele.

The set of alleles found on a single sequence is called *haplotype*.

## Describing DNA variation - heterozygosity {.smaller}

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

Once we have a sample of sequences we want to describe the observed
variation. At any position the *i*th allele has sample frequency
$p_i$, where the sum of all allele frequencies is 1. For instance, at
site 1, $p_T=1$ (and by extension $p_A=p_C=p_G=0$), and at site 2
$p_C=1/4$ and $p_T=3/4$.

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

<h4>Heterozygosity</h4>

The *heterozygosity* at a site $j$ is given by

$$
h_j = \frac{n}{n-1}\left(1 - \sum_i p_i^2\right)
$$

where the summation is over all alleles and $p_i$ is the frequency of
the $i$-th allele

:::

:::

::: {.column width="50%"}

::: {.fragment}

<h4>Exercise: calculate the heterozygosity at sites 1, 2 and 5</h4>

:::

::: {.fragment}

<!-- markdownlint-disable MD013 -->

$$
\begin{align*}
h_1 & = \frac{4}{3} \left(1 - p_T^2 \right) = 0 \\[10pt]
h_2 & = \frac{4}{3} \left(1 - \left(p_C^2 + p_T^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{16} + \frac{9}{16}\right)\right) = \frac{1}{2}\\[10pt]
h_5 & = \frac{4}{3} \left(1 - \left(p_A^2 + p_G^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{4} + \frac{1}{4}\right)\right) = \frac{2}{3}
\end{align*}
$$

<!-- markdownlint-enable MD013 -->

:::

:::

::::

::: {.notes}

In a randomly mating population, the heterozygosity is equal to the
frequency of heterozygotes. Note however that the definition of
heterozygosity only relies on allele frequencies, which means it can
be applied to populations that are not in Hardy-Weinberg equilibrium,
or to more general variation, such as protein isoforms. It can also be
applied to haploid organisms, like bacteria.

:::

## Describing DNA variation - nucleotide diversity {.smaller}

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

<h4>Nucleotide diversity $\pi$</h4>

The *nucleotide diversity* is the sum of site heterozygosities:

$$
\pi = \sum_{j=1}^S h_j
$$

where $S$ is the number of segregating sites

:::

:::

::: {.column width="50%"}

::: {.fragment}

<h4>Calculate the nucleotide diversity</h4>

:::

::: {.fragment}

Observation: $h_i$ either 1/2 or 2/3 (for sites with
$p_{major}=p_{minor}$).

:::

::: {.fragment}

<!-- markdownlint-disable MD013 -->

$$
\pi = \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{1}{2} = 3\frac{1}{3}
$$

<!-- markdownlint-enable MD013 -->

:::

::: {.fragment}

Often we report $\pi$ per site:

$$
\pi = 3.33/15 = 0.222
$$

:::

:::

::::

::: {.notes}

@hahn_MolecularPopulationGenetics_2019 implicitly assumes we are
looking at DNA polymorphism. The expression actually holds for any
genetic variation at a locus, and is sometimes called the *gene
diversity* [@nei_MolecularEvolutionPhylogenetics_2000, p. 245].

Under the infinite sites model, $E(\pi)=\theta=4N_e\mu$, for which
reason $\pi$ sometimes is called $\theta_\pi$. The measure gives the
average number of pairwise nucleotide differences between two
sequences, so an alternative expression is

$$
\pi = \frac{\sum_{i<j}k_{ij}}{n(n-1)/2}
$$

The latter expression is called the *nucleotide diversity*
[@nei_MolecularEvolutionPhylogenetics_2000 p. 251].

:::

## The real data

```{bash }
#| label: samtools-faidx-reference
#| echo: false
#| eval: true
samtools faidx ref/M_aurantiacus_v1.fasta
```

:::: {.columns}

::: {.column width="40%"}

```{bash }
#| label: show-fastq-real-data
#| echo: false
#| eval: true
zcat fastq/PUN-Y-INJ_R1.fastq.gz | head --lines 8 | cut --characters -30
```

:::

::: {.column width="60%"}

```{bash }
#| label: samtools-tview-PUN-Y-real-data
#| echo: false
#| eval: true
#| results: asis
samtools tview -p LG4:30430 -d H -w 47 \
   bam/PUN-Y-INJ.bam \
   ref/M_aurantiacus_v1.fasta
```

:::

::::

:::{}

Before getting to variants and genotypes a lot of processing has to be
done, from FASTQ input, to mapped data, to variant and genotype calls.

:::

::: {.notes}

Before getting to variants and genotypes a lot of processing has to be
done, from FASTQ input, to mapped data, to variant and genotype calls.

:::

## The process

:::{}

![Schematic overview of population genomic workflow from sample
collection via sequencing and data analysis to genotype
output](/slides/foundations/assets/images/popgen.svg)\

:::

## Bibliography {.unnumbered .unlisted .smaller}

::: { #refs }
:::
