---
title: "Population genetics"
subtitle: "Foundations"
author:
  - "Per Unneberg"
format:
  nbis-quarto-revealjs:
    toc: false
    smaller: false
    footer: Population genetics foundations
---

## Setup  {.hidden .unnumbered .unlisted visibility="hidden"}

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}

```{r libs}
#| echo: false
#| eval: true
#| cache: false
library(viridisLite)
library(gganimate)
library(patchwork)
library(expm)
library(gridExtra)
library(igraph)
library(ggraph)
library(tidygraph)
```

## Intended learning outcomes {.unnumbered .unlisted}

Introduction to foundations of population genetics with an emphasis on
genealogies

- Description of DNA variation data
- Wright-Fisher population model and genealogies
- Genetic drift
- Wright-Fisher model with mutation
- Mutation-drift balance
- Neutral theory
- Selection basics

::: {.notes}

Objective: introduce the foundations of population genetics, with
an emphasis on genealogies. Framing population genetic concepts in
trees will facilitate the introduction of the coalescent in the
afternoon.

The lecture presents sequence data, the Wright-Fisher model,
genetic drift, mutation, and selection. The focus is to
familiarise students with concepts and basic theory.

:::

# DNA variation

:::{}

![](assets/images/popgen.svg)

:::

::: {.notes }

Goal of section: look at the data that forms the foundation for
population genomic analyses

From [@nei_MolecularEvolutionPhylogenetics_2000, p. 231]:

> The main subject of population genetics is to study the generation
> and maintenance of genetic polymorphism and to understand the
> mechanisms of evolution at the population level

[@casillas_MolecularPopulationGenetics_2017, p. 1026]:

> Big data samples of complete genome sequences of many individuals
> from natural populations of many species have transformed population
> genetics inferences on samples of loci to population genomics: the
> analysis of genome-wide patterns of DNA variation within and between
> species.

[@gillespie_PopulationGeneticsConcise_2004, p. 1]

> Population geneticists spend most of their time doing one of two
> things: describing the genetic structure of populations or
> theorizing on the evolutionary forces acting on populations. On a
> good day, these two activities mesh and true insights emerge.

:::

## DNA variation - monomorphic sites

::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| * |   | * | * |   | * | * |   | * | *  |    | *  |    | *  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

The alignment has 4 DNA sequences where each sequence has length
$L=15$. A site where all nucleotides (alleles) are identical is called
a *monomorphic* site (indicated with asterisks above). There are 9
monomorphic sites.

## DNA variation - segregating sites

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

A site where there are different nucleotides (alleles) is called a
*segregating* site (indicated with asterisks above), often denoted
*S*. There are $S=6$ segregating sites.

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

Alternative names for *segregating site* are:

:::

::: {.incremental}

- *polymorphism*
- *mutation*
- *single nucleotide polymorphism (SNP)*

:::

:::

::: {.column width="50%"}

::: {.fragment}

*mutation* here and onwards refers to the process that generates new
variation and the new variants generated by this process

In contrast to *mutation* which corresponds to within-species
variation, a *substitution* refers to DNA differences between species.

:::

:::

::::

## DNA variation - major and minor alleles

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

Much of the nucleotide variation we study consists of *bi-allelic*
SNPs. The most common variant is called the *major* allele, and the
least common the *minor* allele.

The set of alleles found on a single sequence is called *haplotype*.

## Describing DNA variation - heterozygosity {.smaller}

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

Once we have a sample of sequences we want to describe the observed
variation. At any position the *i*th allele has sample frequency
$p_i$, where the sum of all allele frequencies is 1. For instance, at
site 1, $p_T=1$ (and by extension $p_A=p_C=p_G=0$), and at site 2
$p_C=1/4$ and $p_T=3/4$.

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

<h4>Heterozygosity</h4>

The *heterozygosity* at a site $j$ is given by

$$
h_j = \frac{n}{n-1}\left(1 - \sum_i p_i^2\right)
$$

where the summation is over all alleles and $p_i$ is the frequency of
the $i$-th allele

:::

:::

::: {.column width="50%"}

::: {.fragment}

<h4>Exercise: calculate the heterozygosity at sites 1, 2 and 5</h4>

:::

::: {.fragment}

<!-- markdownlint-disable MD013 -->

$$
\begin{align*}
h_1 & = \frac{4}{3} \left(1 - p_T^2 \right) = 0 \\[10pt]
h_2 & = \frac{4}{3} \left(1 - \left(p_C^2 + p_T^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{16} + \frac{9}{16}\right)\right) = \frac{1}{2}\\[10pt]
h_5 & = \frac{4}{3} \left(1 - \left(p_A^2 + p_G^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{4} + \frac{1}{4}\right)\right) = \frac{2}{3}
\end{align*}
$$

<!-- markdownlint-enable MD013 -->

:::

:::

::::

::: {.notes}

In a randomly mating population, the heterozygosity is equal to the
frequency of heterozygotes. Note however that the definition of
heterozygosity only relies on allele frequencies, which means it can
be applied to populations that are not in Hardy-Weinberg equilibrium,
or to more general variation, such as protein isoforms. It can also be
applied to haploid organisms, like bacteria.

:::

## Describing DNA variation - nucleotide diversity {.smaller}

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

<h4>Nucleotide diversity $\pi$</h4>

The *nucleotide diversity* is the sum of site heterozygosities:

$$
\pi = \sum_{j=1}^S h_j
$$

where $S$ is the number of segregating sites

:::

:::

::: {.column width="50%"}

::: {.fragment}

<h4>Calculate the nucleotide diversity</h4>

:::

::: {.fragment}

Observation: $h_i$ either 1/2 or 2/3 (for sites with
$p_{major}=p_{minor}$).

:::

::: {.fragment}

<!-- markdownlint-disable MD013 -->

$$
\pi = \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{1}{2} = 3\frac{1}{3}
$$

<!-- markdownlint-enable MD013 -->

:::

::: {.fragment}

Often we report $\pi$ per site:

$$
\pi = 3.33/15 = 0.222
$$

:::

:::

::::

::: {.notes}

@hahn_MolecularPopulationGenetics_2019 implicitly assumes we are
looking at DNA polymorphism. The expression actually holds for any
genetic variation at a locus, and is sometimes called the *gene
diversity* [@nei_MolecularEvolutionPhylogenetics_2000, p. 245].

Under the infinite sites model, $E(\pi)=\theta=4N_e\mu$, for which
reason $\pi$ sometimes is called $\theta_\pi$. The measure gives the
average number of pairwise nucleotide differences between two
sequences, so an alternative expression is

$$
\pi = \frac{\sum_{i<j}k_{ij}}{n(n-1)/2}
$$

The latter expression is called the *nucleotide diversity*
[@nei_MolecularEvolutionPhylogenetics_2000 p. 251].

:::

## Alleles as algebraic entities

:::: {.columns}

::: {.column width="50%"}

Recall: **alleles** refer to *different variants* of a sequence at a
locus (genomic position).

::: {.fragment}

Whatever the underlying molecular nature (gene, chromosome,
nucleotide, protein), let's represent a locus by a letter, e.g., $A$
($B$ if two loci, and so on)

:::

::: {.fragment}

If locus has many alleles $1, 2, ...$ , could use indexing $A_1, A_2, ...$.

:::

::: {.fragment}

Will use combination $A$, $a$ for *bi-allelic* loci from now on

:::

:::

::: {.column width="50%"}

::: {.fragment}

::: {.callout-important appearance="simple" icon=false}

### Example: gene coding for flower color

:::::: {.columns}

::::: {.column width="50%"}

[{{< fa solid circle >}}]{style="color: #ff2a2a;"} $A$ red color

{{< fa regular circle >}} $a$ white color

::: {.fragment}

[Punnett square](https://en.wikipedia.org/wiki/Punnett_square)

:::: {style="display: grid; grid-template-columns: auto auto auto;
grid-row-gap: 0px; grid-column-gap: 0px; text-align: center;
align-items: center;"}

:::{}

{{< fa venus >}} \\ {{< fa mars >}}

:::

:::{}

A

:::

:::{}

a

:::

:::{}

A

:::

:::{}

![](assets/images/flower-red.svg){width=75}

:::

:::{}

![](assets/images/flower-pink.svg){width=75}

:::

:::{}

a

:::

:::{}

![](assets/images/flower-pink.svg){width=75}

:::

:::{}

![](assets/images/flower-white.svg){width=75}

:::

::::

:::

:::::

:::::{.column width="5%"}

:::::

::::: {.column width="45%" .fragment}

:::: {style="display: grid; grid-template-columns: 1fr 1fr 1fr 1fr;
grid-row-gap: 0px; grid-column-gap: 10px; text-align: center; align-items: center;"}

:::{}

Genotype

:::

:::{}

aa

:::

:::{}

Aa

:::

:::{}

AA

:::

:::{}

Phenotype

:::

:::{}

![](assets/images/flower-white.svg){width=100}

:::

:::{}

![](assets/images/flower-pink.svg){width=100}

:::

:::{}

![](assets/images/flower-red.svg){width=100}

:::

::::

Heterozygote has intermediate color phenotype (pink).

:::::

::::::

:::

:::

:::

::::

::: {.notes}

Until now the examples have been based on nucleotide sequences.
However, much of population genetic theory was developed before the
nature of heredity (DNA) was known. In these early days, an allele
would refer to variant forms of a gene, observed as differences in
phenotypes. Genes, or loci, would be denoted using alphabetic
characters, such as $A$, and allelic types could be referenced with
indices, e.g., $A_1, A_2, ..., A_n$.

To simplify calculations, we often look at one locus and we assume two
alleles, whereby we skip the indices and denote the allelic pairs $A$
and $a$ (although note that notations differs from author to author;
for instance @gillespie_PopulationGeneticsConcise_2004 uses $A_1, A_2$
for bi-allelic loci). For two-locus systems we simply denote the
second allele with $B, b$, and so on.

The example shows a hypothetical locus having two alleles $A$ and $a$
that have phenotypes red and white flower color, and where
heterozygotes are colored pink. The *Punnett square* shows how gamete
combinations form genotypes and their corresponding phenotypes.

:::

## Alleles and frequencies

We will be interested in looking at the dynamics of alleles, i.e., how
their abundances in the population change over time. Therefore we want
to measure the *frequencies* of alleles $A$ and $a$.

:::::: {.columns}

::::: {.column width="55%"}

::: {.fragment fragment-index=1}

::: {.callout-important appearance="simple" icon=false}

### Example

Assume following population ($n=10$, with $n_{AA}=5$, $n_{Aa}=4$,
$n_{aa}=1$):

![](assets/images/flower-red.svg){width=50}
![](assets/images/flower-red.svg){width=50}
![](assets/images/flower-red.svg){width=50}
![](assets/images/flower-red.svg){width=50}
![](assets/images/flower-red.svg){width=50}
![](assets/images/flower-pink.svg){width=50}
![](assets/images/flower-pink.svg){width=50}
![](assets/images/flower-pink.svg){width=50}
![](assets/images/flower-pink.svg){width=50}
![](assets/images/flower-white.svg){width=50}

::: {.fragment fragment-index=2}

Let **$p$** be frequency of $A$ alleles, **$q=1-p$** frequency of $a$ alleles; then

5 $AA$ individuals, 4 $Aa$ individuals $\Rightarrow p=\frac{5\cdot2 +
4\cdot1}{10\cdot2}=\frac{14}{20}=0.7$

and $q=1-p=\frac{6}{20}=0.3$

:::

::: {.fragment fragment-index=3}

Inserting frequencies into **Punnett square** gives expected frequency
of offspring genotypes.

:::

:::

:::

:::::

::::: {.column width="45%"}

:::: {.fragment fragment-index=4}

:::: {style="display: grid; grid-template-columns: auto auto auto;
grid-row-gap: 0px; grid-column-gap: 0px; text-align: center;
align-items: center; font-size: 0.8em;"}

:::{}

{{< fa venus >}} \\ {{< fa mars >}}

:::

:::{}

$A$ ($p=0.7$)

:::

:::{}

$a$ ($q=0.3$)

:::

:::{}

$A$ ($p=0.7$)

:::

:::{}

![](assets/images/flower-red.svg){width=50}

$p\cdot p = 0.49$

:::

:::{}

![](assets/images/flower-pink.svg){width=50}

$p\cdot q = 0.21$

:::

:::{}

$a$ ($q=0.3$)

:::

:::{}

![](assets/images/flower-pink.svg){width=50}

$q\cdot p = 0.21$

:::

:::{}

![](assets/images/flower-white.svg){width=50}

$q\cdot q = 0.09$

:::

::::

::::

::: {.fragment fragment-index=5 .small}

**Expected** allele frequencies after mating: $p=p^2 + pq=0.7$, $q=1-p=0.3$

:::

:::::

::::::

::: {.notes}

Given $n$ diploid individuals, there are $2n$ alleles in the
population. The frequency of allele $A$ is then the homozygote $AA$
times two, plus one times the individuals carrying one $A$.

The Punnett square shows that the expected homozygote frequencies are
$0.49$ and $0.09$ for $AA$ and $aa$, and hence, the frequency of $Aa$
is $1 - 0.49 - 0.09 = 0.42 = 0.21 + 0.21 = 2pq$

:::

## In absence of evolutionary forces alleles are in equilibrium{.smaller}

### The Hardy-Weinberg equilibrium

For a locus, let $A$ and $a$ be two different alleles and let $p$ be
the frequency of the $A$ allele and $q=1-p$ the frequency of the $a$
allele. In the **absence** of **mutation**, **drift**, **migration**,
and other evolutionary processes, the equilibrium state is given by
the **Hardy-Weinberg equilibrium** (**HWE**).

:::: {.columns=center}

::: {.column width="45%"}

|           | $A$ ($p$) | $a$ ($q$) |
|-----------|-----------|-----------|
| $A$ ($p$) | $p^2$     | $pq$      |
| $a$ ($q$) | $qp$      | $q^2$     |

:::

::: {.column width="45%"}

|            |          |          |          |
|------------|----------|----------|----------|
| Genotype:  | $AA$     | $Aa$     | $aa$     |
| Frequency: | $p^2$    | $2pq$    | $q^2$    |
|            | $f_{AA}$ | $f_{Aa}$ | $f_{aa}$ |

:::

::::

Under HWE assumption, neither allele nor genotype frequencies change
over time.

Importantly, we can calculate allele frequencies from genotype
frequencies and vice versa:

$$
p = f_{AA} + \frac{f_{Aa}}{2} = p^2 + pq\\
q = f_{aa} + \frac{f_{Aa}}{2} = q^2 + pq\\
$$

::: {.notes }

Segue: apart from describing the variation via e.g., diversity
measures, we want to model how allele frequencies *change* in time. As
[@gillespie_PopulationGeneticsConcise_2004, preface p. xi] points out,
"While genotype frequencies are easily measured, their change is not"

IOW: describing variation fine, but where does it come from and how
does it change?

If assumptions of HWE hold, we have no change of variation. However,
we **want** to look at **change** of variation and disentangle the
forces that impose change

HWE assumption gives us a way to calculate allele frequencies from
genotype frequencies. How well do these assumptions hold in real data?
See next slide.

:::

## Natural populations do mate randomly?

:::{}

```{r }
#| label: fig-hwe-1000genomes
#| echo: false
#| eval: true
#| out-width: 80%
#| fig-height: 6
#| fig-width: 14
#| fig-align: center
#| fig-cap:
#|     Hardy-Weinberg proportions in 10,000 SNPs on chromosome
#|     22 from three populations based on 1000 genomes data.
#|     For each SNP, genotypes are given as counts (minor/heterozygote/major),
#|     converted to frequencies and plotted on the y-axis. Allele
#|     frequencies are obtained from genotype frequencies and
#|     plotted on the x-axis. Most observations follow HWE proportions.
#|     Deviations from HWE can indicate sample QC issues, or that
#|     there is population structure.
#|     Illustration inspired by @cooplab_PopulationGeneticsCourse_2011.
#|     [{{< fa rectangle-list >}}]({{< var recipes.base >}}#sec-recipe-slides-hwe)
include_graphics("assets/images/hwe-1000genomes.webp")
```

:::

::: {.notes }

So how do the HWE assumptions hold up in real data? The figure shows
three human populations from the 1000 genomes data, for which 10,000
SNPs have been selected. For each SNP, we know the genotype
frequencies $AA$, $Aa$, $aa$ and can therefore calculate the allele
frequencies for $A$ and $a$ using the equations on the preceding
slide (e.g., $p=p_{AA}+p_{Aa}/2$).

Note that 1000 genomes data have sorted genotype frequencies, so to
produce nice symmetrical plots, half the genotype frequencies have
been reversed (minor/het/major -> major/het/minor).

<!-- markdownlint-disable MD013 -->

See
<https://stackoverflow.com/questions/26587940/ggplot2-different-legend-symbols-for-points-and-lines>
for legend customization.

<!-- markdownlint-enable MD013 -->

:::

## The obsession of population genetics

Population genetics is about [@gillespie_PopulationGeneticsConcise_2004]

1. describing the genetic structure of populations
2. constructing theories on the forces that influence genetic
   variation

::: {.fragment}

:::: {.columns}

::: {.column width="50%"}

Questions to ponder:

- why does variation look the way it does?
- how is variation maintained?
- how does variation change over time ($\Delta p$)?
- what forces shape the genetic structure of *populations*?

:::

::: {.column width="50%" .fragment}

:::: {style="display: grid; grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
grid-row-gap: 0px; grid-column-gap: 0px; align-items: center;"}

:::{}

![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}

$$p=0.1$$

:::

:::{}

$$\large\rightarrow$$

:::

:::{}

![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}
![](assets/images/flower-white.svg){width=27}

$$p=0.5$$

:::

:::{}

$$\large\rightarrow$$

:::

:::{}

![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-red.svg){width=27}
![](assets/images/flower-white.svg){width=27}

$$p=0.9$$

:::

::::

:::

::::

:::

::: {.notes }

From HWE: we want to look a the creation, maintenance and loss of
variation and what forces affect it

Goal: describe theory behind evolving populations. This has been an
obsession for a long time, even before DNA was known. We need to get
back to basics.

:::

## Bibliography {.unnumbered .unlisted .smaller}

::: { #refs }
:::
