<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Linkage disequilibrium</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-dd75e976e0414ec343b1912506d158bc.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Linkage disequilibrium</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="a-first-look-at-haplotype-structure" class="slide level2 ultitle">
<h2>A first look at haplotype structure</h2>
<!-- markdownlint-disable MD013 -->
<div class="center">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-mcdonald-kreitman" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-mcdonald-kreitman-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-mcdonald-kreitman-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;1: Shared haplotypes in the Drosophila ADH gene [@kreitman_NucleotidePolymorphismAlcohol_1983]"><img data-src="linkage_files/figure-revealjs/fig-mcdonald-kreitman-1.svg" class="quarto-figure quarto-figure-center" style="width:120.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mcdonald-kreitman-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Shared haplotypes in the <em>Drosophila</em> ADH gene <span class="citation" data-cites="kreitman_NucleotidePolymorphismAlcohol_1983">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kreitman, 1983</a>)</span>
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<h4 id="observations">Observations</h4>
<ul>
<li>Particular combinations of alleles at different SNPs occur together</li>
<li>Non-random assortment of alleles is called <strong>linkage disequilibrium</strong> (<strong>LD</strong>)
<ul>
<li>Note: detecting <strong>LD</strong> does <strong>not</strong> ensure linkage or lack of equilibrium</li>
</ul></li>
<li>Unfortunate term: barrier to understanding <span class="citation" data-cites="slatkin_LinkageDisequilibriumUnderstanding_2008">(<a href="#/bibliography" role="doc-biblioref" onclick="">Slatkin, 2008</a>)</span></li>
</ul>
<div class="fragment">
<p>Going forward: discuss how <strong>linkage generates LD</strong> and <strong>recombination breaks it down</strong></p>
</div>
</section>
<section id="linkage-generates-haplotype-structure" class="slide level2 ultitle">
<h2>Linkage generates haplotype structure</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-genealogy" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-genealogy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-genealogy-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;2: A simple genealogy with mutations mapped onto branches"><img data-src="linkage_files/figure-revealjs/fig-genealogy-1.svg" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-genealogy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: A simple genealogy with mutations mapped onto branches
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-haplotypes" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-haplotypes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-haplotypes-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;3: Individual haplotypes and their mutations."><img data-src="linkage_files/figure-revealjs/fig-haplotypes-1.svg" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-haplotypes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Individual haplotypes and their mutations.
</figcaption>
</figure>
</div>
</div>
</div>
</div></div>
<div>
<ul>
<li>Mutations 1, 2 on same branch <span class="math inline">\Rightarrow</span> appear together</li>
<li>Mutations 1, 5 on derived branches <span class="math inline">\Rightarrow</span> <em>usually</em> appear together</li>
</ul>
</div>
<aside class="notes">
<p>Key assumption in genealogies: infinite sites model; a mutation happens once and once only at a given site</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linkage-disequilibrium-in-the-absence-of-recombination" class="slide level2 ultitle">
<h2>Linkage disequilibrium in the absence of recombination</h2>
<div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-ld-no-recomb" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-ld-no-recomb-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-ld-no-recomb-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;4: In the absence of recombination we can observe 2 or 3 possible haplotypes out of 4. Here AB is the ancestral haplotype."><img data-src="linkage_files/figure-revealjs/fig-ld-no-recomb-1.svg" style="width:100.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ld-no-recomb-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: In the absence of recombination we can observe 2 or 3 possible haplotypes out of 4. Here AB is the ancestral haplotype.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div>
<ul>
<li>we assume no homoplasy (back mutations)</li>
<li>haplotypes consistent with single tree (without recombination) form <strong>perfect phylogeny</strong></li>
</ul>
</div>
<aside class="notes">
<p>Note the implicit assumption of the infinite sites model, that is, we are only allowed to place each mutation <strong>once</strong> on the tree.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="recombination-breaks-association-between-loci" class="slide level2 smaller ultitle">
<h2>Recombination breaks association between loci</h2>
<div class="columns">
<div class="column" style="width:60%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../../slides/foundations/assets/images/meiosis.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img data-src="../../slides/foundations/assets/images/meiosis.webp" class="quarto-figure quarto-figure-center" style="width:80.0%"></a></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<div class="flushright smallr translatey50">
<p><span class="citation" data-cites="miller_HumanBiology_2020">Miller (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span>, Fig. 5.12.3</p>
</div>
<p>One (at least) <strong>crossover</strong> in meiosis I per chromosome! But: rates vary between loci (hotspots), sex chromosomes vs autosomes, and in some species, recombination only occurs in one sex (e.g., <em>D.melanogaster</em>).</p>
</div><div class="column fragment" style="width:40%;">
<h4 id="genetic-distance-and-recombination-rate">Genetic distance and recombination rate</h4>
<div class="center">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><a href="linkage_files/figure-revealjs/ld-model-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="linkage_files/figure-revealjs/ld-model-1.svg" style="width:100.0%"></a></p>
</figure>
</div>
</div>
</div>
</div>
<p><span class="math inline">d</span> - physical distance in bp</p>
<p>Definiton: <strong>genetic distance</strong> between two points is <span class="math inline">x</span> centiMorgans(cM) if the average number of crossovers between points <span class="math inline">x/100</span> per meiosis</p>
<p>For short distances, genetic distance in cM <span class="math inline">\approx</span> <span class="math inline">\mathrm{Pr}(\text{crossover})</span></p>
<p>Definition: <strong>recombination rate</strong> <span class="math inline">r</span> relates genetic distance to base pair distance; commonly measured in cM/Mb</p>
<p>Example: in human average is 1.2 cM per Mb - the probability of crossover is a 1.2%</p>
</div></div>
<aside class="notes">
<p>The mechanism that breaks up association between loci is recombination. Cartoon to right shows the possible (haploid) outcomes of meiosis. <span class="math inline">r</span> (the recombination rate) denotes the probability that there is a recombination event between the two loci.</p>
<p>The farther apart two loci on the same chromosome, the larger the probability that a recombination event occurs between them. Note that this can be applied to loci on different chromosomes as well; they segregate independently.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="recombination-generates-new-combinations-of-alleles" class="slide level2 ultitle">
<h2>Recombination generates new combinations of alleles</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-parent-gamete" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-parent-gamete-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-parent-gamete-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;5: Recombination mixes haplotypes"><img data-src="linkage_files/figure-revealjs/fig-parent-gamete-1.svg" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-parent-gamete-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Recombination mixes haplotypes
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<h4 id="rules-of-thumb-human">Rules of thumb (human)</h4>
<ul>
<li>For close SNPs (less than ~0.01-0.1 cM, or ~10-100Kb) linkage is stronger force than recombination</li>
<li>At larger (&gt;0.1cM) recombination is stronger</li>
</ul>
</div></div>
</section>
<section id="measuring-ld" class="slide level2 smaller ultitle">
<h2>Measuring LD</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-alleles-haplotypes" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-alleles-haplotypes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-alleles-haplotypes-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;6: Allele and haplotype frequencies at two SNPs. p denotes a frequency."><img data-src="linkage_files/figure-revealjs/fig-alleles-haplotypes-1.svg" style="width:70.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-alleles-haplotypes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Allele and haplotype frequencies at two SNPs. p denotes a frequency.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div>
<p>Given only information about SNP allele frequencies <span class="math inline">p_A</span> and <span class="math inline">p_B</span>, what would guess be for <span class="math inline">p_{AB}</span>?</p>
</div>
<div class="fragment">
<p>If independent, then <span class="math inline">p_{AB} = p_Ap_B</span>, else <span class="math inline">p_{AB} \neq p_Ap_B</span>. We measure the deviation <span class="math inline">D</span></p>
<p><span class="math display">
D = p_{AB} - p_Ap_B
</span></p>
<p>and say that there is <strong>linkage equilibrium</strong> if <span class="math inline">D=0</span>!</p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<h4 id="alternative-measures">Alternative measures</h4>
<p>Unfortunate property of <span class="math inline">D</span>: its magnitude depends on allele frequencies!</p>
</div>
<div class="fragment">
<p><span class="math display">
\begin{align*}
D^\prime = \frac{D}{D_\mathrm{max}} &amp;
    = \frac{D}{\min(p_Ap_b,p_ap_B)}, \quad\mathrm{for\  }D&gt;0\\
    &amp; = \frac{D}{\min(p_Ap_B,p_ap_b)}, \quad\mathrm{for\ }D&lt;0
\end{align*}
</span></p>
<p><span class="math inline">|D^\prime| &lt; 1</span> implies there must have been recombination</p>
</div>
<div class="fragment">
<p><span class="math display">
r^2 = \frac{D^2}{p_Ap_ap_Bp_b}
</span></p>
<ul>
<li><span class="math inline">r^2=1</span> perfect LD</li>
<li><span class="math inline">r^2</span> natural parameter for measuring contribution of LD to genetic associations</li>
</ul>
</div>
</div></div>
<aside class="notes">
<p>Alternatively put:</p>
<p><span class="math display">
\max{|D|} = \min(p(1-q), (1-p)q)
</span></p>
<p>Better use <span class="math inline">R^2</span> or <span class="math inline">D^\prime</span></p>
<p>Square of correlation coefficient between alleles at two loci:</p>
<p><span class="math display">
R^2 = \frac{D^2}{pq(1-p)(1-q)}
</span></p>
<p>Toy example to show how mixing subpopulations can introduce linkage: assume two subpopulations fixed for AB and ab. Then for either population <span class="math inline">D=p_{AB} - p_Ap_B=1-1=0</span>. Mixing the populations leads to <span class="math inline">p_{AB}=p_{ab}=p_A=p_A=0.5</span>, so <span class="math inline">D=0.5-0.5^2=0.25</span>, <span class="math inline">D_{max}=\min(p_Ap_B, p_ap_b)=0.25</span> and <span class="math inline">D^\prime=1</span>.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="ld-decay" class="slide level2 ultitle">
<h2>LD decay</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-plot-ld" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-plot-ld-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="linkage_files/figure-revealjs/fig-plot-ld-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;7: Decay of LD over time due to recombination, starting from D_0."><img data-src="linkage_files/figure-revealjs/fig-plot-ld-1.svg" style="width:100.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-plot-ld-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Decay of LD over time due to recombination, starting from <span class="math inline">D_0</span>.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="fragment">
<h4 id="observation">Observation</h4>
<p>Even for free recombination (<span class="math inline">r=0.5</span>) LD decay takes time. Comparison: HWE which takes one generation.</p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<!-- markdownlint-disable MD013 -->
<div id="fig-ld-human-mouse" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-ld-human-mouse-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../../slides/foundations/assets/images/ld_mice.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;8: LD between pairs of autosomal SNPs for human and mouse. From [@laurie_LinkageDisequilibriumWild_2007, Fig. 2]"><img data-src="../../slides/foundations/assets/images/ld_mice.png"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ld-human-mouse-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: LD between pairs of autosomal SNPs for human and mouse. From <span class="citation" data-cites="laurie_LinkageDisequilibriumWild_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Laurie et al., 2007, Figure 2</a>)</span>
</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
</div></div>
</section>
<section id="ld-importance-and-applications" class="slide level2 ultitle">
<h2>LD importance and applications</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h3 class="ultitle" id="importance">Importance</h3>
<ul>
<li>information about past events</li>
<li>constrains potential response to natural and artificial selection</li>
<li>genome-wide LD is reflection of
<ul>
<li>population history</li>
<li>breeding system</li>
<li>pattern of geographic subdivision</li>
</ul></li>
<li>regional LD reflection of
<ul>
<li>natural selection</li>
<li>gene conversion</li>
<li>mutation and more</li>
</ul></li>
</ul>
<p>Due to lack of data LD didn’t become important until the end of the 70’s</p>
</div><div class="column" style="width:50%;">
<h3 class="ultitle" id="applications">Applications</h3>
<h4 id="mutation-and-gene-mapping">Mutation and gene mapping</h4>
<p>Most loci are in close linkage with a variable site which can be used as a <em>marker</em> to study the inheritance of a trait of interest.</p>
<h4 id="linked-selection">Linked selection</h4>
<p>Reduced variation close to a site under selection due to linkage</p>
<h4 id="estimating-allele-age">Estimating allele age</h4>
<p>Strong LD in large region indication of young allele.</p>
<h4 id="determining-window-size-for-genome-scans">Determining window size for genome scans</h4>
<p>Ideal window size ~ distance at which LD between markers approaches background levels</p>
</div><div class="citation flushright">
<p><span class="citation" data-cites="slatkin_LinkageDisequilibriumUnderstanding_2008">(<a href="#/bibliography" role="doc-biblioref" onclick="">Slatkin, 2008</a>)</span></p>
</div>
</div>
</section>
<section id="bibliography" class="slide level2 ultitle smaller scrollable">
<h2>Bibliography</h2>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-kreitman_NucleotidePolymorphismAlcohol_1983" class="csl-entry" role="listitem">
Kreitman, M. (1983). Nucleotide polymorphism at the alcohol dehydrogenase locus of <span>Drosophila</span> melanogaster. <em>Nature</em>, <em>304</em>(5925), 412. <a href="https://doi.org/10.1038/304412a0">https://doi.org/10.1038/304412a0</a>
</div>
<div id="ref-laurie_LinkageDisequilibriumWild_2007" class="csl-entry" role="listitem">
Laurie, C. C., Nickerson, D. A., Anderson, A. D., Weir, B. S., Livingston, R. J., Dean, M. D., Smith, K. L., Schadt, E. E., &amp; Nachman, M. W. (2007). Linkage <span>Disequilibrium</span> in <span>Wild Mice</span>. <em>PLOS Genetics</em>, <em>3</em>(8), e144. <a href="https://doi.org/10.1371/journal.pgen.0030144">https://doi.org/10.1371/journal.pgen.0030144</a>
</div>
<div id="ref-miller_HumanBiology_2020" class="csl-entry" role="listitem">
Miller, C. (2020). <em>Human <span>Biology</span></em>. <span>Thompson Rivers University</span>.
</div>
<div id="ref-pritchard_OwnerGuideHuman_" class="csl-entry" role="listitem">
Pritchard, J. K. (n.d.). <em>An <span>Owner</span>’s <span>Guide</span> to the <span>Human Genome</span></em>. Retrieved August 18, 2025, from <a href="https://web.stanford.edu/group/pritchardlab/HGbook.html">https://web.stanford.edu/group/pritchardlab/HGbook.html</a>
</div>
<div id="ref-slatkin_LinkageDisequilibriumUnderstanding_2008" class="csl-entry" role="listitem">
Slatkin, M. (2008). Linkage disequilibrium — understanding the evolutionary past and mapping the medical future. <em>Nature Reviews Genetics</em>, <em>9</em>(6), 477–485. <a href="https://doi.org/10.1038/nrg2361">https://doi.org/10.1038/nrg2361</a>
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Linkage disequilibrium</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>