---
title: Linkage disequilibrium
description: >-
  Basic introduction to linkage disequilibrium and recombination.
  Based on @pritchard_OwnerGuideHuman_
author:
  - Per Unneberg
format:
  nbis-quarto-revealjs:
    footer: Linkage disequilibrium
---

## Setup {visibility="hidden" .hidden .unnumbered .unlisted}

{{< include /slides/_knitr.qmd >}}

{{< include /slides/_rlibs.qmd >}}

```{r }
#| label: custom-libs
#| echo: false
#| eval: true
library(data.table)
```

## A first look at haplotype structure

<!-- markdownlint-disable MD013 -->

:::{.center}

```{r, engine="tikz", fig.ext="svg" }
#| label: fig-mcdonald-kreitman
#| echo: false
#| eval: true
#| out-width: 120%
#| fig-align: center
#| fig-cap: Shared haplotypes in the *Drosophila* ADH gene [@kreitman_NucleotidePolymorphismAlcohol_1983]
\begin{tikzpicture}
\tikzset {
  every node/.style={anchor=base,font=\bfseries\sffamily\Huge, minimum size=1.5em},
  table/.style = {
    matrix of nodes,
    text height=1.5em,
    node distance=0pt,
    column sep=2pt,
    row sep=10pt,
  },
  ld/.style = {
    color=red,
    fill=red!5,
    outer sep=0,
  },
}

\matrix (m) [table] {
  C & C & G & C & A & A & T & A & |[ld]| A & |[ld]| T & G & G & C & G & C & T & A & C & |[ld]| T & |[ld]| T & C & |[ld]| A & |[ld]| C & |[ld]| A & A & |[ld]| T & |[ld]| A & |[ld]| A & |[ld]| C & C & C & A & C & T & A & G & A & C & A & G & C & C & T \\
  %%
  C & C & |[ld]| C & C & A & A & T & A & T & G & G & G & C & G & C & T & A & C & |[ld]| T & |[ld]| T & C & |[ld]| A & |[ld]| C & |[ld]| A & A & |[ld]| T & |[ld]| A & |[ld]| A & |[ld]| C & C & C & A & C & T & A & G & A & C & A & G & C & C & T \\
  %%
  C & C & G & C & A & A & T & A & T & G & G & G & C & G & C & T & A & C & C & C & C & C & G & G & A & A & T & C & T & C & C & A & C & T & A & |[ld]| A & A & C & A & G & |[ld]| T & C & |[ld]| A \\
  %%
  C & C & G & C & A & A & T & A & T & G & G & G & C & G & C & T & |[ld]| G & |[ld]| T & C & C & C & C & G & G & A & A & T & C & T & C & C & A & C & T & A & |[ld]| A & A & C & |[ld]| T & |[ld]| A & C & C & T \\
  %%
  C & C & G & |[ld]| A & |[ld]| G & A & T & A & |[ld]| A & G & |[ld]| T & |[ld]| C & C & G & |[ld]| A & |[ld]| G & |[ld]| G & |[ld]| T & C & C & C & C & G & G & A & A & T & C & T & C & C & A & C & T & A & G & |[ld]| C & C & A & G & C & C & T \\
  %%
  C & C & |[ld]| C & C & A & A & T & A & T & G & G & G & C & G & C & |[ld]| G & A & C & C & C & C & C & G & G & A & A & T & C & T & C & |[ld]| T & A & |[ld]| T & T & |[ld]| C & |[ld]| A & |[ld]| C & C & A & G & C & |[ld]| T & T \\
  %%
  C & C & |[ld]| C & C & A & A & T & A & T & G & G & G & C & G & C & |[ld]| G & A & C & C & C & C & C & G & G & A & A & T & C & T & |[ld]| G & |[ld]| T & |[ld]| C & |[ld]| T & |[ld]| C & |[ld]| C & G & |[ld]| C & C & A & G & C & C & T \\
  %%
  |[ld]| T & |[ld]| G & |[ld]| C & |[ld]| A & |[ld]| G & A & T & A & A & G & |[ld]| T & |[ld]| C & |[ld]| G & G & C & |[ld]| G & A & C & C & C & C & C & G & G & A & A & T & C & T & |[ld]| G & |[ld]| T & |[ld]| C & |[ld]| T & |[ld]| C & |[ld]| C & G & |[ld]| C & |[ld]| G & A & G & C & C & T \\
  %%
  |[ld]| T & |[ld]| G & |[ld]| C & |[ld]| A & |[ld]| G & A & T & A & A & G & |[ld]| T & |[ld]| C & |[ld]| G & G & C & |[ld]| G & A & C & C & C & C & C & G & G & A & A & T & C & T & |[ld]| G & |[ld]| T & |[ld]| C & |[ld]| T & |[ld]| C & |[ld]| C & G & |[ld]| C & |[ld]| G & A & G & C & C & T \\
  %%
  |[ld]| T & |[ld]| G & |[ld]| C & |[ld]| A & |[ld]| G & A & T & A & A & G & |[ld]| T & |[ld]| C & |[ld]| G & G & C & |[ld]| G & A & C & C & C & C & C & G & G & A & A & T & C & T & |[ld]| G & |[ld]| T & |[ld]| C & |[ld]| T & |[ld]| C & |[ld]| C & G & |[ld]| C & |[ld]| G & A & G & C & C & T \\
  %%
  |[ld]| T & |[ld]| G & |[ld]| C & |[ld]| A & |[ld]| G & |[ld]| G & |[ld]| G & |[ld]| G & |[ld]| A & G & G & G & C & |[ld]| T & C & |[ld]| G & A & C & C & C & |[ld]| A & C & G & G & |[ld]| G & A & T & C & T & |[ld]| G & |[ld]| T & |[ld]| C & |[ld]| T & |[ld]| C & |[ld]| C & G & |[ld]| C & C & A & G & C & C & T \\
};
%%\scoped[on background layer] {
  \node[fit={(m-8-1.north west) (m-11-5.south east)},
        ultra thick, draw=red]{};
  \node[fit={(m-8-11.north west) (m-10-13.south east)},
        ultra thick, draw=red]{};
  \node[fit={(m-7-30.north west) (m-11-35.south east)},
        ultra thick, draw=red]{};
  \node[fit={(m-1-19.north west) (m-2-29.south east)},
        ultra thick, draw=red]{};
%%}
\end{tikzpicture}
```

:::

<!-- markdownlint-enable MD013 -->

#### Observations

- Particular combinations of alleles at different SNPs occur together
- Non-random assortment of alleles is called **linkage
  disequilibrium** (**LD**)

## Linkage generates haplotype structure

:::: {.columns}

::: {.column width="50%"}

```{r, engine='tikz', fig.ext="svg" }
#| label: fig-genealogy
#| echo: false
#| eval: true
#| fig-cap: |
#|   A simple genealogy with mutations mapped onto branches
#| out-width: "70%"
\begin{tikzpicture}[ultra thick, x=1cm, y=1cm, font=\Large\sffamily]
\tikzset{
  every node/.style={inner sep=0pt, outer sep=0pt},
  label/.style={inner sep=0pt, outer sep=3pt},
  mut/.style={circle, fill=red!70, minimum size=8pt},
  mlab/.style={text=red!70, node distance=8pt}
}
\def\samples{A,...,I}
\foreach \x [count=\xi] in \samples {\node[label] (\x) at (\xi,0) {\x};}
\def\coal{E/F/1, B/C/1.5, EF/G/1.65, BC/D/2, EFG/H/2.2, A/BCD/2.7,
  EFGH/I/3.4, ABCD/EFGHI/5.7}
\foreach \left/ \right / \time [count=\xi] in \coal
{
  \node (\left\right) at ($(A -| \left) !.5! (A -| \right) + (0, \time)$) {};
  \draw (\left) |- (\left\right) -| (\right);
}
\def\mut{ABCD/2,ABCD/1.2,EFGH/0.9,A/1.2,BCD/0.2,EFG/0.2,H/1.2}
\foreach \m/\time [count=\mi] in \mut
{
  \node[mut, above of=\m, node distance=\time cm] (m\mi) {};
  \node[mlab, left of=m\mi] {\mi};
}
\node[above of=ABCDEFGHI] (root) {};
\draw (ABCDEFGHI) -- (root);
\end{tikzpicture}
```

:::

::: {.column width="50%"}

```{r, engine='tikz', fig.ext="svg" }
#| label: fig-haplotypes
#| echo: false
#| eval: true
#| fig-cap: |
#|    Individual haplotypes and their mutations.
#| out-width: 70%
\begin{tikzpicture}[ultra thick, x=1cm, font=\Large\sffamily]
\tikzset{
  mut/.style={circle, fill=red!70, minimum size=8pt},
  mlab/.style={text=red!70}
}

\def\samples{I,...,A}
\foreach \x [count=\xi] in \samples {
  \node (\x1) at (0,\xi) {};
  \node (\x2) at (10,\xi) {};
  \draw (\x1) -- (\x2);
  \node[left of=\x1] {\x};
}
\def\mutpos{1,4.2,8,2,2.8,6,8.8}
\foreach \p [count=\pi] in \mutpos { \node[mlab] (m\pi) at (\p, 10) {\pi}; }
\def\mut{%%
  A/1,A/4,A/2,B/1,B/5,B/2,%%
  C/1,C/5,C/2,D/1,D/5,D/2,%%
  E/6,E/3,F/6,F/3,G/6,G/3,%%
  H/3,H/7,I/3}
\foreach \sample/\id in \mut {
  \node[mut] at ($(\sample1 -| m\id)$) {};
}
\end{tikzpicture}
```

:::

::::

:::{}

- Mutations 1, 2 on same branch $\Rightarrow$ appear together
- Mutations 1, 5 on derived branches $\Rightarrow$ *usually* appear
  together

:::

## Linkage disequilibrium in the absence of recombination

:::{}

```{r, engine='tikz', fig.ext="svg" }
#| label: fig-ld-no-recomb
#| echo: false
#| eval: true
#| fig-cap: |
#|    In the absence of recombination we can observe 2 or 3 possible
#|    haplotypes out of 4. Here AB is the ancestral haplotype.
#| out-width: 100%
\begin{tikzpicture}[ultra thick, x=1cm, font=\Large\sffamily]
\tikzset{
  every node/.style={inner sep=0pt, outer sep=0pt},
  label/.style={text=blue!70,inner sep=0pt, outer sep=3pt},
  mut/.style={circle, fill=blue!70, minimum size=8pt},
  mlab/.style={text=blue!70, anchor=east, outer sep=0pt}
}
\newcommand{\Aa}{$\mathsf{A\rightarrow a}$}
\newcommand{\Bb}{$\mathsf{B\rightarrow b}$}

\begin{scope}
\def\haplotypes{A/AB,B/ab,C/ab,D/ab}
\foreach \hap/\x [count=\xi] in \haplotypes {\node[label] (\hap) at (\xi,0) {\x};}
\def\coal{C/D/1,B/CD/2.2,A/BCD/4}
\foreach \left/\right/\time [count=\i] in \coal
{
  \node (\left\right) at ($(A -| \left) !.5! (A -| \right) + (0, \time)$) {};
  \draw (\left) |- (\left\right) -| (\right);
}
\node[above of=ABCD] (root) {};
\draw (ABCD) -- (root);
\node[mut, above of=BCD, node distance=1.2 cm] (m1) {};
\node[mlab, right of=m1] {\Aa};
\node[mut, above of=BCD, node distance=.6 cm] (m2) {};
\node[mlab, right of=m2] {\Bb};
\end{scope}

\begin{scope}[xshift=5cm]
\def\haplotypes{A/aB,B/Ab,C/Ab,D/Ab}
\foreach \hap/\x [count=\xi] in \haplotypes {\node[label] (\hap) at (\xi,0) {\x};}
\def\coal{C/D/1,B/CD/2.3,A/BCD/4}
\foreach \left/\right/\time [count=\i] in \coal
{
  \node (\left\right) at ($(A -| \left) !.5! (A -| \right) + (0, \time)$) {};
  \draw (\left) |- (\left\right) -| (\right);
}
\node[above of=ABCD] (root) {};
\draw (ABCD) -- (root);
\node[mut, above of=A, node distance=3.4 cm] (m1) {};
\node[mlab, right of=m1] {\Aa};
\node[mut, above of=BCD, node distance=.6 cm] (m2) {};
\node[mlab, right of=m2] {\Bb};
\end{scope}

\begin{scope}[xshift=10cm]
\def\haplotypes{A/AB,B/aB,C/ab,D/ab}
\foreach \hap/\x [count=\xi] in \haplotypes {\node[label] (\hap) at (\xi,0) {\x};}
\def\coal{C/D/1,B/CD/2.3,A/BCD/4}
\foreach \left/\right/\time [count=\i] in \coal
{
  \node (\left\right) at ($(A -| \left) !.5! (A -| \right) + (0, \time)$) {};
  \draw (\left) |- (\left\right) -| (\right);
}
\node[above of=ABCD] (root) {};
\draw (ABCD) -- (root);
\node[mut, above of=BCD, node distance=1.2 cm] (m1) {};
\node[mlab, right of=m1] {\Aa};
\node[mut, above of=CD, node distance=.6 cm] (m2) {};
\node[mlab, right of=m2] {\Bb};
\end{scope}

\begin{scope}[xshift=15cm]
\def\haplotypes{A/aB,B/AB,C/AB,D/Ab}
\foreach \hap/\x [count=\xi] in \haplotypes {\node[label] (\hap) at (\xi,0) {\x};}
\def\coal{C/D/1,B/CD/2.3,A/BCD/4}
\foreach \left/\right/\time [count=\i] in \coal
{
  \node (\left\right) at ($(A -| \left) !.5! (A -| \right) + (0, \time)$) {};
  \draw (\left) |- (\left\right) -| (\right);
}
\node[above of=ABCD] (root) {};
\draw (ABCD) -- (root);
\node[mut, above of=A, node distance=3.4 cm] (m1) {};
\node[mlab, right of=m1] {\Aa};
\node[mut, above of=D, node distance=.6 cm] (m2) {};
\node[mlab, right of=m2] {\Bb};
\end{scope}

\end{tikzpicture}
```

:::

:::{}

- we assume no homoplasy (back mutations)
- haplotypes consistent with single tree (without recombination) form
  **perfect phylogeny**

:::

## Recombination breaks association between loci {.smaller}

:::: {.columns}

::: {.column width="60%"}

<!-- markdownlint-disable MD013 -->

![](/slides/foundations/assets/images/meiosis.webp){width=80% fig-align=center}

<!-- markdownlint-enable MD013 -->

::: {.flushright .smallr .translatey50}

@miller_HumanBiology_2020, Fig. 5.12.3

:::

Once per chromosome! But: rates vary between loci (hotspots), sex
chromosomes vs autosomes, and in some species, recombination only
occurs in one sex (e.g., *D.melanogaster*).

:::

::: {.column width="40%"}

:::{}

```{r, engine='tikz', fig.ext="svg" }
#| label: fig-parent-gamete
#| echo: false
#| eval: true
#| fig-cap: Recombination mixes haplotypes
#| out-width: 70%
\begin{tikzpicture}[ultra thick, font=\sffamily, x=1cm]
\tikzset{
  r70/.style={red!70, text=red!70},
  b70/.style={blue!70, text=blue!70},
  mut/.style={circle, minimum size=2pt},
  bmut/.style={mut, b70, fill=blue!70},
  rmut/.style={mut, fill=red!70},
  ul/.style={inner sep=0pt, outer sep=5pt},
  bl/.style={inner sep=0pt, outer sep=5pt, yshift=-1em, anchor=mid},
}

\begin{scope}
\draw[b70] (0,1) -- (6,1);
\foreach \x [count=\xi] in {A,...,E} {
  \node[bmut, label={[b70, ul] \x}] at (\xi, 1) {};
}
\draw[r70] (0,0) -- (6,0);
\foreach \x [count=\xi] in {a,...,e} {
  \node[rmut, label={[r70, bl]below: \x}] at (\xi,0) {};
}
\node (upper) at (3, -1) {};
\end{scope}

\begin{scope}[yshift=-5cm]
\node (lower) at (3, 2) {};
\draw[b70] (0,1) -- (2.5,1);
\draw[b70] (2.5,0) -- (6,0);
\draw[r70] (0,0) -- (2.5,0);
\draw[r70] (2.5,1) -- (6,1);
\foreach \x/\m/\l [count=\xi] in {
  A/bmut/b70,B/bmut/b70,c/rmut/r70,d/rmut/r70,e/rmut/r70%%
} {
  \node[\m, label={[ul, \l] \x}] at (\xi,1) {};
}
\foreach \x/\m/\l [count=\xi] in {
  a/rmut/r70,b/rmut/r70,C/bmut/b70,D/bmut/b70,E/bmut/b70%%
} {
  \node[\m, label={[bl, \l]below: \x}] at (\xi,0) {};
}
\end{scope}

\draw[->, >=latex] (upper) -- (lower);

\end{tikzpicture}
```

:::

:::

::::

::: {.notes}

The mechanism that breaks up association between loci is
recombination. Cartoon to right shows the possible (haploid) outcomes
of meiosis. $r$ (the recombination rate) denotes the probability that
there is a recombination event between the two loci.

The farther apart two loci on the same chromosome, the larger the
probability that a recombination event occurs between them. Note that
this can be applied to loci on different chromosomes as well; they
segregate independently.

:::

## Measuring LD{.smaller}

:::: {.columns}

::: {.column width="50%"}

:::{}

```{r, engine='tikz', fig.ext="svg" }
#| label: fig-alleles-haplotypes
#| echo: false
#| eval: true
#| fig-cap: |
#|   Allele and haplotype frequencies at two SNPs. p denotes a
#|   frequency.
#| out-width: 70%
\begin{tikzpicture}[ultra thick, x=1cm, font=\Large\sffamily]
\tikzset{
  bl/.style={blue!70, text=blue!70},
  mut/.style={circle, minimum size=2pt, fill=blue!70},
}
\newcommand{\pA}{$\mathsf{p_A}$}
\newcommand{\pa}{$\mathsf{p_a}$}
\newcommand{\pB}{$\mathsf{p_B}$}
\newcommand{\pb}{$\mathsf{p_b}$}
\newcommand{\pAB}{$\mathsf{p_{AB}}$}
\newcommand{\pAb}{$\mathsf{p_{Ab}}$}
\newcommand{\paB}{$\mathsf{p_{aB}}$}
\newcommand{\pab}{$\mathsf{p_{ab}}$}

\begin{scope}
\node[text=black] at (2, 5.5) {SNPs};
\foreach \l/\x/\pos [count=\xi] in {b/\pb/2, B/\pB/2, a/\pa/1, A/\pA/1} {
  \draw[bl] (0, \xi) -- (3, \xi);
  \node[mut, label={[bl] \l}] at (\pos, \xi) {};
  \node[bl] at (4, \xi) {\x};
}
\end{scope}

\begin{scope}[xshift=8cm]
\node[text=black] at (2, 5.5) {Haplotypes};
\foreach \a/\b/\l [count=\xi] in {a/b/\pab, a/B/\paB, A/b/\pAb, A/B/\pAB} {
  \draw[bl] (0, \xi) -- (3, \xi);
  \node[mut, label={[bl] \a}] at (1, \xi) {};
  \node[mut, label={[bl] \b}] at (2, \xi) {};
  \node[bl] at (4, \xi) {\l};
}
\end{scope}

\end{tikzpicture}
```

:::

:::{}

Given only information about SNP allele frequencies $p_A$ and $p_B$,
what would guess be for $p_{AB}$?

:::

:::{.fragment}

If independent, then $p_{AB} = p_Ap_B$, else $p_{AB} \neq p_Ap_B$. We
measure the deviation $D$

$$
D = p_{AB} - p_Ap_B
$$

and say that there is **linkage equilibrium** if $D=0$!

:::

:::

::: {.column width="50%"}

::: {.fragment}

#### Alternative measures

Unfortunate property of $D$: its magnitude depends on allele
frequencies!

:::

::: {.fragment}

$$
\begin{align*}
D^\prime = \frac{D}{D_\mathrm{max}} &
    = \frac{D}{\min(p_Ap_b,p_ap_B)}, \quad\mathrm{for\  }D>0\\
    & = \frac{D}{\min(p_Ap_B,p_ap_b)}, \quad\mathrm{for\ }D<0
\end{align*}
$$

$|D^\prime| < 1$ implies there must have been recombination

:::

::: {.fragment}

$$
r^2 = \frac{D^2}{p_Ap_ap_Bp_b}
$$

- $r^2=1$ perfect LD
- $r^2$ natural parameter for measuring contribution of LD to genetic
  associations

:::

:::

::::

## LD decay

:::: {.columns}

::: {.column width="50%"}

```{r }
#| label: fig-plot-ld
#| echo: false
#| eval: true
#| fig-height: 6
#| fig-width: 8
#| fig-cap: |
#|   Decay of LD over time due to recombination, starting from $D_0$.
#| out-width: 100%
decay <- function(r, d0, x) {
  d <- (1 - r)**x * d0
  d
}

x <- seq(0, 100, 1)

df <- data.frame(
    cbind(x, decay(0.01, 1, x), decay(0.1, 1, x), decay(0.5, 1, x))
)
colnames(df) <- c("x", "r=0.01", "r=0.1", "r=0.5")
df <- melt(data.table(df), id.vars = "x")
ggplot(df, aes(x = x, y = value, color = variable)) +
  geom_line() +
  labs(title = "LD decay",
       x = "Time (generations)",
       y = "Linkage disequilibrium (D)") +
  scale_y_continuous(breaks=c(0.0, 1.0), labels=c("0", TeX("$D_0$")))
```

::: {.fragment}

#### Observation

Even for free recombination ($r=0.5$) LD decay takes time

:::

:::

::: {.column width="50%"}

:::{.fragment}

<!-- markdownlint-disable MD013 -->

![LD between pairs of autosomal SNPs for human and mouse. From
[@laurie_LinkageDisequilibriumWild_2007, Fig.
2]](/slides/foundations/assets/images/ld_mice.png){#fig-ld-human-mouse}

<!-- markdownlint-enable MD013 -->

:::

:::

::::

## Bibliography
