<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/all.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/latex-fontsize.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – Selection</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-dd75e976e0414ec343b1912506d158bc.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Selection</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section>
<section id="selection" class="title-slide slide level1 center">
<h1>Selection</h1>
<div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="assets/images/gould_finches.webp" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="assets/images/gould_finches.webp" class="quarto-figure quarto-figure-center" width="500"></a></p>
</figure>
</div>
</div>
<aside class="notes">
<p>For a list of methods see <a href="https://methodspopgen.com/methods-to-detect-selection/" class="uri">https://methodspopgen.com/methods-to-detect-selection/</a></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="selection-and-fitness" class="slide level2 smaller ultitle">
<h2>Selection and fitness</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-selection-model-tikz" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selection-model-tikz-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selection-model-tikz-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;1: The life cycle used in the fundamental model of selection [@gillespie_PopulationGeneticsConcise_2004, Fig. 3.2]"><img data-src="selection_files/figure-revealjs/fig-selection-model-tikz-1.svg" class="quarto-figure quarto-figure-center" width="700"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selection-model-tikz-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: The life cycle used in the fundamental model of selection <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004, Figure 3.2</a>)</span>
</figcaption>
</figure>
</div>
</div>
</div>
<div class="fragment">
<blockquote>
<p>Much confusion exists in the literature regarding how various types of selection are defined, in particular because some of the terminology is used slightly differently within different scientific communities <span class="citation" data-cites="nielsen_MolecularSignaturesNatural_2005">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen, 2005</a>)</span></p>
</blockquote>
</div>
<div class="columns">
<div class="column fragment" style="width:50%;">
<p><span class="math display">
\begin{matrix}
\mathrm{Genotype} &amp; AA &amp; Aa &amp; aa \\
\mathrm{Frequency\ in\ newborns} &amp; p^2 &amp; 2pq &amp; q^2\\
\mathrm{Viability} &amp; w_{AA} &amp; w_{Aa} &amp; w_{aa}\\
\mathrm{Frequency\ after\ selection} &amp; p^2w_{AA} / \bar{w} &amp;
    2pqw_{Aa} / \bar{w} &amp; q^2w_{aa} / \bar{w} \\
\mathrm{Relative\ fitness} &amp; 1 &amp; 1-hs &amp; 1-s\\
\end{matrix}
</span></p>
<p>where <span class="math inline">\bar{w} = p^2w_{AA} + 2pqw_{Aa} + q^2w_{aa}</span> is the <strong>mean fitness</strong>.</p>
</div><div class="column fragment" style="width:50%;">
<table class="caption-top">
<tbody>
<tr class="odd">
<td><span class="math inline">h=0</span></td>
<td><span class="math inline">A</span> dominant, <span class="math inline">a</span> recessive</td>
</tr>
<tr class="even">
<td><span class="math inline">h=1</span></td>
<td><span class="math inline">a</span> dominant, <span class="math inline">A</span> recessive</td>
</tr>
<tr class="odd">
<td><span class="math inline">0&lt;h&lt;1</span></td>
<td>incomplete dominance</td>
</tr>
<tr class="even">
<td><span class="math inline">h&lt;0</span></td>
<td>overdominance (heterozygote advantage)</td>
</tr>
<tr class="odd">
<td><span class="math inline">h&gt;1</span></td>
<td>underdominance</td>
</tr>
</tbody>
</table>
<div class="flushright smallr">
<p>Notation follows <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">Gillespie (<a href="#/bibliography" role="doc-biblioref" onclick="">2004</a>)</span>, pp.&nbsp;61–64.</p>
</div>
</div></div>
<aside class="notes">
<p>The life cycle figure starts with newborns carrying the <span class="math inline">A</span> allele at frequency <span class="math inline">p</span>, like parents. To reproduce they must reach adulthood, but due to selection, the allele frequency of <span class="math inline">A</span> may shift to <span class="math inline">p^\prime</span> because different genotypes affect survival probabilities. This is typically summarized in genotype <em>viabilities</em>.</p>
<p>The Nielsen quote is a reminder that many different notations and parametrizations exist in the literature.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-most-important-equation-in-population-genetics" class="slide level2 ultitle">
<h2>The most important equation in population genetics</h2>
<p><span class="math display">
p^\prime - p = \Delta_sp = \frac{pq[p(w_{AA} - w_{Aa}) + q(w_{Aa} -
w_{aa})]}{p^2w_{AA} + 2pqw_{Aa} + q^2w_{aa}}
</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-selection-fig-p" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selection-fig-p-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selection-fig-p-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;2: Allele frequency change over time for directional, balancing, and disruptive selection, for different values of p_0."><img data-src="selection_files/figure-revealjs/fig-selection-fig-p-1.svg" class="quarto-figure quarto-figure-center" width="400" height="400"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selection-fig-p-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Allele frequency change over time for directional, balancing, and disruptive selection, for different values of <span class="math inline">p_0</span>.
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-selection-fig-dp" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selection-fig-dp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selection-fig-dp-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;3: Rate of allele frequency change as a function of allele frequency for directional, balancing, and disruptive selection."><img data-src="selection_files/figure-revealjs/fig-selection-fig-dp-1.svg" class="quarto-figure quarto-figure-center" width="400" height="400"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selection-fig-dp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Rate of allele frequency change as a function of allele frequency for directional, balancing, and disruptive selection.
</figcaption>
</figure>
</div>
</div>
</div>
</div></div>
<aside class="notes">
<p>Given the genotype fitnesses/viabilities, we can work out the difference in allele frequency <span class="math inline">p-p^\prime</span> between successive generations as a function of viabilities (or relative fitnesses). Equation relates allele frequency change to viabilities for different genotypes. It is instructive to study the equation and plot trajectories to gain an intuition for how allele frequencies evolve given the different selection regimes.</p>
<p>The left panel shows that the three modes of selection have different equilibrium points: for directional selection the favored homozygote will eventually attain 100%, balancing selection has a stable equilibrium point (attractor) for a given allelic ratio (here 25:75), and disruptive selection has an <em>unstable</em> equilibrium point (repeller) for a given allelic ratio.</p>
<p>The right panel shows the rate of <em>change</em> as a function of allele frequency. Note how the directionality changes at the equilibrium point for balancing and directional selection.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="selection-and-drift---population-size-matters" class="slide level2 ultitle">
<h2>Selection and drift - population size matters</h2>
<div class="columns">
<div class="column" style="width:80%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-selection-and-drift" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selection-and-drift-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selection-and-drift-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;4: The fixation probability relative to the neutral probability of fixation (p=1/2N) under the assumption s<0.1. Red highlights region where |N_es|<0.05. Adapted from @lynch_OriginsGenomeArchitecture_2007, Fig. 4.2."><img data-src="selection_files/figure-revealjs/fig-selection-and-drift-1.svg" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selection-and-drift-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: The fixation probability relative to the neutral probability of fixation (<span class="math inline">p=1/2N</span>) under the assumption <span class="math inline">s&lt;0.1</span>. Red highlights region where <span class="math inline">|N_es|&lt;0.05</span>. Adapted from <span class="citation" data-cites="lynch_OriginsGenomeArchitecture_2007">Lynch (<a href="#/bibliography" role="doc-biblioref" onclick="">2007</a>)</span>, Fig. 4.2.
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:20%;">
<p>In red region (<span class="math inline">|N_es|&lt;0.05</span>) the probability of fixation is within 10% of neutral fixation.</p>
<p><strong>Consequence</strong>: for any population size there exists range of selection coefficients where mutant alleles <span class="math inline">\approx</span> neutral (<strong>effective neutrality</strong>).</p>
</div></div>
<aside class="notes">
<p>Segue: note on the strength of s and the relation to <span class="math inline">N_e</span>. For small <span class="math inline">N_e</span> the relation will hold for higher <span class="math inline">s</span>, meaning small populations will have difficulties purging deleterious mutations.</p>
<p>Even if a new mutation has, say, <span class="math inline">s=0.01</span>, it still needs to overcome effects of drift. Only has 2% change of fixation.</p>
<p>In figure, we assume <span class="math inline">s&lt;0.1</span>, which is almost always the case <span class="citation" data-cites="lynch_OriginsGenomeArchitecture_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lynch, 2007, p. 74</a>)</span>.</p>
<p><span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 133</a>)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="direct-selection-can-be-inferred-from-protein-substitutions" class="slide level2 smaller ultitle">
<h2>Direct selection can be inferred from protein substitutions</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>For genes, the ratio of nonsynonymous to synonymous substitutions can tell us about protein evolution:</p>
<div class="columns">
<div class="column" style="width:50%;">
<pre><code>Synonymous substitution

Protein          L
DNA         --- CTT ---
                  *
DNA         --- CTC ---
Protein          L</code></pre>
</div><div class="column" style="width:50%;">
<pre><code>Nonsynonymous substitution

Protein          L
DNA         --- CTT ---
                 *
DNA         --- CHT ---
Protein          H</code></pre>
</div></div>
<div>
<dl>
<dt><strong><span class="math inline">\mathbf{d_N/d_S &lt;&lt; 1}</span></strong></dt>
<dd>
negative (purifying) selection
</dd>
<dt><strong><span class="math inline">\mathbf{d_N/d_S &lt; 1}</span></strong></dt>
<dd>
majority nonsynonymous deleterious, some advantageous
</dd>
<dt><strong><span class="math inline">\mathbf{d_N/d_S = 1}</span></strong></dt>
<dd>
neutral or mix neutral / advantageous / deleterious mutations
</dd>
<dt><strong><span class="math inline">\mathbf{d_N/d_S &gt; 1}</span></strong></dt>
<dd>
positive selection
</dd>
</dl>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-ensembl-hsapiens-rat-dnds" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-ensembl-hsapiens-rat-dnds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-ensembl-hsapiens-rat-dnds-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;5: d_n/d_s comparisons for human-rat orthologs. For most genes, d_n/d_s << 1 indicating purifying selection. A handful of genes (n=9) have d_n/d_s > 1.0 which could indicate positive selection. "><img data-src="selection_files/figure-revealjs/fig-ensembl-hsapiens-rat-dnds-1.svg" width="768" height="400"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ensembl-hsapiens-rat-dnds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: <span class="math inline">d_n/d_s</span> comparisons for human-rat orthologs. For most genes, <span class="math inline">d_n/d_s &lt;&lt; 1</span> indicating purifying selection. A handful of genes (n=9) have <span class="math inline">d_n/d_s &gt; 1.0</span> which could indicate positive selection. <a href="../../recipes/index.html#sec-recipe-slides-dn-ds-ratio-human-rat"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a>
</figcaption>
</figure>
</div>
</div>
</div>
</div></div>
<div class="fragment">
<div class="translatey50">
<p>Not all mutations fall in genes. Methods for detecting direct selection not applicable to studying selection on <em>single</em> mutation, or e.g., balancing. This requires looking for specific patterns of diversity surrounding locus under selection.</p>
</div>
</div>
<aside class="notes">
<p>Segue: how do we detect selection?</p>
<p>Early approach: look at protein-coding sequence where nucleotide substitutions can change the amino acid (non-synonymous, and presumably non-neutral) or be silent (synonymous, assumed neutral).</p>
<p>Not all mutations in genes. Also want to look at single advantageous mutations, or balanced polymorphisms. Must therefore look at tell-tale signs of linked neutral variants.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linked-selection-reduces-diversity-at-neighbour-loci" class="slide level2 ultitle">
<h2>Linked selection reduces diversity at neighbour loci</h2>
<div class="cell">
<div class="cell-output-display">
<div id="fig-selective-sweep" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selective-sweep-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selective-sweep-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;6: A selective sweep of an advantageous mutation (gray dot). Adapted from @charlesworth_ElementsEvolutionaryGenetics_2010, Fig. 8.13"><img data-src="selection_files/figure-revealjs/fig-selective-sweep-1.svg" width="960" height="300"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selective-sweep-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: A selective sweep of an advantageous mutation (gray dot). Adapted from <span class="citation" data-cites="charlesworth_ElementsEvolutionaryGenetics_2010">Charlesworth &amp; Charlesworth (<a href="#/bibliography" role="doc-biblioref" onclick="">2010</a>)</span>, Fig. 8.13
</figcaption>
</figure>
</div>
</div>
</div>
<p>Example of a selective sweep. If a sweep completes at a locus, it will become monomorphic, as will the <em>neighbouring</em> sites. Mutation could reintroduce variation. Recombination could increase diversity in neighbourhood, but in a manner that depends on the <em>distance</em> from the locus under selection.</p>
<aside class="notes">
<p>Alternatives to direct selection tests are those that look at linked selection. In essence, we look for characteristic patterns in the vicinity of a selected locus. A sweep leads to the selected site and neigbouring sites being monomorphic (and therefore invisible to tests based on direct selection). Without recombination, mutation could reintroduce variation. In the presence of recombination, the effect of reduced diversity will be dependent on distance from the locus under selection.</p>
<p>Note here that every sequence can be treated as an allele, such that we in essence are looking multi- and not a bi-allelic locus.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity" class="slide level2 ultitle">
<h2>The effect of a selective sweep on diversity</h2>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a></a><span class="ex">pgip-slim</span> <span class="at">--seed</span> 42 <span class="at">-n</span> 1000 <span class="at">-r</span> 1e-6 <span class="at">-m</span> 1e-7 <span class="at">--threads</span> 12 recipes/slim/selective_sweep.slim <span class="at">-l</span> 1000000 <span class="at">--outdir</span> results/slim</span>
<span id="cb3-2"><a></a><span class="ex">pgip-tsstat</span> results/slim/slim<span class="pp">*</span>.trees <span class="at">-n</span> 10 <span class="at">--seed</span> 31 <span class="at">-s</span> pi <span class="at">-s</span> S <span class="at">-s</span> TajD <span class="at">-w</span> 500 <span class="at">--threads</span> 10 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-v</span> <span class="at">-</span> <span class="op">&gt;</span> results/slim/selective_sweep.w500.csv.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- markdownlint-enable MD013 -->
<!-- markdownlint-disable MD013 -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-pgip-slim-sweep-plot-tajd" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-pgip-slim-sweep-plot-tajd-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-tajd-1.gif" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;7: The effect of a selective sweep on diversity. The arrow points to the site under selection. The y-axis shows Tajima’s D which is proportional to the difference between two measures of diversity, nucleotide diversity \pi and Watterson’s \theta_W. "><img data-src="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-tajd-1.gif" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pgip-slim-sweep-plot-tajd-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: The effect of a selective sweep on diversity. The arrow points to the site under selection. The y-axis shows Tajima’s D which is proportional to the difference between two measures of diversity, nucleotide diversity <span class="math inline">\pi</span> and Watterson’s <span class="math inline">\theta_W</span>. <a href="../../recipes/slides/slim.html#sec-recipe-slim-selective-sweep"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a>
</figcaption>
</figure>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<aside class="notes">
<p>Simulation: N=1e4, mu=5e-8, pi=4<em>N</em>mu=0.004, seek S=a*pi=0.01, so a=S/pi which gives n.&nbsp;One needs to ramp up recombination rate to see the dip in TajD clearly.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity-1" class="slide level2 ultitle">
<h2>The effect of a selective sweep on diversity</h2>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a></a><span class="ex">pgip-slim</span> <span class="at">--seed</span> 42 <span class="at">-n</span> 1000 <span class="at">-r</span> 1e-6 <span class="at">-m</span> 1e-7 <span class="at">--threads</span> 12 recipes/slim/selective_sweep.slim <span class="at">-l</span> 1000000 <span class="at">--outdir</span> results/slim</span>
<span id="cb4-2"><a></a><span class="ex">pgip-tsstat</span> results/slim/slim<span class="pp">*</span>.trees <span class="at">-n</span> 10 <span class="at">--seed</span> 31 <span class="at">-s</span> pi <span class="at">-s</span> S <span class="at">-s</span> TajD <span class="at">-w</span> 500 <span class="at">--threads</span> 10 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-v</span> <span class="at">-</span> <span class="op">&gt;</span> results/slim/selective_sweep.w500.csv.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- markdownlint-enable MD013 -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-pgip-slim-sweep-plot-tajd-2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-pgip-slim-sweep-plot-tajd-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-tajd-2-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;8: The effect of a selective sweep on diversity. The arrow points to the site under selection. The y-axis shows Tajima’s D which is proportional to the difference between two measures of diversity, nucleotide diversity \pi and Watterson’s \theta_W. "><img data-src="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-tajd-2-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pgip-slim-sweep-plot-tajd-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: The effect of a selective sweep on diversity. The arrow points to the site under selection. The y-axis shows Tajima’s D which is proportional to the difference between two measures of diversity, nucleotide diversity <span class="math inline">\pi</span> and Watterson’s <span class="math inline">\theta_W</span>. <a href="../../recipes/slides/slim.html#sec-recipe-slim-selective-sweep"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a>
</figcaption>
</figure>
</div>
</div>
</div>
<aside class="notes">
<p>Important: individual runs show a lot of stochasticity, but the signal is clear for the <em>mean</em>.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity-2" class="slide level2 ultitle">
<h2>The effect of a selective sweep on diversity</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-pgip-slim-sweep-plot-pi-S" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-pgip-slim-sweep-plot-pi-S-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-pi-S-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;9: The effect of a selective sweep on diversity. The figure shows the mean of 1000 simulations with the selected locus indicated with an arrow."><img data-src="selection_files/figure-revealjs/fig-pgip-slim-sweep-plot-pi-S-1.svg" class="quarto-figure quarto-figure-center" width="600"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pgip-slim-sweep-plot-pi-S-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: The effect of a selective sweep on diversity. The figure shows the mean of 1000 simulations with the selected locus indicated with an arrow.
</figcaption>
</figure>
</div>
</div>
</div>
<aside class="notes">
<p>cf <span class="citation" data-cites="nielsen_MolecularSignaturesNatural_2005">Nielsen (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span>. Not clear how scaling of statistics to have mean=1 is done.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>

<section id="the-phases-of-a-selective-sweep" class="slide level2 ultitle">
<h2>The phases of a selective sweep</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-selective-sweep-phases" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-selective-sweep-phases-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-selective-sweep-phases-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Figure&nbsp;10: Time goes from left to right. As sweep progresses, tree topology changes. Adapted from @hahn_MolecularPopulationGenetics_2019, Figure 8.1"><img data-src="selection_files/figure-revealjs/fig-selective-sweep-phases-1.svg" width="960" height="400"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-selective-sweep-phases-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: Time goes from left to right. As sweep progresses, tree topology changes. Adapted from <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">Hahn (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span>, Figure 8.1
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column fragment" style="width:30%;">
<p>Amount of diversity depends on fixation time. A neutral locus fixes in <span class="math inline">4N_e</span> generations; for <span class="math inline">s=0.0001</span>, it takes approximately <span class="math inline">0.29N_e</span> generations.</p>
<p>Selections changes the genealogy (different topology, shorter branches), an aspect used in many linkage-based tests for selection.</p>
</div></div>
<aside class="notes">
<p>A selective sweep changes the genealogy (shorter branches, different <em>topology</em>) compared to the neutral case, a fact that can be utilised in linkage-based tests for selection.</p>
<p>Note that the amount of variation surrounding a sweep depends heavily on the <em>rate</em> of the sweep. <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 167</a>)</span> points out that an advantageous allele with <span class="math inline">s=0.0001</span> will take approximately <span class="math inline">0.29N_e</span> generations to fix (compared to <span class="math inline">4N_e</span> generations for a neutral locus)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linked-selection-may-constrain-levels-of-diversity" class="slide level2 ultitle">
<h2>Linked selection may constrain levels of diversity</h2>
<div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-hitchhiking-vs-background-selection" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-hitchhiking-vs-background-selection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="selection_files/figure-revealjs/fig-hitchhiking-vs-background-selection-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="Figure&nbsp;11: Hitchhiking (left) versus background selection (right)."><img data-src="selection_files/figure-revealjs/fig-hitchhiking-vs-background-selection-1.svg" width="1000"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-hitchhiking-vs-background-selection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11: Hitchhiking (left) versus background selection (right).
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<h4 id="hitchhiking">Hitchhiking</h4>
<ul>
<li>alleles linked to locus under selection “hitchhike” to high frequencies <span class="citation" data-cites="smith_HitchhikingEffectFavourable_1974">(<a href="#/bibliography" role="doc-biblioref" onclick="">Smith &amp; Haigh, 1974</a>)</span></li>
<li>evidence: positive correlation between putative neutral diversity and recombination <span class="citation" data-cites="corbett-detig_NaturalSelectionConstrains_2015">(<a href="#/bibliography" role="doc-biblioref" onclick="">Corbett-Detig et al., 2015</a>)</span></li>
</ul>
</div><div class="column" style="width:50%;">
<h4 id="background-selection">Background selection</h4>
<ul>
<li>loci linked to a deleterious locus will be purged from population and thus reduce diversity <span class="citation" data-cites="charlesworth_EffectDeleteriousMutations_1993">(<a href="#/bibliography" role="doc-biblioref" onclick="">Charlesworth et al., 1993</a>)</span></li>
<li>similar patterns to hitchhiking</li>
</ul>
</div></div>
<aside class="notes">
<p>Linked selection has been proposed as solution to the low range of diversity seen between species with orders of magnitude ranges in census population sizes.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="summary" class="slide level2 ultitle">
<h2>Summary</h2>
<p>We have looked at the <strong>Wright-Fisher model</strong> as a model of populations and <strong>genealogies</strong></p>
<p><strong>Genetic drift</strong> moves allele frequencies up and down at random and removes variation at rate <span class="math inline">\propto 1/2N</span></p>
<p><strong>Mutation</strong> reintroduces variation. The <strong>Neutral theory</strong> posits most mutations are neutral and dynamics follow <strong>mutation drift</strong> equilibrium.</p>
<p>Methods to detect <strong>selection</strong> are based on <strong>direct selection</strong> or studying patterns of variation caused by <strong>linked selection</strong>.</p>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller ultitle scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-charlesworth_ElementsEvolutionaryGenetics_2010" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Charlesworth, D. (2010). <em>Elements of <span>Evolutionary Genetics</span></em>. <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-charlesworth_EffectDeleteriousMutations_1993" class="csl-entry" role="listitem">
Charlesworth, B., Morgan, M. T., &amp; Charlesworth, D. (1993). <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1205596">The <span>Effect</span> of <span>Deleterious Mutations</span> on <span>Neutral Molecular Variation</span></a>. <em>Genetics</em>, <em>134</em>(4), 1289–1303.
</div>
<div id="ref-corbett-detig_NaturalSelectionConstrains_2015" class="csl-entry" role="listitem">
Corbett-Detig, R. B., Hartl, D. L., &amp; Sackton, T. B. (2015). Natural <span>Selection Constrains Neutral Diversity</span> across <span>A Wide Range</span> of <span>Species</span>. <em>PLOS Biology</em>, <em>13</em>(4), e1002112. <a href="https://doi.org/10.1371/journal.pbio.1002112">https://doi.org/10.1371/journal.pbio.1002112</a>
</div>
<div id="ref-gillespie_PopulationGeneticsConcise_2004" class="csl-entry" role="listitem">
Gillespie, J. H. (2004). <em>Population <span>Genetics</span>: <span>A Concise Guide</span></em> (2nd edition). <span>Johns Hopkins University Press</span>.
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-lynch_OriginsGenomeArchitecture_2007" class="csl-entry" role="listitem">
Lynch, M. (2007). <em>The origins of genome architecture</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-nielsen_MolecularSignaturesNatural_2005" class="csl-entry" role="listitem">
Nielsen, R. (2005). Molecular <span>Signatures</span> of <span>Natural Selection</span>. <em>Annual Review of Genetics</em>, <em>39</em>(1), 197–218. <a href="https://doi.org/10.1146/annurev.genet.39.073003.112420">https://doi.org/10.1146/annurev.genet.39.073003.112420</a>
</div>
<div id="ref-smith_HitchhikingEffectFavourable_1974" class="csl-entry" role="listitem">
Smith, J. M., &amp; Haigh, J. (1974). The hitch-hiking effect of a favourable gene. <em>Genetics Research</em>, <em>23</em>(1), 23–35. <a href="https://doi.org/10.1017/S0016672300014634">https://doi.org/10.1017/S0016672300014634</a>
</div>
</div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Selection</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>