<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-541d1840fe0299ec4c9bf9ceaf6b9fb6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Per Unneberg">
  <title>Population Genomics in Practice 2025 – The Wright-Fisher model</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-dd75e976e0414ec343b1912506d158bc.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">The Wright-Fisher model</h1>
  <p class="subtitle">Of population models and genealogies</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Per Unneberg 
</div>
</div>
</div>

</section>

<section id="models-of-populations" class="slide level2 ultitle">
<h2>Models of populations</h2>
<div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wright-fisher-drift-models-of-population-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="wrightfisher_files/figure-revealjs/wright-fisher-drift-models-of-population-1.svg" class="quarto-figure quarto-figure-center" style="width:30.0%"></a></p>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Segue: take a step back to the days before the nature of genetic inheritance was known. First focus on the population.</p>
<p>Goal of lecture is also to understand genetic relationships in the context of genealogies -&gt; coalescent.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li class="fragment">discrete and non-overlapping generations</li>
<li class="fragment">haploid individuals or two subpopulations (males and females)</li>
<li class="fragment">constant population size</li>
<li class="fragment">all individuals are equally fit</li>
<li class="fragment">population has no geographical or social structure</li>
<li class="fragment">no recombination</li>
</ul>
</div><div class="column" style="width:50%;">

</div></div>
<aside class="notes">
<p>See <span class="citation" data-cites="hein_GeneGenealogies_2004">Hein et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2004</a>)</span> for more assumptions. The second assumption means we can use 2N chromosome interchangeably for haploid (n=2N) and diploid (n=N) populations.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model-1" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-1-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img data-src="wrightfisher_files/figure-revealjs/wf-1-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-2" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-1">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-2-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img data-src="wrightfisher_files/figure-revealjs/wf-2-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-3" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-2">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3.0-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img data-src="wrightfisher_files/figure-revealjs/wf-3.0-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-4" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-3">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3.1-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img data-src="wrightfisher_files/figure-revealjs/wf-3.1-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-5" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-4">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3.1b-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="wrightfisher_files/figure-revealjs/wf-3.1b-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-6" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-5">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3.2-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="wrightfisher_files/figure-revealjs/wf-3.2-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-7" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-6">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img data-src="wrightfisher_files/figure-revealjs/wf-3-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-8" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-7">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-3-untangled-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img data-src="wrightfisher_files/figure-revealjs/wf-3-untangled-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-9" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 class="ultitle" id="algorithm-8">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-4-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img data-src="wrightfisher_files/figure-revealjs/wf-4-1.svg" class="quarto-figure quarto-figure-center" style="width:90.0%"></a></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="wright-fisher-model-10" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-wright-fisher-model-latex-animate" class="quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-wright-fisher-model-latex-animate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="wrightfisher_files/figure-revealjs/fig-wright-fisher-model-latex-animate-.gif" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="Figure&nbsp;1: Wright-Fisher model"><img data-src="wrightfisher_files/figure-revealjs/fig-wright-fisher-model-latex-animate-.gif" style="width:40.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-wright-fisher-model-latex-animate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Wright-Fisher model
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>See <a href="https://mikeyharper.uk/animated-plots-with-r/" class="uri">https://mikeyharper.uk/animated-plots-with-r/</a></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model-11" class="slide level2 smaller ultitle">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-wright-fisher-model-gganimation-last" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-wright-fisher-model-gganimation-last-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="wrightfisher_files/figure-revealjs/fig-wright-fisher-model-gganimation-last-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;2: WF model indicating time direction from past (top) to present (bottom)."><img data-src="wrightfisher_files/figure-revealjs/fig-wright-fisher-model-gganimation-last-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-wright-fisher-model-gganimation-last-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: WF model indicating time direction from past (top) to present (bottom).
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column fragment" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy-all" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-wf-model-genealogy-all-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="wrightfisher_files/figure-revealjs/fig-wf-model-genealogy-all-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Figure&nbsp;3: WF model tracing the genealogies of all extant chromosomes"><img data-src="wrightfisher_files/figure-revealjs/fig-wf-model-genealogy-all-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-wf-model-genealogy-all-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: WF model tracing the genealogies of all extant chromosomes
</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-wf-model-genealogy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="wrightfisher_files/figure-revealjs/fig-wf-model-genealogy-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="Figure&nbsp;4: WF model tracing the genealogies of three extant chromosomes"><img data-src="wrightfisher_files/figure-revealjs/fig-wf-model-genealogy-1.svg" class="quarto-figure quarto-figure-center" style="width:80.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-wf-model-genealogy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: WF model tracing the genealogies of three extant chromosomes
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div></div>
<div class="columns">
<div class="column" style="width:70%;">
<div class="fragment">
<h5 id="observations">Observations</h5>
<ul>
<li>most lineages are lost over time</li>
<li>looking backwards in time genes eventually <strong>coalesce</strong> at a common ancestor</li>
<li>looking backwards in time sampled genes can be described by a <strong>genealogy</strong></li>
</ul>
</div>
</div><div class="column fragment" style="width:30%;">
<div class="center">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-sampled-genealogy-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img data-src="wrightfisher_files/figure-revealjs/wf-sampled-genealogy-1.svg" width="960" height="150"></a></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
</section>
<section id="the-wright-fisher-sampling-model" class="slide level2 ultitle">
<h2>The Wright-Fisher sampling model</h2>
<p>Let’s formalise the sampling process of the Wright-Fisher model<sup>1</sup>. We assume</p>
<ol type="1">
<li>a single locus in a haploid population of size <span class="math inline">2N</span> (or diploid of size <span class="math inline">N</span> when random mating)</li>
<li>no mutation and selection</li>
<li>discrete generations</li>
</ol>
<p>Each generation we sample <span class="math inline">2N</span> new chromosomes from the previous generation. The probability of choosing a chromosome <span class="math inline">v</span> is <span class="math inline">1/2N</span> (coin flip with probability of success <span class="math inline">1/2N</span>). Since the trials are independent, and we perform <span class="math inline">2N</span> trials, the number of offspring <span class="math inline">k</span> of a given chromosome <span class="math inline">v</span> is <a href="https://en.wikipedia.org/wiki/Binomial_distribution"><em>binomially distributed</em></a> <span class="math inline">\mathrm{Bin}(m, p)</span>, with parameters <span class="math inline">m=2N</span> and probability of success <span class="math inline">p=\frac{1}{2N}</span>.</p>
<aside class="notes">
<p>We assume some familiarity with Binomial sampling. For reference, some mathematical formality follows.</p>
<p>Let <span class="math inline">p</span> be the probability of heads (=1) of a coin flip, and <span class="math inline">q=1-p</span> the probability of tails (=0). A coin flip can be modelled by a random variable (r.v.) <span class="math inline">X</span> that follows a <em>Bernoulli distribution</em>, where <span class="math inline">\mathrm{Pr(}X=1\mathrm{)}=p</span>, <span class="math inline">\mathrm{Pr(}X=0\mathrm{)}=1-p=q</span>. One can show (e.g., with transforms such as the <a href="https://en.wikipedia.org/wiki/Probability-generating_function">probability-generating function</a>) that a binomially distributed variable <span class="math inline">Y\sim Bin(n, p)</span> is the sum of <span class="math inline">n</span> independent Bernoulli variables <span class="math inline">X</span> (“Bernoulli trials”), or <span class="math inline">Y=\sum^nX</span>. In our case, there are <span class="math inline">2N</span> trials, each with <span class="math inline">p=1/2N</span>.</p>
<p><span class="math display">
P(v=k) = {2N\choose k}\left( \frac{1}{2N} \right)^k
\left(1 - \frac{1}{2N} \right)^{(2N - k)}
</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<aside><ol class="aside-footnotes"><li id="fn1"><p>For a more extensive treatment, see <span class="citation" data-cites="hermisson_MathematicalPopulationGenetics_2017">Hermisson (<a href="#/bibliography" role="doc-biblioref" onclick="">2017</a>)</span> or <span class="citation" data-cites="hein2005gene">Hein et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span></p></li></ol></aside></section>
<section id="properties-of-wright-fisher-sampling" class="slide level2 smaller ultitle">
<h2>Properties of Wright-Fisher sampling</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="fragment">
<h5 id="the-expected-number-of-offspring-is-one">The expected number of offspring is one</h5>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-expected-number-of-offspring-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img data-src="wrightfisher_files/figure-revealjs/wf-expected-number-of-offspring-1.svg" class="quarto-figure quarto-figure-center" style="width:70.0%"></a></p>
</figure>
</div>
</div>
</div>
</div>
<div class="fragment">
<h5 id="poisson-approximation-for-large-n">Poisson approximation for large <span class="math inline">N</span></h5>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-poisson-approximation-large-n-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img data-src="wrightfisher_files/figure-revealjs/wf-poisson-approximation-large-n-1.svg" class="quarto-figure quarto-figure-center" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p><span class="math display">
P(v=k) \approx \frac{1}{k!}e^{-k}
</span></p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<h5 id="sec-prob-sequence-pick-same-parent">Prob(pick same parent) = 1/2N</h5>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-prob-parents-identical-fig-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img data-src="wrightfisher_files/figure-revealjs/wf-prob-parents-identical-fig-1.svg" class="quarto-figure quarto-figure-center" width="800"></a></p>
</figure>
</div>
</div>
</div>
</div>
<div class="fragment">
<h5 id="time-for-two-sequences-to-coalesce-sim-12n">Time for two sequences to coalesce <span class="math inline">\sim 1/2N</span></h5>
<div class="columns">
<div class="column" style="width:15%;">

</div><div class="column" style="width:70%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="wrightfisher_files/figure-revealjs/wf-two-chrom-find-common-ancestor-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img data-src="wrightfisher_files/figure-revealjs/wf-two-chrom-find-common-ancestor-1.svg" class="quarto-figure quarto-figure-center" style="width:100.0%"></a></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:15%;">

</div></div>
</div>
</div></div>
<aside class="notes">
<p>Many of the results above follow from standard results in probability theory and are provided here in brevity for reference to the interested reader. The details are not necessary to understand in detail.</p>
<h4 id="expected-number-of-offspring">Expected number of offspring</h4>
<p>The expected value of a <span class="math inline">\mathrm{Bin}(m,p)</span> (binomially distributed) variable with parameters <span class="math inline">m=2N</span> and <span class="math inline">p=\frac{1}{2N}</span> is <span class="math inline">mp</span>, hence:</p>
<p><span class="math display">
E(v) = mp = 2N\frac{1}{2N} = 1
</span></p>
<h4 id="poisson-approximation">Poisson approximation</h4>
<p>When <span class="math inline">2N</span> large it holds that the probability that sequence <span class="math inline">v</span> has <span class="math inline">k</span> offspring is</p>
<p><span class="math display">
P(v=k) \approx \frac{1}{k!}e^{-k}
</span></p>
<p>Since <span class="math inline">P(v=0) = e^{-1} \approx 0.37</span>, a fraction 0.37 of sequences lack descendants</p>
<h4 id="probability-that-two-sequences-pick-same-parent">Probability that two sequences pick same parent</h4>
<p>Let <span class="math inline">u</span> and <span class="math inline">v</span> be two sequences. Pick a parent <span class="math inline">p_u</span> of <span class="math inline">u</span>. Then the probability that <span class="math inline">v</span> picks the same parent is (solid lines)</p>
<p><span class="math display">
P(p_u=p_v) = \frac{1}{2N}
</span></p>
<p>The probability that they pick different parents is</p>
<p><span class="math display">
P(p_u \neq p_v) = 1 - \frac{1}{2N}
</span></p>
<h5 id="time-for-two-sequences-to-coalescent">Time for two sequences to coalescent</h5>
<p>Time for two sequences <span class="math inline">u</span> and <span class="math inline">v</span> to find common ancestor distributed as <span class="math inline">(1 - \frac{1}{2N})^{j-1}\frac{1}{2N}</span> (<span class="math inline">j-1</span> failures followed by success). This is the <a href="https://en.wikipedia.org/wiki/Geometric_distribution">geometric distribution</a> <span class="math inline">Ge(p)</span>, with parameter <span class="math inline">p=\frac{1}{2N}</span>, and expected value <span class="math inline">\frac{1}{p}={2N}</span>. That is, the <strong>expected number of generations for two sequences to find a common ancestor (i.e., coalesce) is <span class="math inline">2N</span> generations</strong>.</p>
<p>Since a large fraction of genes lack descendants, very quickly (compared to 2N) a population will descend from a small proportion of genes.</p>
<p>Derivation of average time to having same parent (i.e., coalescence) requires knowledge of geometric distribution. The probability that the genes find common parent <span class="math inline">j</span> generations ago is</p>
<p><span class="math display">
(1-\frac{1}{2N})^{(j-1)}\frac{1}{2N}
</span></p>
<p>due to independence between generations. This is the geometric distribution <span class="math inline">Ge(p)</span> with <span class="math inline">p=1/2N</span>, which has expected value <span class="math inline">1/p</span>, i.e., 2N in this case.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller ultitle scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-hein2005gene" class="csl-entry" role="listitem">
Hein, J., Schierup, M. H., &amp; Wiuf, C. (2005). <em>Gene genealogies, variation and evolution: <span>A</span> primer in coalescent theory</em>. <span>Oxford University Press</span>. <a href="https://books.google.se/books?id=CCmLNAEACAAJ">https://books.google.se/books?id=CCmLNAEACAAJ</a>
</div>
<div id="ref-hein_GeneGenealogies_2004" class="csl-entry" role="listitem">
Hein, J., Schierup, M., &amp; Wiuf, C. (2004). Gene genealogies, variation and evolution. A primer in coalescent theory. In <em>Systematic Biology - SYST BIOL</em> (Vol. 54).
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2017" class="csl-entry" role="listitem">
Hermisson, J. (2017). <em>Mathematical population genetics</em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017</a>
</div>
</div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-pgip-H25"><img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/nbislogo-green-teal.svg" style="height:50px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<img data-src="../../site_libs/quarto-contrib/quarto-project/percyfal/nbis-quarto/assets/logo/SciLifeLab_symbol_green.png" style="height:50px;max-width:none;">
</div>
</div>
</section>

    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>Wright-Fisher model</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip-H25\/");
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>